// lemma-sdk/dist/openapi_client/services/AgentRuntimeService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class AgentRuntimeService {
    /**
     * List Available Agent Harnesses
     * @returns AgentHarnessListResponse Successful Response
     * @throws ApiError
     */
    static agentRuntimeHarnessesList() {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/agent-runtime/harnesses',
        });
    }
    /**
     * List Available Agent Runtime Profiles
     * @param orgId
     * @returns AgentRuntimeProfileListResponse Successful Response
     * @throws ApiError
     */
    static agentRuntimeProfilesList(orgId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{org_id}/agent-runtime/profiles',
            path: {
                'org_id': orgId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Agent Runtime Profile
     * @param orgId
     * @param requestBody
     * @returns AgentRuntimeProfileResponse Successful Response
     * @throws ApiError
     */
    static agentRuntimeProfilesCreate(orgId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/{org_id}/agent-runtime/profiles',
            path: {
                'org_id': orgId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
