// lemma-sdk/dist/openapi_client/services/AgentSurfacesIngressService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class AgentSurfacesIngressService {
    /**
     * Teams Admin Consent Callback
     * @param tenant
     * @param adminConsent
     * @param state
     * @param error
     * @param errorDescription
     * @returns any Successful Response
     * @throws ApiError
     */
    static agentSurfaceTeamsAdminConsentCallback(tenant, adminConsent, state, error, errorDescription) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/surfaces/teams/admin-consent/callback',
            query: {
                'tenant': tenant,
                'admin_consent': adminConsent,
                'state': state,
                'error': error,
                'error_description': errorDescription,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Verify surface webhook using the platform callback URL
     * Webhook verification endpoint for platforms that require it.
     * @param platform
     * @returns any Successful Response
     * @throws ApiError
     */
    static surfaceWebhookVerify(platform) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/surfaces/webhooks/{platform}',
            path: {
                'platform': platform,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Handle platform-level surface webhook
     * Handle platform-level webhook callbacks.
     * @param platform
     * @returns any Successful Response
     * @throws ApiError
     */
    static surfaceWebhookHandlePlatform(platform) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/surfaces/webhooks/{platform}',
            path: {
                'platform': platform,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Verify surface webhook using a surface-level callback URL
     * Webhook verification endpoint for platforms that require it.
     * @param surfaceId
     * @returns any Successful Response
     * @throws ApiError
     */
    static surfaceWebhookVerifySurface(surfaceId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/surfaces/{surface_id}/webhook',
            path: {
                'surface_id': surfaceId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Handle surface-level webhook
     * Handle webhooks addressed to one concrete surface.
     * @param surfaceId
     * @returns any Successful Response
     * @throws ApiError
     */
    static surfaceWebhookHandleSurface(surfaceId) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/surfaces/{surface_id}/webhook',
            path: {
                'surface_id': surfaceId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
