// lemma-sdk/dist/openapi_client/services/AgentSurfacesService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class AgentSurfacesService {
    /**
     * List Surfaces
     * @param podId
     * @param limit
     * @param pageToken
     * @returns AgentSurfaceListResponse Successful Response
     * @throws ApiError
     */
    static agentSurfaceList(podId, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/surfaces',
            path: {
                'pod_id': podId,
            },
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete Surface
     * @param podId
     * @param platform
     * @returns void
     * @throws ApiError
     */
    static agentSurfaceDelete(podId, platform) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/pods/{pod_id}/surfaces/{platform}',
            path: {
                'pod_id': podId,
                'platform': platform,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Surface
     * @param podId
     * @param platform
     * @returns any Successful Response
     * @throws ApiError
     */
    static agentSurfaceGet(podId, platform) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/surfaces/{platform}',
            path: {
                'pod_id': podId,
                'platform': platform,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Upsert Surface
     * Create the surface for a platform, or merge updates into the existing one.
     *
     * A surface is unique per ``pod_id + platform``, so this single idempotent
     * write covers create, config edits, channel routing, account/credential
     * changes, and enable/disable. Only fields present in the request are applied
     * on update.
     * @param podId
     * @param platform
     * @param requestBody
     * @returns any Successful Response
     * @throws ApiError
     */
    static agentSurfaceUpsert(podId, platform, requestBody) {
        return __request(OpenAPI, {
            method: 'PUT',
            url: '/pods/{pod_id}/surfaces/{platform}',
            path: {
                'pod_id': podId,
                'platform': platform,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List Surface Channels
     * List the channels/groups this surface bot can be configured to respond in.
     *
     * Returns an empty list for platforms without an enumerable channel concept
     * (Telegram groups, WhatsApp, email).
     * @param podId
     * @param platform
     * @returns AvailableSurfaceChannelsResponse Successful Response
     * @throws ApiError
     */
    static agentSurfaceChannels(podId, platform) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/surfaces/{platform}/channels',
            path: {
                'pod_id': podId,
                'platform': platform,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Surface Setup
     * Everything needed to finish setting up this platform's surface.
     *
     * Merges the static platform checklist with live webhook + admin-consent
     * state. Works before the surface exists (guide only) and after (live state).
     * @param podId
     * @param platform
     * @returns SurfaceSetupResponse Successful Response
     * @throws ApiError
     */
    static agentSurfaceSetup(podId, platform) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/surfaces/{platform}/setup',
            path: {
                'pod_id': podId,
                'platform': platform,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
