// lemma-sdk/dist/openapi_client/services/AgentToolsService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class AgentToolsService {
    /**
     * Agent Connector Helper Agent
     * Plan how to use one or more connectors for a goal and return recommended operations.
     * @param requestBody
     * @returns ConnectorHelperAgentResponse Successful Response
     * @throws ApiError
     */
    static agentToolConnectorHelperAgent(requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/tools/connector-helper-agent',
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Agent Report Feedback
     * Record a maintainer-facing feedback report about system issues, skill issues, incorrect knowledge, or other unexpected behavior.
     * @param requestBody
     * @returns ReportFeedbackResponse Successful Response
     * @throws ApiError
     */
    static agentToolReportFeedback(requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/tools/report-feedback',
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Agent Web Search
     * Run a raw web search and return structured results.
     * @param requestBody
     * @returns WebSearchResponse Successful Response
     * @throws ApiError
     */
    static agentToolWebSearch(requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/tools/web-search',
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
