// lemma-sdk/dist/openapi_client/services/AgentsService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class AgentsService {
    /**
     * List Agents
     * List pod-owned agent definitions visible to the current user.
     * @param podId
     * @param pageToken
     * @param limit
     * @returns AgentListResponse Successful Response
     * @throws ApiError
     */
    static agentList(podId, pageToken, limit = 100) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/agents',
            path: {
                'pod_id': podId,
            },
            query: {
                'page_token': pageToken,
                'limit': limit,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Agent
     * Create a pod-owned agent definition with runtime, toolsets, and schemas.
     * @param podId
     * @param requestBody
     * @returns AgentActionResponse Successful Response
     * @throws ApiError
     */
    static agentCreate(podId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/agents',
            path: {
                'pod_id': podId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete Agent
     * Delete a pod-owned agent definition by name.
     * @param podId
     * @param agentName
     * @returns AgentMessageResponse Successful Response
     * @throws ApiError
     */
    static agentDelete(podId, agentName) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/pods/{pod_id}/agents/{agent_name}',
            path: {
                'pod_id': podId,
                'agent_name': agentName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Agent
     * Get one pod-owned agent definition by its stable name.
     * @param podId
     * @param agentName
     * @returns AgentDetailResponse Successful Response
     * @throws ApiError
     */
    static agentGet(podId, agentName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/agents/{agent_name}',
            path: {
                'pod_id': podId,
                'agent_name': agentName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Update Agent
     * Update an agent definition, including prompt instruction, runtime, toolsets, and schemas.
     * @param podId
     * @param agentName
     * @param requestBody
     * @returns AgentActionResponse Successful Response
     * @throws ApiError
     */
    static agentUpdate(podId, agentName, requestBody) {
        return __request(OpenAPI, {
            method: 'PATCH',
            url: '/pods/{pod_id}/agents/{agent_name}',
            path: {
                'pod_id': podId,
                'agent_name': agentName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Agent Resource Permissions
     * Get explicit resource grants assigned to an agent.
     * @param podId
     * @param agentName
     * @returns AgentPermissionsResponse Successful Response
     * @throws ApiError
     */
    static agentPermissionsGet(podId, agentName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/agents/{agent_name}/permissions',
            path: {
                'pod_id': podId,
                'agent_name': agentName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Replace Agent Resource Permissions
     * Replace explicit resource grants assigned to an agent.
     * @param podId
     * @param agentName
     * @param requestBody
     * @returns AgentPermissionsResponse Successful Response
     * @throws ApiError
     */
    static agentPermissionsReplace(podId, agentName, requestBody) {
        return __request(OpenAPI, {
            method: 'PUT',
            url: '/pods/{pod_id}/agents/{agent_name}/permissions',
            path: {
                'pod_id': podId,
                'agent_name': agentName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
