// lemma-sdk/dist/openapi_client/services/AppsService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class AppsService {
    /**
     * List Apps
     * @param podId
     * @param limit
     * @param pageToken
     * @returns AppListResponse Successful Response
     * @throws ApiError
     */
    static appList(podId, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/apps',
            path: {
                'pod_id': podId,
            },
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create App
     * @param podId
     * @param requestBody
     * @returns AppDetailResponse Successful Response
     * @throws ApiError
     */
    static appCreate(podId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/apps',
            path: {
                'pod_id': podId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Save Widget As App
     * Promote a conversation widget into a persisted app.
     *
     * The widget and the app are the same artifact at two lifecycle stages: this
     * fetches the widget's stored HTML and deploys it as the app's bundle —
     * identical to what was shown.
     * @param podId
     * @param requestBody
     * @returns AppDetailResponse Successful Response
     * @throws ApiError
     */
    static appCreateFromWidget(podId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/apps/from-widget',
            path: {
                'pod_id': podId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete App
     * @param podId
     * @param appName
     * @returns AppMessageResponse Successful Response
     * @throws ApiError
     */
    static appDelete(podId, appName) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/pods/{pod_id}/apps/{app_name}',
            path: {
                'pod_id': podId,
                'app_name': appName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get App
     * @param podId
     * @param appName
     * @returns AppDetailResponse Successful Response
     * @throws ApiError
     */
    static appGet(podId, appName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/apps/{app_name}',
            path: {
                'pod_id': podId,
                'app_name': appName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Update App
     * @param podId
     * @param appName
     * @param requestBody
     * @returns AppDetailResponse Successful Response
     * @throws ApiError
     */
    static appUpdate(podId, appName, requestBody) {
        return __request(OpenAPI, {
            method: 'PATCH',
            url: '/pods/{pod_id}/apps/{app_name}',
            path: {
                'pod_id': podId,
                'app_name': appName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get App Root Asset
     * @param podId
     * @param appName
     * @returns any Successful Response
     * @throws ApiError
     */
    static appAssetRootGet(podId, appName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/apps/{app_name}/assets',
            path: {
                'pod_id': podId,
                'app_name': appName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get App Asset
     * @param podId
     * @param appName
     * @param assetPath
     * @returns any Successful Response
     * @throws ApiError
     */
    static appAssetGet(podId, appName, assetPath) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/apps/{app_name}/assets/{asset_path}',
            path: {
                'pod_id': podId,
                'app_name': appName,
                'asset_path': assetPath,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Upload App Bundle
     * @param podId
     * @param appName
     * @param formData
     * @returns AppBundleUploadResponse Successful Response
     * @throws ApiError
     */
    static appBundleUpload(podId, appName, formData) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/apps/{app_name}/bundle',
            path: {
                'pod_id': podId,
                'app_name': appName,
            },
            formData: formData,
            mediaType: 'multipart/form-data',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Download App Dist Archive
     * @param podId
     * @param appName
     * @returns binary Zip archive bytes
     * @throws ApiError
     */
    static appDistArchiveGet(podId, appName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/apps/{app_name}/dist/archive',
            path: {
                'pod_id': podId,
                'app_name': appName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Download App Source Archive
     * @param podId
     * @param appName
     * @returns binary Zip archive bytes
     * @throws ApiError
     */
    static appSourceArchiveGet(podId, appName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/apps/{app_name}/source/archive',
            path: {
                'pod_id': podId,
                'app_name': appName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
