// lemma-sdk/dist/openapi_client/services/ConnectorsService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class ConnectorsService {
    /**
     * List Connectors
     * Get all active connectors available for connector
     * @param limit
     * @param pageToken
     * @returns ConnectorListResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorList(limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/connectors',
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * OAuth Callback
     * Handle OAuth callback and complete account connection. This endpoint is public and uses state parameter for security.
     * @param error
     * @param format
     * @returns string Successful Response
     * @throws ApiError
     */
    static connectorOauthCallback(error, format) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/connectors/connect-requests/oauth/callback',
            query: {
                'error': error,
                'format': format,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Connector
     * Get a specific connector by ID along with its operation catalog
     * @param connectorId
     * @returns ConnectorDetailResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorGet(connectorId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/connectors/{connector_id}',
            path: {
                'connector_id': connectorId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Connector Skill
     * Get the skill guide markdown for a connector. Pass `provider=lemma` or `provider=composio` to get provider-specific instructions when the app supports both. Falls back to the generic doc if no provider-specific file exists. Returns 404 if no skill doc has been generated yet.
     * @param connectorId
     * @param provider Provider override: lemma or composio
     * @returns ConnectorSkillResponse Successful Response
     * @throws ApiError
     */
    static connectorSkillGet(connectorId, provider) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/connectors/{connector_id}/skill',
            path: {
                'connector_id': connectorId,
            },
            query: {
                'provider': provider,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List Accounts
     * Get all connected accounts for the current user. Optionally filter by connector_id or connector_name
     * @param organizationId
     * @param connectorId
     * @param limit
     * @param pageToken
     * @returns AccountListResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorAccountList(organizationId, connectorId, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/accounts',
            path: {
                'organization_id': organizationId,
            },
            query: {
                'connector_id': connectorId,
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Account
     * Directly connect a credential-managed native account for an org auth config.
     * @param organizationId
     * @param requestBody
     * @returns AccountResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorAccountCreate(organizationId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/{organization_id}/connectors/accounts',
            path: {
                'organization_id': organizationId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete Account
     * Delete a connected account and revoke the connection
     * @param organizationId
     * @param accountId
     * @returns MessageResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorAccountDelete(organizationId, accountId) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/organizations/{organization_id}/connectors/accounts/{account_id}',
            path: {
                'organization_id': organizationId,
                'account_id': accountId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Account
     * Get a specific account by ID
     * @param organizationId
     * @param accountId
     * @returns AccountResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorAccountGet(organizationId, accountId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/accounts/{account_id}',
            path: {
                'organization_id': organizationId,
                'account_id': accountId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Credentials
     * Get the credentials for a specific account
     * @param organizationId
     * @param accountId
     * @returns AccountCredentialsResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorAccountCredentialsGet(organizationId, accountId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/accounts/{account_id}/credentials',
            path: {
                'organization_id': organizationId,
                'account_id': accountId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List Auth Configs
     * @param organizationId
     * @param limit
     * @param pageToken
     * @returns AuthConfigListResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorAuthConfigList(organizationId, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/auth-configs',
            path: {
                'organization_id': organizationId,
            },
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Auth Config
     * @param organizationId
     * @param requestBody
     * @returns AuthConfigResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorAuthConfigCreate(organizationId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/{organization_id}/connectors/auth-configs',
            path: {
                'organization_id': organizationId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete Auth Config
     * @param organizationId
     * @param authConfigName
     * @returns boolean Successful Response
     * @throws ApiError
     */
    static connectorAuthConfigDelete(organizationId, authConfigName) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/organizations/{organization_id}/connectors/auth-configs/{auth_config_name}',
            path: {
                'organization_id': organizationId,
                'auth_config_name': authConfigName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Auth Config
     * @param organizationId
     * @param authConfigName
     * @returns AuthConfigResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorAuthConfigGet(organizationId, authConfigName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/auth-configs/{auth_config_name}',
            path: {
                'organization_id': organizationId,
                'auth_config_name': authConfigName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Initiate Connect Request
     * Initiate an OAuth connection request for a connector
     * @param organizationId
     * @param requestBody
     * @returns ConnectRequestResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorConnectRequestCreate(organizationId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/{organization_id}/connectors/connect-requests',
            path: {
                'organization_id': organizationId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Connector Status
     * @param organizationId
     * @returns ConnectorStatusResponse Successful Response
     * @throws ApiError
     */
    static connectorStatusGet(organizationId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/status',
            path: {
                'organization_id': organizationId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Discover Connector Operations
     * @param organizationId
     * @param authConfigName
     * @param query
     * @param limit
     * @returns OperationDiscoverResponse Successful Response
     * @throws ApiError
     */
    static connectorOperationDiscover(organizationId, authConfigName, query, limit = 100) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/{auth_config_name}/operations',
            path: {
                'organization_id': organizationId,
                'auth_config_name': authConfigName,
            },
            query: {
                'query': query,
                'limit': limit,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Connector Operation Details In Batch
     * @param organizationId
     * @param authConfigName
     * @param requestBody
     * @returns OperationDetailsBatchResponse Successful Response
     * @throws ApiError
     */
    static connectorOperationDetailsBatch(organizationId, authConfigName, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/{organization_id}/connectors/{auth_config_name}/operations/details',
            path: {
                'organization_id': organizationId,
                'auth_config_name': authConfigName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Connector Operation Details
     * @param organizationId
     * @param authConfigName
     * @param operationName
     * @returns OperationDetail Successful Response
     * @throws ApiError
     */
    static connectorOperationDetail(organizationId, authConfigName, operationName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/{auth_config_name}/operations/{operation_name}',
            path: {
                'organization_id': organizationId,
                'auth_config_name': authConfigName,
                'operation_name': operationName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Execute Connector Operation
     * @param organizationId
     * @param authConfigName
     * @param operationName
     * @param requestBody
     * @returns OperationExecutionResponse Successful Response
     * @throws ApiError
     */
    static connectorOperationExecute(organizationId, authConfigName, operationName, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/{organization_id}/connectors/{auth_config_name}/operations/{operation_name}/execute',
            path: {
                'organization_id': organizationId,
                'auth_config_name': authConfigName,
                'operation_name': operationName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List Connector Triggers
     * @param organizationId
     * @param authConfigName
     * @param search
     * @param limit
     * @returns AppTriggerListResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorTriggerList(organizationId, authConfigName, search, limit = 100) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/{auth_config_name}/triggers',
            path: {
                'organization_id': organizationId,
                'auth_config_name': authConfigName,
            },
            query: {
                'search': search,
                'limit': limit,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Connector Trigger
     * @param organizationId
     * @param authConfigName
     * @param triggerName
     * @returns AppTriggerResponseSchema Successful Response
     * @throws ApiError
     */
    static connectorTriggerGet(organizationId, authConfigName, triggerName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{organization_id}/connectors/{auth_config_name}/triggers/{trigger_name}',
            path: {
                'organization_id': organizationId,
                'auth_config_name': authConfigName,
                'trigger_name': triggerName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
