// lemma-sdk/dist/openapi_client/services/FilesService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class FilesService {
    /**
     * List Files
     * @param podId
     * @param directoryPath
     * @param limit
     * @param pageToken
     * @returns FileListResponse Successful Response
     * @throws ApiError
     */
    static fileList(podId, directoryPath = '/', limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/datastore/files',
            path: {
                'pod_id': podId,
            },
            query: {
                'directory_path': directoryPath,
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Upload File
     * @param podId
     * @param formData
     * @returns FileDetailResponse Successful Response
     * @throws ApiError
     */
    static fileUpload(podId, formData) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/datastore/files',
            path: {
                'pod_id': podId,
            },
            formData: formData,
            mediaType: 'multipart/form-data',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete File Or Folder
     * @param podId
     * @param path
     * @returns void
     * @throws ApiError
     */
    static fileDelete(podId, path) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/pods/{pod_id}/datastore/files/by-path',
            path: {
                'pod_id': podId,
            },
            query: {
                'path': path,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get File
     * @param podId
     * @param path
     * @returns FileDetailResponse Successful Response
     * @throws ApiError
     */
    static fileGet(podId, path) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/datastore/files/by-path',
            path: {
                'pod_id': podId,
            },
            query: {
                'path': path,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Update File
     * @param podId
     * @param formData
     * @returns FileDetailResponse Successful Response
     * @throws ApiError
     */
    static fileUpdate(podId, formData) {
        return __request(OpenAPI, {
            method: 'PATCH',
            url: '/pods/{pod_id}/datastore/files/by-path',
            path: {
                'pod_id': podId,
            },
            formData: formData,
            mediaType: 'multipart/form-data',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List a document's derived child files
     * @param podId
     * @param path
     * @returns FileChildrenResponse Successful Response
     * @throws ApiError
     */
    static fileChildrenList(podId, path) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/datastore/files/children',
            path: {
                'pod_id': podId,
            },
            query: {
                'path': path,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Fetch a document's child artifact by path
     * @param podId
     * @param path Child path, e.g. /folder/report.pdf/document.md, /folder/report.pdf/image_0.png, or /folder/report.pdf/pages/page_0001.jpg
     * @param pageStart
     * @param pageEnd
     * @returns binary File bytes
     * @throws ApiError
     */
    static fileChildGet(podId, path, pageStart, pageEnd) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/datastore/files/children/content',
            path: {
                'pod_id': podId,
            },
            query: {
                'path': path,
                'page_start': pageStart,
                'page_end': pageEnd,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Download File
     * @param podId
     * @param path
     * @returns binary File bytes
     * @throws ApiError
     */
    static fileDownload(podId, path) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/datastore/files/download',
            path: {
                'pod_id': podId,
            },
            query: {
                'path': path,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Folder
     * @param podId
     * @param requestBody
     * @returns FileDetailResponse Successful Response
     * @throws ApiError
     */
    static fileFolderCreate(podId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/datastore/files/folders',
            path: {
                'pod_id': podId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Search Files
     * @param podId
     * @param requestBody
     * @returns FileSearchResponse Successful Response
     * @throws ApiError
     */
    static fileSearch(podId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/datastore/files/search',
            path: {
                'pod_id': podId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create a public, hit-capped signed URL for a file
     * @param podId
     * @param path
     * @param requestBody
     * @returns FileSignedUrlResponse Successful Response
     * @throws ApiError
     */
    static fileSignedUrl(podId, path, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/datastore/files/signed-url',
            path: {
                'pod_id': podId,
            },
            query: {
                'path': path,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Directory Tree
     * @param podId
     * @param rootPath
     * @param filesPerDirectory
     * @returns DirectoryTreeResponse Successful Response
     * @throws ApiError
     */
    static fileTree(podId, rootPath = '/', filesPerDirectory = 3) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/datastore/files/tree',
            path: {
                'pod_id': podId,
            },
            query: {
                'root_path': rootPath,
                'files_per_directory': filesPerDirectory,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get a short-lived URL for a file
     * @param podId
     * @param path
     * @returns FileUrlResponse Successful Response
     * @throws ApiError
     */
    static fileUrl(podId, path) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/datastore/files/url',
            path: {
                'pod_id': podId,
            },
            query: {
                'path': path,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
