// lemma-sdk/dist/openapi_client/services/FunctionsService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class FunctionsService {
    /**
     * List Functions
     * List all functions in a pod
     * @param podId
     * @param limit
     * @param pageToken
     * @returns FunctionListResponse Successful Response
     * @throws ApiError
     */
    static functionList(podId, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/functions',
            path: {
                'pod_id': podId,
            },
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Function
     * Create a new function in a pod. Do not send input_schema, output_schema, or config_schema; the platform derives those schemas from the function code and returns them in the response.
     * @param podId
     * @param requestBody
     * @returns FunctionActionResponse Successful Response
     * @throws ApiError
     */
    static functionCreate(podId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/functions',
            path: {
                'pod_id': podId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete Function
     * Delete a function
     * @param podId
     * @param functionName
     * @returns FunctionMessageResponse Successful Response
     * @throws ApiError
     */
    static functionDelete(podId, functionName) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/pods/{pod_id}/functions/{function_name}',
            path: {
                'pod_id': podId,
                'function_name': functionName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Function
     * Get a function by name
     * @param podId
     * @param functionName
     * @returns FunctionDetailResponse Successful Response
     * @throws ApiError
     */
    static functionGet(podId, functionName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/functions/{function_name}',
            path: {
                'pod_id': podId,
                'function_name': functionName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Update Function
     * Update a function. When code is supplied, the platform re-derives the function input_schema and output_schema and returns the refreshed function.
     * @param podId
     * @param functionName
     * @param requestBody
     * @returns FunctionActionResponse Successful Response
     * @throws ApiError
     */
    static functionUpdate(podId, functionName, requestBody) {
        return __request(OpenAPI, {
            method: 'PATCH',
            url: '/pods/{pod_id}/functions/{function_name}',
            path: {
                'pod_id': podId,
                'function_name': functionName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Function Resource Permissions
     * Get explicit resource grants assigned to a function.
     * @param podId
     * @param functionName
     * @returns FunctionPermissionsResponse Successful Response
     * @throws ApiError
     */
    static functionPermissionsGet(podId, functionName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/functions/{function_name}/permissions',
            path: {
                'pod_id': podId,
                'function_name': functionName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Replace Function Resource Permissions
     * Replace explicit resource grants assigned to a function.
     * @param podId
     * @param functionName
     * @param requestBody
     * @returns FunctionPermissionsResponse Successful Response
     * @throws ApiError
     */
    static functionPermissionsReplace(podId, functionName, requestBody) {
        return __request(OpenAPI, {
            method: 'PUT',
            url: '/pods/{pod_id}/functions/{function_name}/permissions',
            path: {
                'pod_id': podId,
                'function_name': functionName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List Runs
     * List runs for a function
     * @param podId
     * @param functionName
     * @param limit
     * @param pageToken
     * @returns FunctionRunListResponse Successful Response
     * @throws ApiError
     */
    static functionRunList(podId, functionName, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/functions/{function_name}/runs',
            path: {
                'pod_id': podId,
                'function_name': functionName,
            },
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Execute Function
     * Execute a function
     * @param podId
     * @param functionName
     * @param requestBody
     * @returns FunctionRunResponse Successful Response
     * @throws ApiError
     */
    static functionRun(podId, functionName, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/functions/{function_name}/runs',
            path: {
                'pod_id': podId,
                'function_name': functionName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Run
     * Get a specific function run
     * @param podId
     * @param functionName
     * @param runId
     * @returns FunctionRunResponse Successful Response
     * @throws ApiError
     */
    static functionRunGet(podId, functionName, runId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/functions/{function_name}/runs/{run_id}',
            path: {
                'pod_id': podId,
                'function_name': functionName,
                'run_id': runId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
