// lemma-sdk/dist/openapi_client/services/OrganizationsService.js
import { OrganizationInvitationStatus } from '../models/OrganizationInvitationStatus.js';
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class OrganizationsService {
    /**
     * List My Organizations
     * Get all organizations the current user belongs to
     * @param limit
     * @param pageToken
     * @returns OrganizationListResponse Successful Response
     * @throws ApiError
     */
    static orgList(limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations',
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Organization
     * Create a new organization
     * @param requestBody
     * @returns OrganizationResponse Successful Response
     * @throws ApiError
     */
    static orgCreate(requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations',
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List My Invitations
     * Get all pending invitations for the current user
     * @param status
     * @param limit
     * @param pageToken
     * @returns OrganizationInvitationListResponse Successful Response
     * @throws ApiError
     */
    static orgInvitationListMine(status = OrganizationInvitationStatus.PENDING, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/invitations',
            query: {
                'status': status,
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Revoke Invitation
     * Revoke an organization invitation
     * @param invitationId
     * @returns void
     * @throws ApiError
     */
    static orgInvitationRevoke(invitationId) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/organizations/invitations/{invitation_id}',
            path: {
                'invitation_id': invitationId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Organization Invitation
     * Get an invitation by id
     * @param invitationId
     * @returns OrganizationInvitationResponse Successful Response
     * @throws ApiError
     */
    static orgInvitationGet(invitationId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/invitations/{invitation_id}',
            path: {
                'invitation_id': invitationId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Accept Invitation
     * Accept an organization invitation
     * @param invitationId
     * @returns OrganizationMessageResponse Successful Response
     * @throws ApiError
     */
    static orgInvitationAccept(invitationId) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/invitations/{invitation_id}/accept',
            path: {
                'invitation_id': invitationId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Check Organization Slug Availability
     * Check whether an organization slug is available
     * @param slug
     * @returns OrganizationSlugAvailabilityResponse Successful Response
     * @throws ApiError
     */
    static orgSlugAvailability(slug) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/slug-availability',
            query: {
                'slug': slug,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Suggested Organizations
     * Get auto-join organizations matching the current user's email domain
     * @param limit
     * @param pageToken
     * @returns OrganizationListResponse Successful Response
     * @throws ApiError
     */
    static orgSuggested(limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/suggested',
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Organization
     * Get organization details
     * @param orgId
     * @returns OrganizationResponse Successful Response
     * @throws ApiError
     */
    static orgGet(orgId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{org_id}',
            path: {
                'org_id': orgId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Update Organization
     * Update an organization's name or join policy (owner only)
     * @param orgId
     * @param requestBody
     * @returns OrganizationResponse Successful Response
     * @throws ApiError
     */
    static orgUpdate(orgId, requestBody) {
        return __request(OpenAPI, {
            method: 'PATCH',
            url: '/organizations/{org_id}',
            path: {
                'org_id': orgId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List Organization Invitations
     * Get all pending invitations for an organization
     * @param orgId
     * @param status
     * @param limit
     * @param pageToken
     * @returns OrganizationInvitationListResponse Successful Response
     * @throws ApiError
     */
    static orgInvitationList(orgId, status = OrganizationInvitationStatus.PENDING, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{org_id}/invitations',
            path: {
                'org_id': orgId,
            },
            query: {
                'status': status,
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Invite Member
     * Invite a user to join the organization
     * @param orgId
     * @param requestBody
     * @returns OrganizationInvitationResponse Successful Response
     * @throws ApiError
     */
    static orgInvitationInvite(orgId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/{org_id}/invitations',
            path: {
                'org_id': orgId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Join Auto-Join Organization
     * Join an organization when the current user's email domain is allowed to auto-join
     * @param orgId
     * @returns OrganizationResponse Successful Response
     * @throws ApiError
     */
    static orgJoinAutoJoin(orgId) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/organizations/{org_id}/join',
            path: {
                'org_id': orgId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * List Organization Members
     * Get all members of an organization
     * @param orgId
     * @param limit
     * @param pageToken
     * @returns OrganizationMemberListResponse Successful Response
     * @throws ApiError
     */
    static orgMemberList(orgId, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/organizations/{org_id}/members',
            path: {
                'org_id': orgId,
            },
            query: {
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Remove Member
     * Remove a member from the organization
     * @param orgId
     * @param memberId
     * @returns void
     * @throws ApiError
     */
    static orgMemberRemove(orgId, memberId) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/organizations/{org_id}/members/{member_id}',
            path: {
                'org_id': orgId,
                'member_id': memberId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Update Member Role
     * Update a member's role in the organization
     * @param orgId
     * @param memberId
     * @param requestBody
     * @returns OrganizationMemberResponse Successful Response
     * @throws ApiError
     */
    static orgMemberUpdateRole(orgId, memberId, requestBody) {
        return __request(OpenAPI, {
            method: 'PATCH',
            url: '/organizations/{org_id}/members/{member_id}/role',
            path: {
                'org_id': orgId,
                'member_id': memberId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
