// lemma-sdk/dist/openapi_client/services/PodPermissionsService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class PodPermissionsService {
    /**
     * Get Pod Permission Catalog
     * @param podId
     * @returns PodPermissionCatalogResponse Successful Response
     * @throws ApiError
     */
    static podPermissionsCatalog(podId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/permissions/catalog',
            path: {
                'pod_id': podId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get My Pod Permissions
     * @param podId
     * @returns PodEffectivePermissionsResponse Successful Response
     * @throws ApiError
     */
    static podPermissionsMe(podId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/permissions/me',
            path: {
                'pod_id': podId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
