// lemma-sdk/dist/openapi_client/services/PodRolesService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class PodRolesService {
    /**
     * List Pod Roles
     * @param podId
     * @returns PodRoleListResponse Successful Response
     * @throws ApiError
     */
    static podRolesList(podId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/roles',
            path: {
                'pod_id': podId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Pod Role
     * @param podId
     * @param requestBody
     * @returns PodRoleResponse Successful Response
     * @throws ApiError
     */
    static podRolesCreate(podId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/roles',
            path: {
                'pod_id': podId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete Pod Role
     * @param podId
     * @param roleName
     * @returns void
     * @throws ApiError
     */
    static podRolesDelete(podId, roleName) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/pods/{pod_id}/roles/{role_name}',
            path: {
                'pod_id': podId,
                'role_name': roleName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Update Pod Role
     * @param podId
     * @param roleName
     * @param requestBody
     * @returns PodRoleResponse Successful Response
     * @throws ApiError
     */
    static podRolesUpdate(podId, roleName, requestBody) {
        return __request(OpenAPI, {
            method: 'PATCH',
            url: '/pods/{pod_id}/roles/{role_name}',
            path: {
                'pod_id': podId,
                'role_name': roleName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Pod Role Permissions
     * @param podId
     * @param roleName
     * @returns PodRolePermissionsResponse Successful Response
     * @throws ApiError
     */
    static podRolePermissionsGet(podId, roleName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/roles/{role_name}/permissions',
            path: {
                'pod_id': podId,
                'role_name': roleName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Replace Pod Role Permissions
     * @param podId
     * @param roleName
     * @param requestBody
     * @returns PodRolePermissionsResponse Successful Response
     * @throws ApiError
     */
    static podRolePermissionsReplace(podId, roleName, requestBody) {
        return __request(OpenAPI, {
            method: 'PUT',
            url: '/pods/{pod_id}/roles/{role_name}/permissions',
            path: {
                'pod_id': podId,
                'role_name': roleName,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
