// lemma-sdk/dist/openapi_client/services/PodResourceAccessService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class PodResourceAccessService {
    /**
     * Get Resource Access
     * @param podId
     * @param resourceType
     * @param resourceName
     * @returns ResourceAccessResponse Successful Response
     * @throws ApiError
     */
    static podResourceAccessGet(podId, resourceType, resourceName) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/resources/{resource_type}/{resource_name}/access',
            path: {
                'pod_id': podId,
                'resource_type': resourceType,
                'resource_name': resourceName,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete Resource Access Grant
     * @param podId
     * @param resourceType
     * @param resourceName
     * @param granteeType
     * @param granteeId
     * @returns ResourceAccessResponse Successful Response
     * @throws ApiError
     */
    static podResourceAccessGrantDelete(podId, resourceType, resourceName, granteeType, granteeId) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/pods/{pod_id}/resources/{resource_type}/{resource_name}/access/grantees/{grantee_type}/{grantee_id}',
            path: {
                'pod_id': podId,
                'resource_type': resourceType,
                'resource_name': resourceName,
                'grantee_type': granteeType,
                'grantee_id': granteeId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Replace Resource Access Grant
     * @param podId
     * @param resourceType
     * @param resourceName
     * @param granteeType
     * @param granteeId
     * @param requestBody
     * @returns ResourceAccessResponse Successful Response
     * @throws ApiError
     */
    static podResourceAccessGrantReplace(podId, resourceType, resourceName, granteeType, granteeId, requestBody) {
        return __request(OpenAPI, {
            method: 'PUT',
            url: '/pods/{pod_id}/resources/{resource_type}/{resource_name}/access/grantees/{grantee_type}/{grantee_id}',
            path: {
                'pod_id': podId,
                'resource_type': resourceType,
                'resource_name': resourceName,
                'grantee_type': granteeType,
                'grantee_id': granteeId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
