// lemma-sdk/dist/openapi_client/services/SchedulesService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class SchedulesService {
    /**
     * List Schedules
     * List pod schedules.
     * @param podId
     * @param scheduleType
     * @param isActive
     * @param agentName
     * @param workflowName
     * @param name
     * @param limit
     * @param pageToken
     * @returns ScheduleListResponse Successful Response
     * @throws ApiError
     */
    static scheduleList(podId, scheduleType, isActive, agentName, workflowName, name, limit = 100, pageToken) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/schedules',
            path: {
                'pod_id': podId,
            },
            query: {
                'schedule_type': scheduleType,
                'is_active': isActive,
                'agent_name': agentName,
                'workflow_name': workflowName,
                'name': name,
                'limit': limit,
                'page_token': pageToken,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Create Schedule
     * Create a new pod schedule.
     * @param podId
     * @param requestBody
     * @returns ScheduleDetailResponse Successful Response
     * @throws ApiError
     */
    static scheduleCreate(podId, requestBody) {
        return __request(OpenAPI, {
            method: 'POST',
            url: '/pods/{pod_id}/schedules',
            path: {
                'pod_id': podId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Delete Schedule
     * Delete a schedule.
     * @param podId
     * @param scheduleId
     * @returns void
     * @throws ApiError
     */
    static scheduleDelete(podId, scheduleId) {
        return __request(OpenAPI, {
            method: 'DELETE',
            url: '/pods/{pod_id}/schedules/{schedule_id}',
            path: {
                'pod_id': podId,
                'schedule_id': scheduleId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Schedule
     * Get a schedule by ID.
     * @param podId
     * @param scheduleId
     * @returns ScheduleDetailResponse Successful Response
     * @throws ApiError
     */
    static scheduleGet(podId, scheduleId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/pods/{pod_id}/schedules/{schedule_id}',
            path: {
                'pod_id': podId,
                'schedule_id': scheduleId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Update Schedule
     * Update a schedule.
     * @param podId
     * @param scheduleId
     * @param requestBody
     * @returns ScheduleDetailResponse Successful Response
     * @throws ApiError
     */
    static scheduleUpdate(podId, scheduleId, requestBody) {
        return __request(OpenAPI, {
            method: 'PATCH',
            url: '/pods/{pod_id}/schedules/{schedule_id}',
            path: {
                'pod_id': podId,
                'schedule_id': scheduleId,
            },
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
