// lemma-sdk/dist/openapi_client/services/UsageService.js
import { OpenAPI } from '../core/OpenAPI.js';
import { request as __request } from '../core/request.js';
export class UsageService {
    /**
     * List Usage Events
     * @param organizationId
     * @param start
     * @param end
     * @param days
     * @param limit
     * @param podId
     * @param userId
     * @param agentId
     * @param profileId
     * @param profileScope
     * @param modelName
     * @param usageKind
     * @param sourceType
     * @param status
     * @returns UsageListResponse Successful Response
     * @throws ApiError
     */
    static usageOrganizationEventsList(organizationId, start, end, days = 30, limit = 100, podId, userId, agentId, profileId, profileScope, modelName, usageKind, sourceType, status) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/usage/organizations/{organization_id}/events',
            path: {
                'organization_id': organizationId,
            },
            query: {
                'start': start,
                'end': end,
                'days': days,
                'limit': limit,
                'pod_id': podId,
                'user_id': userId,
                'agent_id': agentId,
                'profile_id': profileId,
                'profile_scope': profileScope,
                'model_name': modelName,
                'usage_kind': usageKind,
                'source_type': sourceType,
                'status': status,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Usage Limits
     * @param organizationId
     * @returns UsageLimitsResponse Successful Response
     * @throws ApiError
     */
    static usageOrganizationLimitsGet(organizationId) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/usage/organizations/{organization_id}/limits',
            path: {
                'organization_id': organizationId,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get My Usage
     * @param organizationId
     * @param start
     * @param end
     * @param days
     * @param limit
     * @param podId
     * @param userId
     * @param agentId
     * @param profileId
     * @param profileScope
     * @param modelName
     * @param usageKind
     * @param sourceType
     * @param status
     * @returns UsageSummaryResponse Successful Response
     * @throws ApiError
     */
    static usageOrganizationMeSummaryGet(organizationId, start, end, days = 30, limit = 100, podId, userId, agentId, profileId, profileScope, modelName, usageKind, sourceType, status) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/usage/organizations/{organization_id}/me',
            path: {
                'organization_id': organizationId,
            },
            query: {
                'start': start,
                'end': end,
                'days': days,
                'limit': limit,
                'pod_id': podId,
                'user_id': userId,
                'agent_id': agentId,
                'profile_id': profileId,
                'profile_scope': profileScope,
                'model_name': modelName,
                'usage_kind': usageKind,
                'source_type': sourceType,
                'status': status,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Usage Stats
     * @param organizationId
     * @param start
     * @param end
     * @param days
     * @param limit
     * @param podId
     * @param userId
     * @param agentId
     * @param profileId
     * @param profileScope
     * @param modelName
     * @param usageKind
     * @param sourceType
     * @param status
     * @param granularity
     * @param groupBy
     * @returns UsageStatsResponse Successful Response
     * @throws ApiError
     */
    static usageOrganizationStatsGet(organizationId, start, end, days = 30, limit = 100, podId, userId, agentId, profileId, profileScope, modelName, usageKind, sourceType, status, granularity = 'day', groupBy) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/usage/organizations/{organization_id}/stats',
            path: {
                'organization_id': organizationId,
            },
            query: {
                'start': start,
                'end': end,
                'days': days,
                'limit': limit,
                'pod_id': podId,
                'user_id': userId,
                'agent_id': agentId,
                'profile_id': profileId,
                'profile_scope': profileScope,
                'model_name': modelName,
                'usage_kind': usageKind,
                'source_type': sourceType,
                'status': status,
                'granularity': granularity,
                'group_by': groupBy,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
    /**
     * Get Organization Usage Summary
     * @param organizationId
     * @param start
     * @param end
     * @param days
     * @param limit
     * @param podId
     * @param userId
     * @param agentId
     * @param profileId
     * @param profileScope
     * @param modelName
     * @param usageKind
     * @param sourceType
     * @param status
     * @returns UsageSummaryResponse Successful Response
     * @throws ApiError
     */
    static usageOrganizationSummaryGet(organizationId, start, end, days = 30, limit = 100, podId, userId, agentId, profileId, profileScope, modelName, usageKind, sourceType, status) {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/usage/organizations/{organization_id}/summary',
            path: {
                'organization_id': organizationId,
            },
            query: {
                'start': start,
                'end': end,
                'days': days,
                'limit': limit,
                'pod_id': podId,
                'user_id': userId,
                'agent_id': agentId,
                'profile_id': profileId,
                'profile_scope': profileScope,
                'model_name': modelName,
                'usage_kind': usageKind,
                'source_type': sourceType,
                'status': status,
            },
            errors: {
                422: `Validation Error`,
            },
        });
    }
}
