// lemma-sdk/dist/namespaces/agents.js
import { AgentsService } from "../openapi_client/services/AgentsService.js";
export class AgentsNamespace {
    client;
    podId;
    conversations;
    constructor(client, podId, 
    // Lazy accessor — conversations is constructed alongside agents on the client.
    conversations) {
        this.client = client;
        this.podId = podId;
        this.conversations = conversations;
    }
    list(options = {}) {
        return this.client.request(() => AgentsService.agentList(this.podId(), options.pageToken, options.limit ?? 100));
    }
    /**
     * Run an agent on a single message (the `.run` verb, alongside
     * `functions.run` / `datastore.query`). Note the return contract differs:
     * those return the result directly, whereas an agent reply is asynchronous —
     * this opens a fresh conversation, sends `message`, and returns the created
     * conversation (read the reply via `client.conversations.messages.list(conv.id)`).
     * With `stream: true` it returns the SSE stream so you can consume tokens as
     * they arrive.
     */
    async run(agentName, message, options = {}) {
        if (!this.conversations) {
            throw new Error("agents.run requires the conversations namespace — call it via client.agents.run().");
        }
        const conversations = this.conversations();
        const conversation = await conversations.createForAgent(agentName, {
            title: options.title,
            metadata: options.metadata,
        });
        if (options.stream) {
            return conversations.sendMessageStream(conversation.id, { content: message }, { signal: options.signal });
        }
        await conversations.messages.send(conversation.id, { content: message });
        return conversation;
    }
    create(payload) {
        return this.client.request(() => AgentsService.agentCreate(this.podId(), payload));
    }
    get(agentName) {
        return this.client.request(() => AgentsService.agentGet(this.podId(), agentName));
    }
    update(agentName, payload) {
        return this.client.request(() => AgentsService.agentUpdate(this.podId(), agentName, payload));
    }
    delete(agentName) {
        return this.client.request(() => AgentsService.agentDelete(this.podId(), agentName));
    }
    permissions = {
        get: (agentName) => this.client.request(() => AgentsService.agentPermissionsGet(this.podId(), agentName)),
        replace: (agentName, payload) => this.client.request(() => AgentsService.agentPermissionsReplace(this.podId(), agentName, payload)),
    };
}
