// lemma-sdk/dist/namespaces/agent-runtime.js
import { AgentRuntimeService } from "../openapi_client/services/AgentRuntimeService.js";
export class AgentRuntimeNamespace {
    client;
    constructor(client) {
        this.client = client;
    }
    listHarnesses() {
        return this.client.request(() => AgentRuntimeService.agentRuntimeHarnessesList());
    }
    listAvailableHarnesses() {
        return this.listHarnesses();
    }
    listRuntimes(orgId) {
        return this.listProfiles(orgId);
    }
    listProfiles(orgId) {
        return this.client.request(() => AgentRuntimeService.agentRuntimeProfilesList(orgId));
    }
    createRuntime(orgId, request) {
        return this.createProfile(orgId, request);
    }
    createProfile(orgId, request) {
        return this.client.request(() => AgentRuntimeService.agentRuntimeProfilesCreate(orgId, request));
    }
    /**
     * @deprecated Runtime defaults are now pod config (`default_profile_id`) or
     * organization Agent Runtimes. The backend no longer exposes a global
     * default-runtime mutation endpoint.
     */
    updateDefault(agentRuntime) {
        void agentRuntime;
        return Promise.reject(new Error("agentRuntime.updateDefault is no longer supported. Update pod config.default_profile_id instead."));
    }
}
