// lemma-sdk/dist/namespaces/apps.js
import { AppsService } from "../openapi_client/services/AppsService.js";
export class AppsNamespace {
    client;
    http;
    podId;
    constructor(client, http, podId) {
        this.client = client;
        this.http = http;
        this.podId = podId;
    }
    list(options = {}) {
        return this.client.request(() => AppsService.appList(this.podId(), options.limit ?? 100, options.pageToken));
    }
    create(payload) {
        return this.client.request(() => AppsService.appCreate(this.podId(), payload));
    }
    get(name) {
        return this.client.request(() => AppsService.appGet(this.podId(), name));
    }
    update(name, payload) {
        return this.client.request(() => AppsService.appUpdate(this.podId(), name, payload));
    }
    delete(name) {
        return this.client.request(() => AppsService.appDelete(this.podId(), name));
    }
    /** Promote a conversation widget into a persisted app (save as app). */
    createFromWidget(payload) {
        return this.http.request("POST", `/pods/${this.podId()}/apps/from-widget`, {
            body: payload,
        });
    }
    assets = {
        get: (name, path) => this.http.request("GET", `/pods/${this.podId()}/apps/${name}/assets${path ? `/${path.replace(/^\/+/, "")}` : ""}`),
    };
    bundle = {
        upload: (name, payload) => this.client.request(() => AppsService.appBundleUpload(this.podId(), name, payload)),
    };
    source = {
        download: (name) => this.http.requestBytes("GET", `/pods/${this.podId()}/apps/${name}/source/archive`),
    };
    dist = {
        download: (name) => this.http.requestBytes("GET", `/pods/${this.podId()}/apps/${name}/dist/archive`),
    };
}
