// lemma-sdk/dist/namespaces/connectors.js
import { AgentToolsService } from "../openapi_client/services/AgentToolsService.js";
import { ConnectorsService } from "../openapi_client/services/ConnectorsService.js";
function encodePath(value) {
    return encodeURIComponent(value);
}
export class ConnectorsNamespace {
    client;
    http;
    constructor(client, http) {
        this.client = client;
        this.http = http;
    }
    list(options = {}) {
        return this.client.request(() => ConnectorsService.connectorList(options.limit ?? 100, options.pageToken));
    }
    get(connectorId) {
        return this.client.request(() => ConnectorsService.connectorGet(connectorId));
    }
    operations = {
        discover: (scope, options = {}) => this.client.request(() => ConnectorsService.connectorOperationDiscover(scope.organizationId, scope.authConfigName, options.query, options.limit ?? 100)),
        list: async (scope, options = {}) => {
            const response = await this.client.request(() => ConnectorsService.connectorOperationDiscover(scope.organizationId, scope.authConfigName, options.query, options.limit ?? 100));
            return response.items ?? [];
        },
        get: (scope, operationName) => this.client.request(() => ConnectorsService.connectorOperationDetail(scope.organizationId, scope.authConfigName, operationName)),
        details: (scope, operationNames) => {
            const body = { operation_names: operationNames };
            return this.client.request(() => ConnectorsService.connectorOperationDetailsBatch(scope.organizationId, scope.authConfigName, body));
        },
        execute: (scope, operationName, payload, accountId) => {
            const body = { payload, account_id: accountId };
            return this.client.request(() => ConnectorsService.connectorOperationExecute(scope.organizationId, scope.authConfigName, operationName, body));
        },
    };
    helperAgent(goal, appNames) {
        const body = {
            app_names: appNames,
            goal,
        };
        return this.client.request(() => AgentToolsService.agentToolConnectorHelperAgent(body));
    }
    triggers = {
        list: (scope, options = {}) => this.client.request(() => ConnectorsService.connectorTriggerList(scope.organizationId, scope.authConfigName, options.search, options.limit ?? 100)),
        get: (scope, triggerName) => this.client.request(() => ConnectorsService.connectorTriggerGet(scope.organizationId, scope.authConfigName, triggerName)),
    };
    accounts = {
        list: (organizationId, options = {}) => this.client.request(() => ConnectorsService.connectorAccountList(organizationId, options.connectorId, options.limit ?? 100, options.pageToken)),
        create: (organizationId, payload) => this.client.request(() => ConnectorsService.connectorAccountCreate(organizationId, payload)),
        get: (organizationId, accountId) => this.client.request(() => ConnectorsService.connectorAccountGet(organizationId, accountId)),
        credentials: (organizationId, accountId) => this.client.request(() => ConnectorsService.connectorAccountCredentialsGet(organizationId, accountId)),
        delete: (organizationId, accountId) => this.client.request(() => ConnectorsService.connectorAccountDelete(organizationId, accountId)),
        /**
         * @deprecated Use list/get/create with an organization id. Kept only for
         * callers that still need the response shape while migrating.
         */
        listOrgScoped: (organizationId, options = {}) => this.http.request("GET", `/organizations/${encodePath(organizationId)}/connectors/accounts`, {
            params: {
                connector_id: options.connectorId,
                limit: options.limit ?? 100,
                page_token: options.pageToken,
            },
        }),
    };
    authConfigs = {
        list: (organizationId, options = {}) => this.client.request(() => ConnectorsService.connectorAuthConfigList(organizationId, options.limit ?? 100, options.pageToken)),
        create: (organizationId, payload) => this.client.request(() => ConnectorsService.connectorAuthConfigCreate(organizationId, payload)),
        get: (organizationId, authConfigName) => this.client.request(() => ConnectorsService.connectorAuthConfigGet(organizationId, authConfigName)),
        delete: (organizationId, authConfigName) => this.client.request(() => ConnectorsService.connectorAuthConfigDelete(organizationId, authConfigName)),
    };
    async enableApp(organizationId, connectorId, options = {}) {
        const configs = await this.authConfigs.list(organizationId, { limit: 100 });
        const existing = configs.items.find((config) => config.connector_id === connectorId && config.status === "ACTIVE");
        if (existing)
            return existing;
        return this.authConfigs.create(organizationId, {
            connector_id: connectorId,
            provider: options.provider,
            config_source: options.config_source ?? "SYSTEM_DEFAULT",
            credential_config: options.credential_config ?? options.provider_config,
            name: options.name,
        });
    }
    createConnectRequest(organizationId, input) {
        const payload = typeof input === "string" ? { connector_id: input } : input;
        return this.client.request(() => ConnectorsService.connectorConnectRequestCreate(organizationId, payload));
    }
}
