// lemma-sdk/dist/namespaces/conversations.js
function normalizeConversation(conversation) {
    if (!conversation)
        return conversation;
    const record = conversation;
    return {
        ...record,
        model: record.model ?? record.agent_runtime?.model_name ?? record.model_name ?? null,
        status: record.status ?? "waiting",
    };
}
function normalizeConversationList(response) {
    const items = (response.items ?? []).map((conversation) => normalizeConversation(conversation));
    return {
        ...response,
        items,
    };
}
function normalizeMessage(message) {
    return message;
}
export class ConversationsNamespace {
    http;
    podId;
    runtimeCatalogPromise;
    profileCatalogPromises = new Map();
    constructor(http, podId) {
        this.http = http;
        this.podId = podId;
    }
    resolvePodId(explicitPodId) {
        if (typeof explicitPodId === "string") {
            return explicitPodId;
        }
        try {
            return this.podId();
        }
        catch {
            return undefined;
        }
    }
    requirePodId(explicitPodId) {
        const podId = this.resolvePodId(explicitPodId);
        if (!podId) {
            throw new Error("pod_id is required for this conversation operation.");
        }
        return podId;
    }
    listRuntimeCatalog() {
        this.runtimeCatalogPromise ??= this.http.request("GET", "/agent-runtime/harnesses");
        return this.runtimeCatalogPromise;
    }
    listProfileCatalog(orgId) {
        const key = orgId.trim();
        const existing = this.profileCatalogPromises.get(key);
        if (existing)
            return existing;
        const request = this.http.request("GET", `/organizations/${encodeURIComponent(key)}/agent-runtime/profiles`);
        this.profileCatalogPromises.set(key, request);
        return request;
    }
    modelOptionsFromProfiles(catalog) {
        return catalog.items.flatMap((profile) => {
            const catalogEntries = profile.model_catalog ?? [];
            const entries = catalogEntries.length > 0
                ? catalogEntries
                : profile.default_model_name
                    ? [{
                            name: profile.default_model_name,
                            display_name: null,
                            provider_model_name: profile.default_model_name,
                            capabilities: [],
                            default_model_settings: {},
                            metadata: {},
                        }]
                    : [];
            return entries.map((model) => ({
                id: model.name,
                name: model.display_name ?? model.name,
                agentRuntime: profile,
                agentRuntimeId: profile.id,
                profile,
                profile_id: profile.id,
                harness_kind: profile.derived_harness_kind,
                description: profile.name,
                runtime: {
                    profile_id: profile.id,
                    model_name: model.name,
                },
            }));
        });
    }
    async resolveAgentRuntime(agentRuntime, model, harnessKind, profileId) {
        if (agentRuntime || !model) {
            return agentRuntime;
        }
        if (profileId) {
            return {
                profile_id: profileId,
                model_name: model,
            };
        }
        void harnessKind;
        return undefined;
    }
    list(options = {}) {
        const podId = this.requirePodId(options.pod_id);
        return this.http.request("GET", `/pods/${podId}/conversations`, {
            params: {
                agent_name: options.agent_name,
                parent_id: options.parent_id,
                type: options.type,
                limit: options.limit ?? 20,
                page_token: options.page_token,
            },
        }).then(normalizeConversationList);
    }
    listByAgent(agentName, options = {}) {
        return this.list({ ...options, agent_name: agentName });
    }
    async listModels(options = {}) {
        const orgId = options.orgId?.trim();
        if (orgId) {
            const catalog = await this.listProfileCatalog(orgId);
            const items = this.modelOptionsFromProfiles(catalog);
            return {
                items,
                limit: items.length,
                next_page_token: null,
            };
        }
        const catalog = await this.listRuntimeCatalog();
        const items = catalog.items.flatMap((harness) => (harness.models ?? []).map((model) => ({
            id: model,
            name: model,
            harness_kind: harness.harness_kind,
            description: harness.daemon_display_name,
        })));
        return {
            items,
            limit: items.length,
            next_page_token: null,
        };
    }
    async create(payload = {}) {
        const podId = this.requirePodId(payload.pod_id);
        const { agent_name, harness_kind, model, model_name, pod_id, profile_id, ...requestBody } = payload;
        const agentRuntime = await this.resolveAgentRuntime(requestBody.agent_runtime, model_name ?? model, harness_kind, profile_id);
        const body = {
            ...requestBody,
            agent_name: agent_name ?? undefined,
            agent_runtime: agentRuntime,
        };
        void pod_id;
        return this.http.request("POST", `/pods/${podId}/conversations`, {
            body,
        }).then(normalizeConversation);
    }
    createForAgent(agentName, payload = {}) {
        return this.create({
            ...payload,
            agent_name: agentName,
        });
    }
    get(conversationId, options = {}) {
        const podId = this.requirePodId(options.pod_id);
        return this.http.request("GET", `/pods/${podId}/conversations/${conversationId}`)
            .then(normalizeConversation);
    }
    async update(conversationId, payload, options = {}) {
        const podId = this.requirePodId(options.pod_id);
        const { harness_kind, model, model_name, profile_id, ...requestBody } = payload;
        const agentRuntime = await this.resolveAgentRuntime(requestBody.agent_runtime, model_name ?? model, harness_kind, profile_id);
        const body = {
            ...requestBody,
            agent_runtime: agentRuntime,
        };
        return this.http.request("PATCH", `/pods/${podId}/conversations/${conversationId}`, {
            body,
        }).then(normalizeConversation);
    }
    sendMessageStream(conversationId, payload, options = {}) {
        const podId = this.requirePodId(options.pod_id);
        return this.http.stream(`/pods/${podId}/conversations/${conversationId}/messages`, {
            method: "POST",
            body: payload,
            signal: options.signal,
            headers: {
                "Content-Type": "application/json",
                Accept: "text/event-stream",
            },
        });
    }
    resumeStream(conversationId, options = {}) {
        const podId = this.requirePodId(options.pod_id);
        return this.http.stream(`/pods/${podId}/conversations/${conversationId}/stream`, {
            signal: options.signal,
            headers: {
                Accept: "text/event-stream",
            },
        });
    }
    stopRun(conversationId, options = {}) {
        const podId = this.requirePodId(options.pod_id);
        return this.http.request("POST", `/pods/${podId}/conversations/${conversationId}/stop`, {
            body: {},
        }).then(normalizeConversation);
    }
    messages = {
        list: (conversationId, options = {}) => {
            const podId = this.requirePodId(options.pod_id);
            return this.http.request("GET", `/pods/${podId}/conversations/${conversationId}/messages`, {
                params: {
                    page_token: options.page_token,
                    before_sequence: options.before_sequence,
                    after_sequence: options.after_sequence,
                    limit: options.limit ?? 100,
                },
            }).then((response) => ({
                ...response,
                items: (response.items ?? []).map(normalizeMessage),
            }));
        },
        send: (conversationId, payload, options = {}) => {
            const podId = this.requirePodId(options.pod_id);
            return this.http.request("POST", `/pods/${podId}/conversations/${conversationId}/messages`, {
                body: payload,
            });
        },
    };
    approvals = {
        list: (conversationId, options = {}) => {
            const podId = this.requirePodId(options.pod_id);
            return this.http.request("GET", `/pods/${podId}/conversations/${conversationId}/approvals`);
        },
        resolve: (conversationId, approvalId, payload, options = {}) => {
            const podId = this.requirePodId(options.pod_id);
            const body = {
                ...payload,
                decision: payload.decision,
            };
            return this.http.request("POST", `/pods/${podId}/conversations/${conversationId}/approvals/${approvalId}/decision`, { body });
        },
    };
}
