// lemma-sdk/dist/namespaces/datastore.js
import { QueryService } from "../openapi_client/services/QueryService.js";
import { watchDatastoreChanges, } from "../datastore-changes.js";
export class DatastoreNamespace {
    client;
    podId;
    apiUrl;
    auth;
    constructor(client, podId, apiUrl, auth) {
        this.client = client;
        this.podId = podId;
        this.apiUrl = apiUrl;
        this.auth = auth;
    }
    query(request) {
        const payload = typeof request === "string" ? { query: request } : request;
        return this.client.request(() => QueryService.queryExecute(this.podId(), payload));
    }
    /**
     * Stream live record changes (insert/update/delete) over a WebSocket.
     *
     * Returns a handle; call `handle.close()` (or abort `options.signal`) to stop.
     * RLS tables deliver only the caller's own rows; shared tables deliver all
     * members' changes. Reconnects automatically and resumes from the last change.
     */
    watchChanges(options) {
        return watchDatastoreChanges(this.apiUrl, this.auth, this.podId(), options);
    }
}
