// lemma-sdk/dist/namespaces/files.js
import { SearchMethod } from "../openapi_client/models/SearchMethod.js";
import { FilesService } from "../openapi_client/services/FilesService.js";
function joinDatastorePath(basePath, leaf) {
    const normalizedLeaf = leaf.replace(/^\/+/, "");
    const trimmedBase = (basePath ?? "/").trim();
    const normalizedBase = trimmedBase.length > 0 ? trimmedBase : "/";
    if (normalizedBase === "/") {
        return `/${normalizedLeaf}`;
    }
    return `${normalizedBase.replace(/\/+$/, "")}/${normalizedLeaf}`;
}
function getDirectoryPath(path) {
    const normalized = path.trim();
    if (!normalized || normalized === "/") {
        return "/";
    }
    const withoutTrailing = normalized.replace(/\/+$/, "");
    const index = withoutTrailing.lastIndexOf("/");
    if (index <= 0) {
        return "/";
    }
    return withoutTrailing.slice(0, index);
}
function getBaseName(path) {
    const normalized = path.trim().replace(/\/+$/, "");
    const index = normalized.lastIndexOf("/");
    if (index === -1) {
        return normalized;
    }
    return normalized.slice(index + 1);
}
export class FilesNamespace {
    client;
    http;
    podId;
    constructor(client, http, podId) {
        this.client = client;
        this.http = http;
        this.podId = podId;
    }
    list(options = {}) {
        const directoryPath = options.directoryPath ?? options.parentId ?? "/";
        return this.client.request(() => FilesService.fileList(this.podId(), directoryPath, options.limit ?? 100, options.pageToken));
    }
    get(path) {
        return this.client.request(() => FilesService.fileGet(this.podId(), path));
    }
    /**
     * URLs for a file: a short-lived download `url` plus a permanent
     * authenticated `app_url` deep-link that opens the file in the Lemma
     * frontend (the viewer must be a signed-in pod member).
     */
    getUrl(path) {
        return this.client.request(() => FilesService.fileUrl(this.podId(), path));
    }
    /**
     * Mint a public, hit-capped short signed URL (no login needed to open).
     * Expires after `expiresSeconds` (default 3h, max 24h) and serves the file
     * at most `maxHits` times (default 50, max 100); both bounds are clamped
     * server-side. Use it to share a file outside the pod without unbounded egress.
     */
    createSignedUrl(path, options = {}) {
        const body = {
            expires_seconds: options.expiresSeconds,
            max_hits: options.maxHits,
        };
        return this.client.request(() => FilesService.fileSignedUrl(this.podId(), path, body));
    }
    delete(path) {
        return this.client.request(() => FilesService.fileDelete(this.podId(), path));
    }
    search(query, options = {}) {
        return this.client.request(() => FilesService.fileSearch(this.podId(), {
            query,
            limit: options.limit ?? 10,
            scope_mode: options.scopeMode,
            scope_path: options.scopePath,
            search_method: (options.searchMethod ?? SearchMethod.HYBRID),
        }));
    }
    download(path) {
        const encodedPath = encodeURIComponent(path);
        return this.http.requestBytes("GET", `/pods/${this.podId()}/datastore/files/download?path=${encodedPath}`);
    }
    tree(options = {}) {
        return this.client.request(() => FilesService.fileTree(this.podId(), options.rootPath ?? "/", options.filesPerDirectory ?? 3));
    }
    upload(file, options = {}) {
        const payload = {
            data: file,
            name: options.name ?? (file instanceof File ? file.name : undefined),
            description: options.description,
            directory_path: options.directoryPath ?? options.parentId ?? "/",
            search_enabled: options.searchEnabled ?? true,
        };
        return this.client.request(() => FilesService.fileUpload(this.podId(), payload));
    }
    update(path, options = {}) {
        const targetDirectory = options.directoryPath ?? options.parentId;
        const resolvedNewPath = options.newPath
            ?? (options.name
                ? joinDatastorePath(targetDirectory ?? getDirectoryPath(path), options.name)
                : undefined)
            ?? (targetDirectory
                ? joinDatastorePath(targetDirectory, getBaseName(path))
                : undefined);
        const payload = {
            path,
            data: options.file,
            description: options.description,
            new_path: resolvedNewPath,
            search_enabled: options.searchEnabled,
            visibility: options.visibility,
        };
        return this.client.request(() => FilesService.fileUpdate(this.podId(), payload));
    }
    folder = {
        create: (name, options = {}) => {
            const payload = {
                path: joinDatastorePath(options.directoryPath ?? options.parentId, name),
                description: options.description,
            };
            return this.client.request(() => FilesService.fileFolderCreate(this.podId(), payload));
        },
    };
    // Derived child files of a processed document (converted markdown, extracted
    // figures, and on-demand page renders), addressed by `<file-path>/<artifact>`.
    children = {
        list: (path) => this.client.request(() => FilesService.fileChildrenList(this.podId(), path)),
        content: (childPath, options = {}) => {
            const params = new URLSearchParams({ path: childPath });
            if (options.pageStart != null)
                params.set("page_start", String(options.pageStart));
            if (options.pageEnd != null)
                params.set("page_end", String(options.pageEnd));
            return this.http.requestBytes("GET", `/pods/${this.podId()}/datastore/files/children/content?${params.toString()}`);
        },
        markdown: (path, options = {}) => this.children.content(`${path}/document.md`, options),
    };
}
