// lemma-sdk/dist/namespaces/functions.js
import { FunctionsService } from "../openapi_client/services/FunctionsService.js";
export class FunctionsNamespace {
    client;
    podId;
    constructor(client, podId) {
        this.client = client;
        this.podId = podId;
    }
    list(options = {}) {
        return this.client.request(() => FunctionsService.functionList(this.podId(), options.limit ?? 100, options.pageToken));
    }
    create(payload) {
        return this.client.request(() => FunctionsService.functionCreate(this.podId(), payload));
    }
    get(name) {
        return this.client.request(() => FunctionsService.functionGet(this.podId(), name));
    }
    update(name, payload) {
        return this.client.request(() => FunctionsService.functionUpdate(this.podId(), name, payload));
    }
    delete(name) {
        return this.client.request(() => FunctionsService.functionDelete(this.podId(), name));
    }
    /** Run a function — convenience alias for `functions.runs.create`, matching the
     *  Python SDK's `functions.run(name, input)` and the unified `.run` verb. */
    run(name, options = {}) {
        return this.runs.create(name, options);
    }
    permissions = {
        get: (name) => this.client.request(() => FunctionsService.functionPermissionsGet(this.podId(), name)),
        replace: (name, payload) => this.client.request(() => FunctionsService.functionPermissionsReplace(this.podId(), name, payload)),
    };
    runs = {
        create: (name, options = {}) => this.client.request(() => {
            const payload = { input_data: options.input };
            return FunctionsService.functionRun(this.podId(), name, payload);
        }),
        list: (name, params = {}) => this.client.request(() => FunctionsService.functionRunList(this.podId(), name, params.limit ?? 100, params.pageToken)),
        get: (name, runId) => this.client.request(() => FunctionsService.functionRunGet(this.podId(), name, runId)),
    };
}
