// lemma-sdk/dist/namespaces/organizations.js
import { OrganizationsService } from "../openapi_client/services/OrganizationsService.js";
export class OrganizationsNamespace {
    client;
    http;
    constructor(client, http) {
        this.client = client;
        this.http = http;
    }
    list(options = {}) {
        return this.client.request(() => OrganizationsService.orgList(options.limit ?? 100, options.pageToken ?? options.cursor));
    }
    get(orgId) {
        return this.client.request(() => OrganizationsService.orgGet(orgId));
    }
    create(payload) {
        return this.client.request(() => OrganizationsService.orgCreate(payload));
    }
    members = {
        list: (orgId, options = {}) => this.client.request(() => OrganizationsService.orgMemberList(orgId, options.limit ?? 100, options.pageToken ?? options.cursor)),
        updateRole: (orgId, memberId, role) => this.client.request(() => OrganizationsService.orgMemberUpdateRole(orgId, memberId, { role })),
        remove: (orgId, memberId) => this.client.request(() => OrganizationsService.orgMemberRemove(orgId, memberId)),
    };
    invitations = {
        listMine: async (options = {}) => {
            if (options.status) {
                return this.client.request(() => OrganizationsService.orgInvitationListMine(options.status, options.limit ?? 100, options.pageToken ?? options.cursor));
            }
            return this.http.request("GET", "/organizations/invitations", {
                params: {
                    limit: options.limit ?? 100,
                    page_token: options.pageToken ?? options.cursor,
                },
            });
        },
        list: async (orgId, options = {}) => {
            if (options.status) {
                return this.client.request(() => OrganizationsService.orgInvitationList(orgId, options.status, options.limit ?? 100, options.pageToken ?? options.cursor));
            }
            return this.http.request("GET", `/organizations/${encodeURIComponent(orgId)}/invitations`, {
                params: {
                    limit: options.limit ?? 100,
                    page_token: options.pageToken ?? options.cursor,
                },
            });
        },
        get: (invitationId) => this.client.request(() => OrganizationsService.orgInvitationGet(invitationId)),
        invite: (orgId, payload) => this.client.request(() => OrganizationsService.orgInvitationInvite(orgId, payload)),
        accept: (invitationId) => this.client.request(() => OrganizationsService.orgInvitationAccept(invitationId)),
        revoke: (invitationId) => this.client.request(() => OrganizationsService.orgInvitationRevoke(invitationId)),
    };
}
