// lemma-sdk/dist/namespaces/pod-members.js
import { PodMembersService } from "../openapi_client/services/PodMembersService.js";
export class PodMembersNamespace {
    client;
    constructor(client) {
        this.client = client;
    }
    list(podId, options = {}) {
        return this.client.request(() => PodMembersService.podMemberList(podId, options.limit ?? 100, options.pageToken ?? options.cursor));
    }
    add(podId, payload) {
        return this.client.request(() => PodMembersService.podMemberAdd(podId, payload));
    }
    get(podId, podMemberId) {
        return this.client.request(() => PodMembersService.podMemberGet(podId, podMemberId));
    }
    lookupByEmail(podId, email) {
        return this.client.request(() => PodMembersService.podMemberLookupByEmail(podId, email));
    }
    lookupByUserId(podId, userId) {
        return this.client.request(() => PodMembersService.podMemberLookupByUserId(podId, userId));
    }
    updateRole(podId, podMemberId, role) {
        return this.client.request(() => PodMembersService.podMemberUpdateRoles(podId, podMemberId, { roles: [role] }));
    }
    updateRoles(podId, podMemberId, roles) {
        return this.client.request(() => PodMembersService.podMemberUpdateRoles(podId, podMemberId, { roles }));
    }
    remove(podId, podMemberId) {
        return this.client.request(() => PodMembersService.podMemberRemove(podId, podMemberId));
    }
}
