// lemma-sdk/dist/namespaces/pod-roles.js
import { PodRolesService } from "../openapi_client/services/PodRolesService.js";
export class PodRolesNamespace {
    client;
    podId;
    constructor(client, podId) {
        this.client = client;
        this.podId = podId;
    }
    list(podId) {
        return this.client.request(() => PodRolesService.podRolesList(podId ?? this.podId()));
    }
    create(payload, podId) {
        return this.client.request(() => PodRolesService.podRolesCreate(podId ?? this.podId(), payload));
    }
    update(roleName, payload, podId) {
        return this.client.request(() => PodRolesService.podRolesUpdate(podId ?? this.podId(), roleName, payload));
    }
    delete(roleName, podId) {
        return this.client.request(() => PodRolesService.podRolesDelete(podId ?? this.podId(), roleName));
    }
    permissions = {
        get: (roleName, podId) => this.client.request(() => PodRolesService.podRolePermissionsGet(podId ?? this.podId(), roleName)),
        replace: (roleName, payload, podId) => this.client.request(() => PodRolesService.podRolePermissionsReplace(podId ?? this.podId(), roleName, payload)),
    };
}
