// lemma-sdk/dist/namespaces/pod-surfaces.js
import { AgentSurfacesService } from "../openapi_client/services/AgentSurfacesService.js";
/**
 * Agent surfaces, addressed by platform (a surface is unique per pod+platform).
 *
 * One `upsert` write covers create, config/channel edits, account and
 * credential changes, and enable/disable (via `is_enabled`). `delete` removes
 * the surface and frees its account for reuse. `setup` merges live readiness,
 * admin-consent, and the platform checklist into one read.
 */
export class PodSurfacesNamespace {
    client;
    constructor(client) {
        this.client = client;
    }
    list(podId, options = {}) {
        return this.client.request(() => AgentSurfacesService.agentSurfaceList(podId, options.limit ?? 100, options.pageToken ?? options.cursor));
    }
    upsert(podId, platform, payload) {
        return this.client.request(() => AgentSurfacesService.agentSurfaceUpsert(podId, platform, payload));
    }
    get(podId, platform) {
        return this.client.request(() => AgentSurfacesService.agentSurfaceGet(podId, platform));
    }
    delete(podId, platform) {
        return this.client.request(() => AgentSurfacesService.agentSurfaceDelete(podId, platform));
    }
    setup(podId, platform) {
        return this.client.request(() => AgentSurfacesService.agentSurfaceSetup(podId, platform));
    }
    channels(podId, platform) {
        return this.client.request(() => AgentSurfacesService.agentSurfaceChannels(podId, platform));
    }
}
