// lemma-sdk/dist/namespaces/pod-join-requests.js
import { OrganizationRole } from "../openapi_client/models/OrganizationRole.js";
import { PodRole } from "../openapi_client/models/PodRole.js";
import { PodJoinRequestsService } from "../openapi_client/services/PodJoinRequestsService.js";
export class PodJoinRequestsNamespace {
    client;
    constructor(client) {
        this.client = client;
    }
    create(podId) {
        return this.client.request(() => PodJoinRequestsService.podJoinRequestCreate(podId));
    }
    me(podId) {
        return this.client.request(() => PodJoinRequestsService.podJoinRequestMe(podId));
    }
    list(podId, options = {}) {
        return this.client.request(() => PodJoinRequestsService.podJoinRequestList(podId, options.status, options.limit ?? 100, options.pageToken ?? options.cursor));
    }
    approve(podId, joinRequestId, payload = {}) {
        return this.client.request(() => PodJoinRequestsService.podJoinRequestApprove(podId, joinRequestId, {
            org_role: payload.org_role ?? OrganizationRole.ORG_MEMBER,
            pod_role: payload.pod_role ?? PodRole.POD_USER,
        }));
    }
}
