// lemma-sdk/dist/namespaces/pods.js
import { PodsService } from "../openapi_client/services/PodsService.js";
export class PodsNamespace {
    client;
    http;
    constructor(client, http) {
        this.client = client;
        this.http = http;
    }
    list(options = {}) {
        if (options.organizationId) {
            return this.listByOrganization(options.organizationId, {
                limit: options.limit,
                pageToken: options.pageToken ?? options.cursor,
            });
        }
        throw new Error("organizationId is required for pods.list(). Use listByOrganization(organizationId).");
    }
    listByOrganization(organizationId, options = {}) {
        return this.client.request(() => PodsService.podList(organizationId, options.limit ?? 100, options.pageToken ?? options.cursor));
    }
    get(podId) {
        return this.client.request(() => PodsService.podGet(podId));
    }
    create(payload) {
        return this.client.request(() => PodsService.podCreate(payload));
    }
    update(podId, payload) {
        return this.client.request(() => PodsService.podUpdate(podId, payload));
    }
    delete(podId) {
        return this.client.request(() => PodsService.podDelete(podId));
    }
}
