// lemma-sdk/dist/namespaces/records.js
import { RecordsService } from "../openapi_client/services/RecordsService.js";
function serializeFilters(filters) {
    if (!filters || filters.length === 0) {
        return undefined;
    }
    return filters.map((filter) => JSON.stringify(filter));
}
function serializeSort(sort) {
    if (!sort || sort.length === 0) {
        return undefined;
    }
    return sort.map((entry) => JSON.stringify(entry));
}
export class RecordsNamespace {
    client;
    podId;
    constructor(client, podId) {
        this.client = client;
        this.podId = podId;
    }
    list(table, options = {}) {
        const { filters, sort, limit, pageToken, offset } = options;
        return this.client.request(() => RecordsService.recordList(this.podId(), table, limit ?? 20, offset, serializeFilters(filters), serializeSort(sort), pageToken));
    }
    create(table, data) {
        return this.client.request(() => RecordsService.recordCreate(this.podId(), table, { data }));
    }
    get(table, recordId) {
        return this.client.request(() => RecordsService.recordGet(this.podId(), table, recordId));
    }
    update(table, recordId, data) {
        return this.client.request(() => RecordsService.recordUpdate(this.podId(), table, recordId, { data }));
    }
    delete(table, recordId) {
        return this.client.request(() => RecordsService.recordDelete(this.podId(), table, recordId));
    }
    query(table, payload) {
        return this.client.request(() => RecordsService.recordList(this.podId(), table, payload.limit ?? 20, payload.offset, serializeFilters(payload.filters), serializeSort(payload.sort), payload.page_token));
    }
    bulk = {
        create: (table, records) => {
            const payload = { records };
            return this.client.request(() => RecordsService.recordBulkCreate(this.podId(), table, payload));
        },
        update: (table, records) => {
            const payload = { records };
            return this.client.request(() => RecordsService.recordBulkUpdate(this.podId(), table, payload));
        },
        delete: (table, recordIds) => {
            const payload = { record_ids: recordIds };
            return this.client.request(() => RecordsService.recordBulkDelete(this.podId(), table, payload));
        },
    };
}
