// lemma-sdk/dist/namespaces/schedules.js
import { SchedulesService } from "../openapi_client/services/SchedulesService.js";
export class SchedulesNamespace {
    client;
    podId;
    constructor(client, podId) {
        this.client = client;
        this.podId = podId;
    }
    list(options = {}) {
        return this.client.request(() => SchedulesService.scheduleList(this.podId(), options.scheduleType, options.isActive, options.agentName, options.workflowName, options.name, options.limit ?? 100, options.pageToken));
    }
    create(payload) {
        return this.client.request(() => SchedulesService.scheduleCreate(this.podId(), payload));
    }
    get(scheduleId) {
        return this.client.request(() => SchedulesService.scheduleGet(this.podId(), scheduleId));
    }
    update(scheduleId, payload) {
        return this.client.request(() => SchedulesService.scheduleUpdate(this.podId(), scheduleId, payload));
    }
    delete(scheduleId) {
        return this.client.request(() => SchedulesService.scheduleDelete(this.podId(), scheduleId));
    }
}
