// lemma-sdk/dist/namespaces/tables.js
import { TablesService } from "../openapi_client/services/TablesService.js";
function normalizeCreateTablePayload(payload) {
    if ("table_name" in payload) {
        const { table_name, name, ...rest } = payload;
        return {
            ...rest,
            name: name ?? table_name,
        };
    }
    return payload;
}
export class TablesNamespace {
    client;
    podId;
    constructor(client, podId) {
        this.client = client;
        this.podId = podId;
    }
    list(options = {}) {
        return this.client.request(() => TablesService.tableList(this.podId(), options.limit ?? 100, options.pageToken));
    }
    create(payload) {
        return this.client.request(() => TablesService.tableCreate(this.podId(), normalizeCreateTablePayload(payload)));
    }
    get(tableName) {
        return this.client.request(() => TablesService.tableGet(this.podId(), tableName));
    }
    update(tableName, payload) {
        return this.client.request(() => TablesService.tableUpdate(this.podId(), tableName, payload));
    }
    delete(tableName) {
        return this.client.request(() => TablesService.tableDelete(this.podId(), tableName));
    }
    columns = {
        add: (tableName, request) => {
            const payload = "column" in request ? request : { column: request };
            return this.client.request(() => TablesService.tableColumnAdd(this.podId(), tableName, payload));
        },
        remove: (tableName, columnName) => this.client.request(() => TablesService.tableColumnRemove(this.podId(), tableName, columnName)),
    };
}
