// lemma-sdk/dist/namespaces/workflows.js
import { WorkflowsService } from "../openapi_client/services/WorkflowsService.js";
export class WorkflowsNamespace {
    client;
    http;
    podId;
    constructor(client, http, podId) {
        this.client = client;
        this.http = http;
        this.podId = podId;
    }
    list(options = {}) {
        return this.client.request(() => WorkflowsService.workflowList(this.podId(), options.limit ?? 100, options.pageToken));
    }
    create(payload) {
        return this.client.request(() => WorkflowsService.workflowCreate(this.podId(), payload));
    }
    get(workflowName) {
        return this.client.request(() => WorkflowsService.workflowGet(this.podId(), workflowName));
    }
    update(workflowName, payload) {
        return this.client.request(() => WorkflowsService.workflowUpdate(this.podId(), workflowName, payload));
    }
    delete(workflowName) {
        return this.client.request(() => WorkflowsService.workflowDelete(this.podId(), workflowName));
    }
    visualize(workflowName) {
        return this.client.request(() => WorkflowsService.workflowVisualize(this.podId(), workflowName));
    }
    graph = {
        update: (workflowName, graph) => this.client.request(() => WorkflowsService.workflowGraphUpdate(this.podId(), workflowName, graph)),
    };
    runs = {
        /**
         * Create a run. Runs take no inputs: when the workflow starts with a
         * form, the returned run is WAITING with `active_wait` describing the
         * form to render and submit via `runs.submitForm`.
         */
        create: (workflowName) => this.client.request(() => WorkflowsService.workflowRunCreate(this.podId(), workflowName)),
        list: (workflowName, options = {}) => this.client.request(() => WorkflowsService.workflowRunList(this.podId(), workflowName, options.limit ?? 100, options.pageToken)),
        waitingAssignedToMe: (options = {}) => this.client.request(() => WorkflowsService.workflowRunWaitingAssignedToMe(this.podId(), options.limit ?? 100, options.pageToken)),
        get: (runId, podId = this.podId()) => this.client.request(() => WorkflowsService.workflowRunGet(podId, runId)),
        /** Submit the form a run is waiting on. node_id must match active_wait.node_id. */
        submitForm: (runId, payload, podId = this.podId()) => this.client.request(() => WorkflowsService.workflowRunFormSubmit(podId, runId, payload)),
        visualize: (runId, podId = this.podId()) => this.client.request(() => WorkflowsService.workflowRunVisualize(podId, runId)),
        cancel: (runId, podId = this.podId()) => this.client.request(() => WorkflowsService.workflowRunCancel(podId, runId)),
    };
}
