// lemma-sdk/dist/openapi_client/models/ResourceType.js
/* generated using openapi-typescript-codegen -- do not edit */
/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */
export var ResourceType;
(function (ResourceType) {
    ResourceType["ORGANIZATION"] = "organization";
    ResourceType["POD"] = "pod";
    ResourceType["POD_MEMBER"] = "pod_member";
    ResourceType["ROLE"] = "role";
    ResourceType["DATASTORE_TABLE"] = "datastore_table";
    ResourceType["DATASTORE_RECORD"] = "datastore_record";
    ResourceType["FOLDER"] = "folder";
    ResourceType["DOCUMENT"] = "document";
    ResourceType["APP"] = "app";
    ResourceType["AGENT"] = "agent";
    ResourceType["FUNCTION"] = "function";
    ResourceType["WORKFLOW"] = "workflow";
    ResourceType["SCHEDULE"] = "schedule";
    ResourceType["CONVERSATION"] = "conversation";
    ResourceType["CONNECTOR"] = "connector";
    ResourceType["CONNECTOR_ACCOUNT"] = "connector_account";
    ResourceType["CONNECTOR_AUTH_CONFIG"] = "connector_auth_config";
})(ResourceType || (ResourceType = {}));
