// lemma-sdk/dist/auth.js
/**
 * Auth module — cookie-based auth (production) with Bearer token fallback
 * for agent/dev testing.
 *
 * Auth resolution order on init:
 * 1. localStorage.getItem("lemma_token")
 * 2. Session cookie (credentials: "include") — production path
 *
 * If a token is found in (1), all requests use Authorization: Bearer <token>.
 * Otherwise requests rely on cookies, and the server must set the session cookie
 * after the user authenticates at the auth service. In cookie mode we initialise
 * the SuperTokens browser SDK so fetch/XHR automatically handles anti-CSRF and
 * refresh-token flows for mutating requests.
 *
 * Auth state is determined by calling GET /users/me (user.current.get).
 * 401 → unauthenticated. 200 → authenticated.
 */
import Session from "supertokens-web-js/recipe/session";
import { ensureCookieSessionSupport } from "./supertokens.js";
const DEFAULT_BLOCKED_REDIRECT_PATHS = ["/login", "/signup", "/auth"];
const SUPERTOKENS_FRONTEND_MARKER_KEYS = [
    "sFrontToken",
    "st-last-access-token-update",
    "sIRTFrontend",
    "sAntiCsrf",
    "st-access-token",
    "st-refresh-token",
];
const LOCALSTORAGE_TOKEN_KEY = "lemma_token";
function readStorageToken() {
    if (typeof window === "undefined")
        return null;
    try {
        return localStorage.getItem(LOCALSTORAGE_TOKEN_KEY);
    }
    catch {
        return null;
    }
}
function writeStorageToken(token) {
    if (typeof window === "undefined")
        return;
    try {
        localStorage.setItem(LOCALSTORAGE_TOKEN_KEY, token);
    }
    catch {
        // ignore storage errors
    }
}
function removeStorageToken() {
    if (typeof window === "undefined")
        return;
    try {
        localStorage.removeItem(LOCALSTORAGE_TOKEN_KEY);
    }
    catch {
        // ignore storage errors
    }
}
export function setTestingToken(token) {
    writeStorageToken(token);
}
export function getTestingToken() {
    return readStorageToken();
}
export function clearTestingToken() {
    removeStorageToken();
}
function detectInjectedToken() {
    if (typeof window === "undefined")
        return null;
    // 1. localStorage — the only supported browser testing path
    const localToken = readStorageToken();
    if (localToken)
        return localToken;
    return null;
}
function normalizePath(path) {
    const trimmed = path.trim();
    if (!trimmed)
        return "/";
    if (trimmed === "/")
        return "/";
    const withLeadingSlash = trimmed.startsWith("/") ? trimmed : `/${trimmed}`;
    return withLeadingSlash.endsWith("/") ? withLeadingSlash.slice(0, -1) : withLeadingSlash;
}
function resolveAuthPath(basePath, path) {
    const normalizedBase = normalizePath(basePath);
    if (!path || !path.trim()) {
        return normalizedBase;
    }
    const segment = path.trim().replace(/^\/+/, "");
    if (!segment) {
        return normalizedBase;
    }
    return `${normalizedBase}/${segment}`.replace(/\/{2,}/g, "/");
}
function isBlockedLocalPath(pathname, blockedPaths) {
    const normalizedPathname = normalizePath(pathname);
    return blockedPaths.some((rawBlockedPath) => {
        const blockedPath = normalizePath(rawBlockedPath);
        return normalizedPathname === blockedPath || normalizedPathname.startsWith(`${blockedPath}/`);
    });
}
function normalizeOrigin(rawOrigin) {
    const parsed = new URL(rawOrigin);
    return parsed.origin;
}
function normalizeHostnameSuffix(rawSuffix) {
    return rawSuffix.trim().replace(/^\.+/, "").toLowerCase();
}
function isLoopbackHost(hostname) {
    const normalized = hostname.toLowerCase().replace(/^\[/, "").replace(/\]$/, "");
    return normalized === "localhost" || normalized === "127.0.0.1" || normalized === "::1";
}
function isAllowedRedirectOrigin(parsed, siteOrigin, options) {
    if (parsed.origin === siteOrigin) {
        return true;
    }
    if (options.allowLoopback && isLoopbackHost(parsed.hostname)) {
        return true;
    }
    for (const allowedOrigin of options.allowedOrigins ?? []) {
        try {
            if (normalizeOrigin(allowedOrigin) === parsed.origin) {
                return true;
            }
        }
        catch {
            // Ignore invalid allowlist entries.
        }
    }
    const hostname = parsed.hostname.toLowerCase();
    const siteProtocol = new URL(siteOrigin).protocol;
    return (options.allowedOriginSuffixes ?? []).some((rawSuffix) => {
        const suffix = normalizeHostnameSuffix(rawSuffix);
        const hostnameMatches = Boolean(suffix) && (hostname === suffix || hostname.endsWith(`.${suffix}`));
        return hostnameMatches && (siteProtocol !== "https:" || parsed.protocol === "https:");
    });
}
function resolveFallbackRedirectUri(rawFallback, siteOrigin, blockedPaths) {
    const rootFallback = new URL("/", siteOrigin).toString();
    try {
        const parsed = new URL(rawFallback, siteOrigin);
        if (parsed.origin !== siteOrigin || !["http:", "https:"].includes(parsed.protocol)) {
            return rootFallback;
        }
        if (isBlockedLocalPath(parsed.pathname, blockedPaths)) {
            return rootFallback;
        }
        return parsed.toString();
    }
    catch {
        return rootFallback;
    }
}
export function buildAuthUrl(authUrl, options = {}) {
    const url = new URL(authUrl);
    url.pathname = resolveAuthPath(url.pathname, options.path);
    for (const [key, value] of Object.entries(options.params ?? {})) {
        if (value === null || value === undefined)
            continue;
        if (Array.isArray(value)) {
            url.searchParams.delete(key);
            for (const item of value) {
                url.searchParams.append(key, String(item));
            }
            continue;
        }
        url.searchParams.set(key, String(value));
    }
    if (options.mode === "signup") {
        url.searchParams.set("show", "signup");
    }
    if (options.redirectUri && options.redirectUri.trim()) {
        url.searchParams.set("redirect_uri", options.redirectUri);
    }
    return url.toString();
}
export function buildFederatedLogoutUrl(authUrl, options = {}) {
    const url = new URL(authUrl);
    url.pathname = resolveAuthPath(url.pathname, options.path ?? "logout");
    for (const [key, value] of Object.entries(options.params ?? {})) {
        if (value === null || value === undefined)
            continue;
        if (Array.isArray(value)) {
            url.searchParams.delete(key);
            for (const item of value) {
                url.searchParams.append(key, String(item));
            }
            continue;
        }
        url.searchParams.set(key, String(value));
    }
    if (options.redirectUri && options.redirectUri.trim()) {
        url.searchParams.set(options.redirectParam ?? "redirect_uri", options.redirectUri);
    }
    return url.toString();
}
export function resolveSafeRedirectUri(rawValue, options) {
    const siteOrigin = normalizeOrigin(options.siteOrigin);
    const blockedPaths = options.blockedPaths ?? DEFAULT_BLOCKED_REDIRECT_PATHS;
    const fallbackTarget = options.fallback ?? "/";
    const fallback = resolveFallbackRedirectUri(fallbackTarget, siteOrigin, blockedPaths);
    if (!rawValue || !rawValue.trim()) {
        return fallback;
    }
    try {
        const parsed = new URL(rawValue, siteOrigin);
        if (!["http:", "https:"].includes(parsed.protocol)) {
            return fallback;
        }
        if (!isAllowedRedirectOrigin(parsed, siteOrigin, options)) {
            return fallback;
        }
        if (parsed.origin === siteOrigin && isBlockedLocalPath(parsed.pathname, blockedPaths)) {
            return fallback;
        }
        return parsed.toString();
    }
    catch {
        return fallback;
    }
}
export class AuthManager {
    apiUrl;
    authUrl;
    injectedToken;
    state = { status: "loading", user: null };
    listeners = new Set();
    constructor(apiUrl, authUrl) {
        this.apiUrl = apiUrl;
        this.authUrl = authUrl;
        this.injectedToken = detectInjectedToken();
        if (!this.injectedToken) {
            ensureCookieSessionSupport(this.apiUrl, () => this.markUnauthenticated());
        }
    }
    /** Whether requests will use an injected Bearer token (testing mode). */
    get isTokenMode() {
        return this.injectedToken !== null;
    }
    /** The current injected Bearer token, if token-mode auth is active. */
    getBearerToken() {
        return this.injectedToken;
    }
    /** The current auth state. */
    getState() {
        return this.state;
    }
    /** True if currently authenticated (status === "authenticated"). */
    isAuthenticated() {
        return this.state.status === "authenticated";
    }
    /** Subscribe to auth state changes. Returns an unsubscribe function. */
    subscribe(listener) {
        this.listeners.add(listener);
        return () => this.listeners.delete(listener);
    }
    notify() {
        this.listeners.forEach((l) => l(this.state));
    }
    setState(state) {
        this.state = state;
        this.notify();
    }
    assertBrowserContext() {
        if (typeof window === "undefined") {
            throw new Error("This auth method is only available in browser environments.");
        }
    }
    getCookie(name) {
        if (typeof document === "undefined")
            return undefined;
        const escaped = name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
        const match = document.cookie.match(new RegExp(`(?:^|; )${escaped}=([^;]*)`));
        return match ? decodeURIComponent(match[1]) : undefined;
    }
    getCookieDomainCandidates() {
        if (typeof window === "undefined") {
            return [undefined];
        }
        const host = window.location.hostname;
        const isIpv4 = /^\d{1,3}(?:\.\d{1,3}){3}$/.test(host);
        const isIpv6 = host.includes(":");
        if (!host || host === "localhost" || isIpv4 || isIpv6) {
            return [undefined];
        }
        const domains = new Set();
        const parts = host.split(".").filter(Boolean);
        for (let i = 0; i < parts.length - 1; i += 1) {
            const candidate = parts.slice(i).join(".");
            if (!candidate)
                continue;
            domains.add(candidate);
            domains.add(`.${candidate}`);
        }
        return [undefined, ...domains];
    }
    expireCookie(name, domain) {
        if (typeof document === "undefined")
            return;
        const domainPart = domain ? `;domain=${domain}` : "";
        document.cookie = `${name}=;expires=Thu, 01 Jan 1970 00:00:00 GMT;max-age=0;path=/${domainPart};samesite=lax`;
        document.cookie = `${name}=;expires=Thu, 01 Jan 1970 00:00:00 GMT;max-age=0;path=/${domainPart}`;
    }
    /**
     * Defensive cleanup for stale SuperTokens frontend marker cookies/storage.
     * This helps recover when signout/session-expiry paths leave local markers behind.
     */
    clearFrontendSessionMarkers() {
        if (typeof window === "undefined")
            return;
        for (const key of SUPERTOKENS_FRONTEND_MARKER_KEYS) {
            try {
                window.localStorage.removeItem(key);
            }
            catch {
                // ignore storage errors
            }
            try {
                window.sessionStorage.removeItem(key);
            }
            catch {
                // ignore storage errors
            }
        }
        const domains = this.getCookieDomainCandidates();
        for (const key of SUPERTOKENS_FRONTEND_MARKER_KEYS) {
            for (const domain of domains) {
                this.expireCookie(key, domain);
            }
        }
    }
    applyUnauthenticatedState() {
        const next = { status: "unauthenticated", user: null };
        this.setState(next);
        return next;
    }
    clearInjectedToken() {
        this.injectedToken = null;
        clearTestingToken();
    }
    async rawSignOutViaBackend() {
        const antiCsrf = this.getCookie("sAntiCsrf");
        const headers = {
            Accept: "application/json",
            "Content-Type": "application/json",
            rid: "anti-csrf",
            "fdi-version": "4.2",
            "st-auth-mode": "cookie",
        };
        if (antiCsrf) {
            headers["anti-csrf"] = antiCsrf;
        }
        const separator = this.apiUrl.includes("?") ? "&" : "?";
        const signOutUrl = `${this.apiUrl.replace(/\/$/, "")}/st/auth/signout${separator}superTokensDoNotDoInterception=true`;
        await fetch(signOutUrl, {
            method: "POST",
            credentials: "include",
            headers,
        });
    }
    /**
     * Check whether a cookie-backed session is active without mutating auth state.
     */
    async isAuthenticatedViaCookie() {
        if (this.injectedToken) {
            return this.isAuthenticated();
        }
        try {
            const response = await fetch(`${this.apiUrl}/users/me`, {
                method: "GET",
                credentials: "include",
                headers: { Accept: "application/json" },
            });
            return response.status !== 401;
        }
        catch {
            return false;
        }
    }
    /**
     * Return a browser access token from the session layer.
     * Throws if no token is available.
     */
    async getAccessToken() {
        if (this.injectedToken) {
            return this.injectedToken;
        }
        this.assertBrowserContext();
        ensureCookieSessionSupport(this.apiUrl, () => this.markUnauthenticated());
        const token = await Session.getAccessToken();
        if (!token) {
            throw new Error("Token unavailable");
        }
        return token;
    }
    /**
     * Force a refresh-token flow and return the new access token.
     */
    async refreshAccessToken() {
        if (this.injectedToken) {
            return this.injectedToken;
        }
        this.assertBrowserContext();
        ensureCookieSessionSupport(this.apiUrl, () => this.markUnauthenticated());
        const refreshed = await Session.attemptRefreshingSession();
        if (!refreshed) {
            throw new Error("Session refresh failed");
        }
        const token = await Session.getAccessToken();
        if (!token) {
            throw new Error("Token unavailable");
        }
        return token;
    }
    /**
     * Build request headers for an API call.
     * Uses Bearer token if one was injected, otherwise omits Authorization
     * and lets cookies carry the session.
     */
    getRequestInit(init = {}) {
        const headers = {
            "Content-Type": "application/json",
            Accept: "application/json",
            ...init.headers,
        };
        if (this.injectedToken) {
            headers["Authorization"] = `Bearer ${this.injectedToken}`;
        }
        return {
            ...init,
            credentials: this.injectedToken ? "omit" : "include",
            headers,
        };
    }
    /**
     * Call GET /users/me to determine auth state.
     * Sets internal state and notifies listeners.
     */
    async checkAuth() {
        this.setState({ status: "loading", user: null });
        // Cookie mode: short-circuit when no session exists locally instead of
        // hitting /users/me. A 401 there makes the SuperTokens fetch interceptor
        // treat it as "refresh me" and hammer the session-refresh endpoint — and
        // when the app's configured auth domain can't share cookies with the
        // current origin (e.g. an embedded app booting on localhost), that refresh
        // can never succeed, so every app on the pod home storms it endlessly.
        // `doesSessionExist()` reads the local front token only (no network) and
        // returns false when there's nothing to refresh, ending the loop at the source.
        if (!this.injectedToken && typeof window !== "undefined") {
            ensureCookieSessionSupport(this.apiUrl, () => this.markUnauthenticated());
            try {
                if (!(await Session.doesSessionExist())) {
                    return this.applyUnauthenticatedState();
                }
            }
            catch {
                return this.applyUnauthenticatedState();
            }
        }
        try {
            const response = await fetch(`${this.apiUrl}/users/me`, this.getRequestInit({ method: "GET" }));
            // Only 401 means not authenticated — 403 means authenticated but forbidden
            if (response.status === 401) {
                return this.applyUnauthenticatedState();
            }
            if (!response.ok) {
                // For non-401 errors on /users/me, treat as unauthenticated (conservative)
                return this.applyUnauthenticatedState();
            }
            const user = (await response.json());
            const next = { status: "authenticated", user };
            this.setState(next);
            return next;
        }
        catch {
            return this.applyUnauthenticatedState();
        }
    }
    /**
     * Mark the session as unauthenticated (e.g. after a 401 response).
     * Does NOT redirect — call redirectToAuth() explicitly if desired.
     */
    markUnauthenticated() {
        this.applyUnauthenticatedState();
    }
    /**
     * Sign out the current user session.
     * Returns true when the session is no longer active.
     */
    async signOut() {
        if (this.injectedToken) {
            this.clearInjectedToken();
            this.markUnauthenticated();
            return true;
        }
        this.assertBrowserContext();
        ensureCookieSessionSupport(this.apiUrl, () => this.markUnauthenticated());
        try {
            await Session.signOut();
        }
        catch {
            // continue with raw fallback
        }
        if (await this.isAuthenticatedViaCookie()) {
            try {
                await this.rawSignOutViaBackend();
            }
            catch {
                // best effort fallback only
            }
        }
        // Always clear frontend markers on logout attempt, even if backend session
        // cleanup is partial. This avoids stale local "EXISTS" signals.
        this.clearFrontendSessionMarkers();
        const isAuthenticated = await this.isAuthenticatedViaCookie();
        if (!isAuthenticated) {
            this.markUnauthenticated();
        }
        return !isAuthenticated;
    }
    /**
     * Build auth URL for login/signup/custom auth sub-path.
     */
    getAuthUrl(options = {}) {
        return buildAuthUrl(this.authUrl, options);
    }
    /**
     * Build upstream/federated logout URL.
     */
    getFederatedLogoutUrl(options = {}) {
        return buildFederatedLogoutUrl(this.authUrl, options);
    }
    /**
     * Redirect to the auth service, passing the current URL as redirect_uri.
     * After the user authenticates, the auth service should redirect back to
     * the original URL and set the session cookie.
     */
    redirectToAuth(options = {}) {
        if (typeof window === "undefined") {
            return;
        }
        const redirectUri = options.redirectUri ?? window.location.href;
        window.location.href = this.getAuthUrl({ ...options, redirectUri });
    }
    /**
     * Optional full logout flow:
     * 1. clear local SDK/session cookies
     * 2. redirect to auth service logout endpoint to terminate upstream SSO
     */
    async redirectToFederatedLogout(options = {}) {
        this.assertBrowserContext();
        const redirectUri = options.redirectUri ?? window.location.href;
        const localSignOut = options.localSignOut ?? true;
        if (localSignOut) {
            await this.signOut();
        }
        window.location.href = this.getFederatedLogoutUrl({
            ...options,
            redirectUri,
        });
    }
}
