// lemma-sdk/dist/client.js
import { resolveConfig } from "./config.js";
import { AuthManager } from "./auth.js";
import { GeneratedClientAdapter } from "./generated.js";
import { HttpClient } from "./http.js";
import { AgentRuntimeNamespace } from "./namespaces/agent-runtime.js";
import { AgentsNamespace } from "./namespaces/agents.js";
import { ConversationsNamespace } from "./namespaces/conversations.js";
import { AppsNamespace } from "./namespaces/apps.js";
import { FilesNamespace } from "./namespaces/files.js";
import { FunctionsNamespace } from "./namespaces/functions.js";
import { IconsNamespace } from "./namespaces/icons.js";
import { ConnectorsNamespace } from "./namespaces/connectors.js";
import { OrganizationsNamespace } from "./namespaces/organizations.js";
import { PodMembersNamespace } from "./namespaces/pod-members.js";
import { PodPermissionsNamespace } from "./namespaces/pod-permissions.js";
import { PodJoinRequestsNamespace } from "./namespaces/pod-join-requests.js";
import { PodsNamespace } from "./namespaces/pods.js";
import { PodRolesNamespace } from "./namespaces/pod-roles.js";
import { PodSurfacesNamespace } from "./namespaces/pod-surfaces.js";
import { RecordsNamespace } from "./namespaces/records.js";
import { ResourceAccessNamespace } from "./namespaces/resource-access.js";
import { SchedulesNamespace } from "./namespaces/schedules.js";
import { TablesNamespace } from "./namespaces/tables.js";
import { UsersNamespace } from "./namespaces/users.js";
import { WorkflowsNamespace } from "./namespaces/workflows.js";
import { WidgetsNamespace } from "./namespaces/widgets.js";
import { DatastoreNamespace } from "./namespaces/datastore.js";
export { AuthManager };
export class LemmaClient {
    _config;
    _podId;
    _currentPodId;
    /** Auth manager — subscribe to auth state, check auth, redirect to auth. */
    auth;
    _http;
    _generated;
    // Namespaces
    tables;
    records;
    files;
    functions;
    agents;
    agentRuntime;
    conversations;
    workflows;
    apps;
    widgets;
    connectors;
    resourceAccess;
    schedules;
    datastore;
    /** Alias of {@link datastore}, matching the Python SDK's `pod.queries`. */
    queries;
    users;
    icons;
    pods;
    podMembers;
    podPermissions;
    podJoinRequests;
    podRoles;
    organizations;
    podSurfaces;
    constructor(overrides = {}, internalOptions = {}) {
        this._config = resolveConfig(overrides);
        this._currentPodId = this._config.podId;
        this._podId = this._config.podId;
        this.auth = internalOptions.authManager ?? new AuthManager(this._config.apiUrl, this._config.authUrl);
        this._http = new HttpClient(this._config.apiUrl, this.auth, {
            timeoutMs: this._config.timeoutMs,
            maxRetries: this._config.maxRetries,
        });
        this._generated = new GeneratedClientAdapter(this._config.apiUrl, this.auth, {
            maxRetries: this._config.maxRetries,
            timeoutMs: this._config.timeoutMs,
        });
        const podIdFn = () => {
            if (!this._currentPodId) {
                throw new Error("pod_id is required. Pass podId in the constructor or call client.setPodId(id).");
            }
            return this._currentPodId;
        };
        this.tables = new TablesNamespace(this._generated, podIdFn);
        this.records = new RecordsNamespace(this._generated, podIdFn);
        this.files = new FilesNamespace(this._generated, this._http, podIdFn);
        this.functions = new FunctionsNamespace(this._generated, podIdFn);
        this.agents = new AgentsNamespace(this._generated, podIdFn, () => this.conversations);
        this.agentRuntime = new AgentRuntimeNamespace(this._generated);
        this.conversations = new ConversationsNamespace(this._http, podIdFn);
        this.workflows = new WorkflowsNamespace(this._generated, this._http, podIdFn);
        this.apps = new AppsNamespace(this._generated, this._http, podIdFn);
        this.widgets = new WidgetsNamespace(this._http, podIdFn);
        this.connectors = new ConnectorsNamespace(this._generated, this._http);
        this.resourceAccess = new ResourceAccessNamespace(this._generated, podIdFn);
        this.schedules = new SchedulesNamespace(this._generated, podIdFn);
        this.datastore = new DatastoreNamespace(this._generated, podIdFn, this._config.apiUrl, this.auth);
        this.queries = this.datastore;
        this.users = new UsersNamespace(this._generated);
        this.icons = new IconsNamespace(this._generated);
        this.pods = new PodsNamespace(this._generated, this._http);
        this.podMembers = new PodMembersNamespace(this._generated);
        this.podPermissions = new PodPermissionsNamespace(this._generated, this._http, podIdFn);
        this.podJoinRequests = new PodJoinRequestsNamespace(this._generated);
        this.podRoles = new PodRolesNamespace(this._generated, podIdFn);
        this.organizations = new OrganizationsNamespace(this._generated, this._http);
        this.podSurfaces = new PodSurfacesNamespace(this._generated);
    }
    /** Change the active pod ID for subsequent calls. */
    setPodId(podId) {
        this._currentPodId = podId;
    }
    /** Return a new client scoped to a specific pod, sharing auth state. */
    withPod(podId) {
        return new LemmaClient({ ...this._config, podId }, { authManager: this.auth });
    }
    get podId() {
        return this._currentPodId;
    }
    get apiUrl() {
        return this._config.apiUrl;
    }
    get authUrl() {
        return this._config.authUrl;
    }
    /**
     * Initialize the client by checking auth state.
     * Call this once on app startup (or let AuthGuard handle it).
     */
    async initialize() {
        return this.auth.checkAuth();
    }
    /** Raw HTTP request — escape hatch for operations not covered by namespaces. */
    request(method, path, options) {
        return this._http.request(method, path, options);
    }
}
