// lemma-sdk/dist/config.js
function fromEnv(key) {
    // Vite: import.meta.env.VITE_*
    // CRA / webpack: process.env.REACT_APP_*
    // Node: process.env.*
    try {
        // @ts-ignore — import.meta is valid in ESM/Vite builds; try/catch guards CJS bundles
        const meta = import.meta.env; // eslint-disable-line
        if (meta) {
            return (meta[`VITE_LEMMA_${key}`] ??
                meta[`REACT_APP_LEMMA_${key}`] ??
                meta[`LEMMA_${key}`]);
        }
    }
    catch {
        // not available in CJS/browser bundle context
    }
    try {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const env = globalThis.process?.env;
        if (env) {
            return env[`LEMMA_${key}`];
        }
    }
    catch {
        // not available
    }
    return undefined;
}
function windowConfig() {
    if (typeof window !== "undefined" && window.__LEMMA_CONFIG__) {
        return window.__LEMMA_CONFIG__;
    }
    return {};
}
export function resolveConfig(overrides = {}) {
    const win = windowConfig();
    const apiUrl = overrides.apiUrl ??
        win.apiUrl ??
        fromEnv("API_URL") ??
        "https://api.lemma.work";
    const authUrl = overrides.authUrl ??
        win.authUrl ??
        fromEnv("AUTH_URL") ??
        "https://lemma.work/auth";
    const podId = overrides.podId ??
        win.podId ??
        fromEnv("POD_ID");
    return {
        apiUrl: apiUrl.replace(/\/$/, ""),
        authUrl: authUrl.replace(/\/$/, ""),
        podId,
        timeoutMs: overrides.timeoutMs ?? win.timeoutMs,
        maxRetries: overrides.maxRetries ?? win.maxRetries,
    };
}
