// lemma-sdk/dist/generated.js
import { NetworkError, apiErrorFromStatus } from "./http.js";
import { ApiError as GeneratedApiError } from "./openapi_client/core/ApiError.js";
import { CancelablePromise } from "./openapi_client/core/CancelablePromise.js";
import { OpenAPI } from "./openapi_client/core/OpenAPI.js";
import { retryDelayForStatus, sleep } from "./run-utils.js";
import { CLIENT_HEADER_NAME, CLIENT_HEADER_VALUE } from "./version.js";
const DEFAULT_MAX_RETRIES = 2;
const DEFAULT_TIMEOUT_MS = 30_000;
function extractMessage(body, fallback) {
    if (body && typeof body === "object" && typeof body.message === "string") {
        return body.message;
    }
    return fallback;
}
function extractCode(body) {
    if (body && typeof body === "object" && typeof body.code === "string") {
        return body.code;
    }
    return undefined;
}
function extractDetails(body) {
    if (body && typeof body === "object" && "details" in body) {
        return body.details;
    }
    return undefined;
}
export class GeneratedClientAdapter {
    apiUrl;
    auth;
    maxRetries;
    timeoutMs;
    constructor(apiUrl, auth, options = {}) {
        this.apiUrl = apiUrl;
        this.auth = auth;
        this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
        this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    }
    configure() {
        OpenAPI.BASE = this.apiUrl;
        OpenAPI.WITH_CREDENTIALS = true;
        OpenAPI.CREDENTIALS = this.auth.isTokenMode ? "omit" : "include";
        OpenAPI.TOKEN = this.auth.getBearerToken() ?? undefined;
        OpenAPI.HEADERS = { [CLIENT_HEADER_NAME]: CLIENT_HEADER_VALUE };
    }
    async request(operation) {
        this.configure();
        for (let attempt = 0;; attempt++) {
            try {
                return await this.runWithTimeout(operation);
            }
            catch (error) {
                if (error instanceof GeneratedApiError) {
                    if (error.status === 401) {
                        this.auth.markUnauthenticated();
                    }
                    // Retry transient gateway/rate-limit statuses. The generated client
                    // doesn't expose response headers, so there's no Retry-After to honor
                    // (retryDelayForStatus falls back to jittered backoff).
                    const retryDelay = retryDelayForStatus(error.status, attempt, this.maxRetries, null);
                    if (retryDelay !== null) {
                        await sleep(retryDelay);
                        continue;
                    }
                    throw apiErrorFromStatus(error.status, extractMessage(error.body, error.message), extractCode(error.body), extractDetails(error.body), error.body);
                }
                // Not an HTTP error (timeout/cancellation, programming error, raw
                // transport failure) — preserve it untouched.
                throw error;
            }
        }
    }
    /**
     * Enforce a per-attempt timeout on the generated client (which exposes no
     * timeout of its own). The generated operation returns a CancelablePromise;
     * on timeout we cancel it (aborting the underlying fetch) and surface a
     * NetworkError, matching HttpClient.fetchWithTimeout. Non-cancelable or
     * disabled-timeout cases fall through untouched.
     */
    async runWithTimeout(operation) {
        const op = operation();
        if (this.timeoutMs <= 0 || !(op instanceof CancelablePromise)) {
            return op;
        }
        let timer;
        try {
            return await Promise.race([
                op,
                new Promise((_, reject) => {
                    timer = setTimeout(() => {
                        op.cancel();
                        reject(new NetworkError(`Request timed out after ${this.timeoutMs}ms`));
                    }, this.timeoutMs);
                }),
            ]);
        }
        finally {
            if (timer) {
                clearTimeout(timer);
            }
        }
    }
}
