// lemma-sdk/dist/http.js
/**
 * Thin HTTP layer that wraps fetch with auth injection, error handling,
 * and automatic 401→unauthenticated state propagation.
 */
import { retryDelayForStatus, serverRetryAfterMs, sleep } from "./run-utils.js";
import { CLIENT_HEADER_NAME, CLIENT_HEADER_VALUE } from "./version.js";
const DEFAULT_TIMEOUT_MS = 30_000;
const DEFAULT_MAX_RETRIES = 2;
export class ApiError extends Error {
    statusCode;
    code;
    details;
    rawResponse;
    /** Server correlation id (X-Request-Id) when present — quote it in bug reports. */
    requestId;
    constructor(statusCode, message, code, details, rawResponse) {
        super(message);
        this.statusCode = statusCode;
        this.code = code;
        this.details = details;
        this.rawResponse = rawResponse;
        this.name = "ApiError";
    }
}
/** 401 — session missing/expired. */
export class UnauthorizedError extends ApiError {
    name = "UnauthorizedError";
}
/** 403 — authenticated but not permitted (often an RLS/grant denial). */
export class ForbiddenError extends ApiError {
    name = "ForbiddenError";
}
/** 404 — resource not found. */
export class NotFoundError extends ApiError {
    name = "NotFoundError";
}
/** 409 — conflict (e.g. duplicate name). */
export class ConflictError extends ApiError {
    name = "ConflictError";
}
/** 429 — rate limited; `retryAfterMs` is the server-advised wait if provided. */
export class RateLimitError extends ApiError {
    retryAfterMs;
    name = "RateLimitError";
    constructor(message, code, details, rawResponse, retryAfterMs) {
        super(429, message, code, details, rawResponse);
        this.retryAfterMs = retryAfterMs;
    }
}
/** 5xx — server-side error. */
export class ServerError extends ApiError {
    name = "ServerError";
}
/** Transport-level failure (DNS, connection refused, timeout) — no HTTP status. */
export class NetworkError extends Error {
    cause;
    name = "NetworkError";
    constructor(message, cause) {
        super(message);
        this.cause = cause;
    }
}
/** Map an HTTP status to the most specific ApiError subclass. */
export function apiErrorFromStatus(status, message, code, details, rawResponse, retryAfterMsValue) {
    switch (status) {
        case 401: return new UnauthorizedError(status, message, code, details, rawResponse);
        case 403: return new ForbiddenError(status, message, code, details, rawResponse);
        case 404: return new NotFoundError(status, message, code, details, rawResponse);
        case 409: return new ConflictError(status, message, code, details, rawResponse);
        case 429: return new RateLimitError(message, code, details, rawResponse, retryAfterMsValue);
        default:
            return status >= 500
                ? new ServerError(status, message, code, details, rawResponse)
                : new ApiError(status, message, code, details, rawResponse);
    }
}
export class HttpClient {
    apiUrl;
    auth;
    timeoutMs;
    maxRetries;
    constructor(apiUrl, auth, options = {}) {
        this.apiUrl = apiUrl;
        this.auth = auth;
        this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
        this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
    }
    getBaseUrl() {
        return this.apiUrl;
    }
    /** fetch with a default timeout, normalizing transport failures into NetworkError. */
    async fetchWithTimeout(url, init, userSignal) {
        const controller = new AbortController();
        let timedOut = false;
        const timer = setTimeout(() => {
            timedOut = true;
            controller.abort();
        }, this.timeoutMs);
        const onAbort = () => controller.abort();
        if (userSignal) {
            if (userSignal.aborted)
                controller.abort();
            else
                userSignal.addEventListener("abort", onAbort, { once: true });
        }
        try {
            return await fetch(url, { ...init, signal: controller.signal });
        }
        catch (error) {
            if (timedOut) {
                throw new NetworkError(`Request timed out after ${this.timeoutMs}ms`, error);
            }
            if (userSignal?.aborted) {
                throw error; // caller-initiated abort — propagate the original AbortError
            }
            throw new NetworkError(`Network request failed: ${String(error?.message ?? error)}`, error);
        }
        finally {
            clearTimeout(timer);
            userSignal?.removeEventListener("abort", onAbort);
        }
    }
    buildUrl(path, params) {
        let url = `${this.apiUrl}${path}`;
        if (!params) {
            return url;
        }
        const qs = Object.entries(params)
            .filter(([, value]) => value !== undefined && value !== null)
            .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`)
            .join("&");
        if (qs) {
            url += `?${qs}`;
        }
        return url;
    }
    mergeHeaders(base, extra) {
        if (!extra) {
            return base;
        }
        const merged = new Headers(base.headers ?? {});
        const extraHeaders = new Headers(extra);
        extraHeaders.forEach((value, key) => merged.set(key, value));
        return {
            ...base,
            headers: merged,
        };
    }
    async parseError(response) {
        let message = response.statusText || "Request failed";
        let code;
        let details;
        let raw = null;
        try {
            const body = await response.json();
            raw = body;
            if (body && typeof body === "object") {
                const record = body;
                if (typeof record.message === "string") {
                    message = record.message;
                }
                if (typeof record.code === "string") {
                    code = record.code;
                }
                details = record.details;
            }
        }
        catch {
            // non-JSON error body
        }
        const retryMs = response.status === 429
            ? serverRetryAfterMs(response.headers.get("retry-after")) ?? undefined
            : undefined;
        const error = apiErrorFromStatus(response.status, message, code, details, raw, retryMs);
        error.requestId = response.headers.get("x-request-id") ?? undefined;
        return error;
    }
    getRequestBody(options) {
        if (options.body === undefined) {
            return undefined;
        }
        if (options.isFormData && options.body instanceof FormData) {
            return options.body;
        }
        return JSON.stringify(options.body);
    }
    buildRequestInit(method, options) {
        const initBase = {
            method,
            body: this.getRequestBody(options),
            signal: options.signal,
        };
        // For FormData, let the browser set Content-Type with boundary.
        const withAuth = options.isFormData
            ? {
                ...this.auth.getRequestInit(initBase),
                headers: Object.fromEntries(Object.entries(this.auth.getRequestInit(initBase).headers ?? {}).filter(([key]) => key.toLowerCase() !== "content-type")),
            }
            : this.auth.getRequestInit(initBase);
        const withClient = this.mergeHeaders(withAuth, { [CLIENT_HEADER_NAME]: CLIENT_HEADER_VALUE });
        return this.mergeHeaders(withClient, options.headers);
    }
    async request(method, path, options = {}) {
        const url = this.buildUrl(path, options.params);
        const init = this.buildRequestInit(method, options);
        for (let attempt = 0;; attempt++) {
            const response = await this.fetchWithTimeout(url, init, options.signal);
            // Only 401 means the session is gone — 403 is a permission/RLS error, not an auth failure
            if (response.status === 401) {
                this.auth.markUnauthenticated();
            }
            const retryDelay = retryDelayForStatus(response.status, attempt, this.maxRetries, response.headers.get("retry-after"));
            if (retryDelay !== null) {
                await sleep(retryDelay, options.signal);
                continue;
            }
            if (!response.ok) {
                throw await this.parseError(response);
            }
            if (response.status === 204) {
                return undefined;
            }
            const contentType = response.headers.get("content-type") ?? "";
            if (contentType.includes("application/json")) {
                return response.json();
            }
            return response.text();
        }
    }
    async stream(path, options = {}) {
        // Streams are deliberately timeout-exempt (they're long-lived), but we still
        // normalize transport failures into NetworkError so the typed-error contract
        // holds on the SSE path too.
        let response;
        try {
            response = await fetch(this.buildUrl(path, options.params), this.buildRequestInit(options.method ?? "GET", {
                ...options,
                headers: {
                    Accept: "text/event-stream",
                    ...options.headers,
                },
            }));
        }
        catch (error) {
            if (options.signal?.aborted) {
                throw error; // caller-initiated abort — propagate the original AbortError
            }
            throw new NetworkError(`Network request failed: ${String(error?.message ?? error)}`, error);
        }
        if (response.status === 401) {
            this.auth.markUnauthenticated();
        }
        if (!response.ok) {
            throw await this.parseError(response);
        }
        if (!response.body) {
            throw new ApiError(response.status, "Stream response had no body.");
        }
        return response.body;
    }
    async requestBytes(method, path) {
        const url = `${this.apiUrl}${path}`;
        const response = await this.fetchWithTimeout(url, this.auth.getRequestInit({ method }));
        if (response.status === 401) {
            this.auth.markUnauthenticated();
        }
        if (!response.ok) {
            throw await this.parseError(response);
        }
        return response.blob();
    }
}
