// lemma-sdk/dist/react/generated/agents.js
// GENERATED by scripts/generate_l2.mjs from x-lemma metadata — do not edit by hand.
// Resource: agent. Regenerate with: npm run generate:l2
//
// Mechanical TanStack-Query hooks over the agents facade. Query keys and the mutation
// invalidation graph are derived from the backend's x-lemma metadata, so a mutation
// refreshes exactly the lists/entities it should — no manual refetch. Types are pulled
// from the real facade via Parameters<>, so the compiler verifies every call.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { resolvePodClient, resolvePodId } from "../utils.js";
export function useAgentCreate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.agents.create(vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["agent", pid] });
        },
    });
}
export function useAgentDelete(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.agents.delete(vars.agentName),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["agent", pid] });
        },
    });
}
export function useAgentGet(client, podId, agentName, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["agent", pid, "get", agentName],
        queryFn: () => scoped.agents.get(agentName),
        ...queryOptions,
    });
}
export function useAgentList(client, podId, options, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["agent", pid, "list", options ?? null],
        queryFn: () => scoped.agents.list(options),
        ...queryOptions,
    });
}
export function useAgentUpdate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.agents.update(vars.agentName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["agent", pid] });
        },
    });
}
