// lemma-sdk/dist/react/generated/apps.js
// GENERATED by scripts/generate_l2.mjs from x-lemma metadata — do not edit by hand.
// Resource: app. Regenerate with: npm run generate:l2
//
// Mechanical TanStack-Query hooks over the apps facade. Query keys and the mutation
// invalidation graph are derived from the backend's x-lemma metadata, so a mutation
// refreshes exactly the lists/entities it should — no manual refetch. Types are pulled
// from the real facade via Parameters<>, so the compiler verifies every call.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { resolvePodClient, resolvePodId } from "../utils.js";
export function useAppCreate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.apps.create(vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["app", pid] });
        },
    });
}
export function useAppDelete(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.apps.delete(vars.appName),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["app", pid] });
        },
    });
}
export function useAppGet(client, podId, appName, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["app", pid, "get", appName],
        queryFn: () => scoped.apps.get(appName),
        ...queryOptions,
    });
}
export function useAppList(client, podId, options, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["app", pid, "list", options ?? null],
        queryFn: () => scoped.apps.list(options),
        ...queryOptions,
    });
}
export function useAppUpdate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.apps.update(vars.appName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["app", pid] });
        },
    });
}
