// lemma-sdk/dist/react/generated/functions.js
// GENERATED by scripts/generate_l2.mjs from x-lemma metadata — do not edit by hand.
// Resource: function. Regenerate with: npm run generate:l2
//
// Mechanical TanStack-Query hooks over the functions facade. Query keys and the mutation
// invalidation graph are derived from the backend's x-lemma metadata, so a mutation
// refreshes exactly the lists/entities it should — no manual refetch. Types are pulled
// from the real facade via Parameters<>, so the compiler verifies every call.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { resolvePodClient, resolvePodId } from "../utils.js";
export function useFunctionCreate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.functions.create(vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["function", pid] });
        },
    });
}
export function useFunctionDelete(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.functions.delete(vars.functionName),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["function", pid] });
        },
    });
}
export function useFunctionGet(client, podId, functionName, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["function", pid, "get", functionName],
        queryFn: () => scoped.functions.get(functionName),
        ...queryOptions,
    });
}
export function useFunctionList(client, podId, options, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["function", pid, "list", options ?? null],
        queryFn: () => scoped.functions.list(options),
        ...queryOptions,
    });
}
export function useFunctionUpdate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.functions.update(vars.functionName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["function", pid] });
        },
    });
}
