// lemma-sdk/dist/react/generated/records.js
// GENERATED by scripts/generate_l2.mjs from x-lemma metadata — do not edit by hand.
// Resource: record. Regenerate with: npm run generate:l2
//
// Mechanical TanStack-Query hooks over the records facade. Query keys and the mutation
// invalidation graph are derived from the backend's x-lemma metadata, so a mutation
// refreshes exactly the lists/entities it should — no manual refetch. Types are pulled
// from the real facade via Parameters<>, so the compiler verifies every call.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { resolvePodClient, resolvePodId } from "../utils.js";
export function useRecordBulkCreate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.records.bulk.create(vars.tableName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["record", pid] });
        },
    });
}
export function useRecordBulkDelete(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.records.bulk.delete(vars.tableName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["record", pid] });
        },
    });
}
export function useRecordBulkUpdate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.records.bulk.update(vars.tableName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["record", pid] });
        },
    });
}
export function useRecordCreate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.records.create(vars.tableName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["record", pid] });
        },
    });
}
export function useRecordDelete(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.records.delete(vars.tableName, vars.recordId),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["record", pid] });
        },
    });
}
export function useRecordGet(client, podId, tableName, recordId, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["record", pid, "get", tableName, recordId],
        queryFn: () => scoped.records.get(tableName, recordId),
        ...queryOptions,
    });
}
export function useRecordList(client, podId, tableName, options, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["record", pid, "list", tableName, options ?? null],
        queryFn: () => scoped.records.list(tableName, options),
        ...queryOptions,
    });
}
export function useRecordUpdate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.records.update(vars.tableName, vars.recordId, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["record", pid] });
        },
    });
}
