// lemma-sdk/dist/react/generated/schedules.js
// GENERATED by scripts/generate_l2.mjs from x-lemma metadata — do not edit by hand.
// Resource: schedule. Regenerate with: npm run generate:l2
//
// Mechanical TanStack-Query hooks over the schedules facade. Query keys and the mutation
// invalidation graph are derived from the backend's x-lemma metadata, so a mutation
// refreshes exactly the lists/entities it should — no manual refetch. Types are pulled
// from the real facade via Parameters<>, so the compiler verifies every call.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { resolvePodClient, resolvePodId } from "../utils.js";
export function useScheduleCreate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.schedules.create(vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["schedule", pid] });
        },
    });
}
export function useScheduleDelete(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.schedules.delete(vars.scheduleId),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["schedule", pid] });
        },
    });
}
export function useScheduleGet(client, podId, scheduleId, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["schedule", pid, "get", scheduleId],
        queryFn: () => scoped.schedules.get(scheduleId),
        ...queryOptions,
    });
}
export function useScheduleList(client, podId, options, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["schedule", pid, "list", options ?? null],
        queryFn: () => scoped.schedules.list(options),
        ...queryOptions,
    });
}
export function useScheduleUpdate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.schedules.update(vars.scheduleId, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["schedule", pid] });
        },
    });
}
