// lemma-sdk/dist/react/generated/tables.js
// GENERATED by scripts/generate_l2.mjs from x-lemma metadata — do not edit by hand.
// Resource: table. Regenerate with: npm run generate:l2
//
// Mechanical TanStack-Query hooks over the tables facade. Query keys and the mutation
// invalidation graph are derived from the backend's x-lemma metadata, so a mutation
// refreshes exactly the lists/entities it should — no manual refetch. Types are pulled
// from the real facade via Parameters<>, so the compiler verifies every call.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { resolvePodClient, resolvePodId } from "../utils.js";
export function useTableCreate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.tables.create(vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["table", pid] });
        },
    });
}
export function useTableDelete(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.tables.delete(vars.tableName),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["table", pid] });
        },
    });
}
export function useTableGet(client, podId, tableName, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["table", pid, "get", tableName],
        queryFn: () => scoped.tables.get(tableName),
        ...queryOptions,
    });
}
export function useTableList(client, podId, options, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["table", pid, "list", options ?? null],
        queryFn: () => scoped.tables.list(options),
        ...queryOptions,
    });
}
export function useTableUpdate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.tables.update(vars.tableName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["table", pid] });
        },
    });
}
