// lemma-sdk/dist/react/generated/workflows.js
// GENERATED by scripts/generate_l2.mjs from x-lemma metadata — do not edit by hand.
// Resource: workflow. Regenerate with: npm run generate:l2
//
// Mechanical TanStack-Query hooks over the workflows facade. Query keys and the mutation
// invalidation graph are derived from the backend's x-lemma metadata, so a mutation
// refreshes exactly the lists/entities it should — no manual refetch. Types are pulled
// from the real facade via Parameters<>, so the compiler verifies every call.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { resolvePodClient, resolvePodId } from "../utils.js";
export function useWorkflowCreate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.workflows.create(vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["workflow", pid] });
        },
    });
}
export function useWorkflowDelete(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.workflows.delete(vars.workflowName),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["workflow", pid] });
        },
    });
}
export function useWorkflowGet(client, podId, workflowName, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["workflow", pid, "get", workflowName],
        queryFn: () => scoped.workflows.get(workflowName),
        ...queryOptions,
    });
}
export function useWorkflowList(client, podId, options, queryOptions) {
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useQuery({
        queryKey: ["workflow", pid, "list", options ?? null],
        queryFn: () => scoped.workflows.list(options),
        ...queryOptions,
    });
}
export function useWorkflowUpdate(client, podId) {
    const queryClient = useQueryClient();
    const scoped = resolvePodClient(client, podId);
    const pid = resolvePodId(client, podId);
    return useMutation({
        mutationFn: (vars) => scoped.workflows.update(vars.workflowName, vars.payload),
        onSuccess: () => {
            void queryClient.invalidateQueries({ queryKey: ["workflow", pid] });
        },
    });
}
