// lemma-sdk/dist/react/useAssistantController.js
import { useCallback, useEffect, useMemo, useRef, useState } from "react";
import { useAssistantRuntime } from "./useAssistantRuntime.js";
import { useAssistantSession } from "./useAssistantSession.js";
const CONVERSATIONS_PAGE_SIZE = 30;
const EMPTY_SCOPE_KEY = JSON.stringify({
    podId: null,
    assistantName: null,
    assistantId: null,
    organizationId: null,
});
function isRecord(value) {
    return !!value && typeof value === "object" && !Array.isArray(value);
}
function parseMaybeJsonObject(value) {
    if (isRecord(value))
        return value;
    if (typeof value === "string") {
        try {
            const parsed = JSON.parse(value);
            return isRecord(parsed) ? parsed : {};
        }
        catch {
            return {};
        }
    }
    return {};
}
function parseMaybeJsonValue(value) {
    if (typeof value !== "string")
        return value;
    try {
        return JSON.parse(value);
    }
    catch {
        return value;
    }
}
function parseTimestampMs(value) {
    if (typeof value === "number" && Number.isFinite(value)) {
        return value;
    }
    if (typeof value === "string") {
        const timestamp = new Date(value).getTime();
        if (Number.isFinite(timestamp) && timestamp > 0) {
            return timestamp;
        }
    }
    return null;
}
function parseDurationMs(value) {
    if (typeof value === "number" && Number.isFinite(value) && value > 0) {
        return Math.round(value);
    }
    if (typeof value === "string") {
        const parsed = Number(value);
        if (Number.isFinite(parsed) && parsed > 0) {
            return Math.round(parsed);
        }
    }
    return undefined;
}
function getFileKey(file) {
    return `${file.name}:${file.size}:${file.lastModified}`;
}
function parseThinkingDurationFromRecord(record) {
    return parseDurationMs(record.duration_ms)
        ?? parseDurationMs(record.durationMs)
        ?? parseDurationMs(record.elapsed_ms)
        ?? parseDurationMs(record.elapsedMs)
        ?? parseDurationMs(record.thought_duration_ms)
        ?? parseDurationMs(record.thoughtDurationMs);
}
function extractThinkingPart(msg) {
    if (msg.kind !== "THINKING")
        return null;
    const text = typeof msg.text === "string" ? msg.text.trim() : "";
    if (!text)
        return null;
    const metadata = getMessageMetadata(msg);
    return {
        text,
        state: "done",
        durationMs: metadata
            ? parseThinkingDurationFromRecord(metadata)
            : undefined,
    };
}
function normalizeToolResult(value) {
    if (isRecord(value))
        return value;
    if (Array.isArray(value))
        return { output: value };
    if (typeof value === "undefined" || value === null)
        return {};
    return { output: value };
}
function getMessageMetadata(msg) {
    return (msg.message_metadata || msg.metadata || undefined);
}
function getNativeToolPayload(msg) {
    const toolName = typeof msg.tool_name === "string" ? msg.tool_name : undefined;
    if (msg.kind === "TOOL_CALL") {
        return {
            kind: "call",
            toolCallId: (typeof msg.tool_call_id === "string" && msg.tool_call_id) || `${msg.id}-tool-call`,
            toolName,
            args: parseMaybeJsonObject(parseMaybeJsonValue(msg.tool_args)),
        };
    }
    if (msg.kind === "TOOL_RETURN") {
        return {
            kind: "result",
            toolCallId: (typeof msg.tool_call_id === "string" && msg.tool_call_id) || `${msg.id}-tool-result`,
            toolName,
            result: normalizeToolResult(msg.tool_result),
        };
    }
    return null;
}
function toolInvocationKey(tool) {
    return `${tool.toolCallId}:${tool.state}`;
}
function mapToolInvocations(msg) {
    const invocations = [];
    const metadata = getMessageMetadata(msg);
    const nativeToolPayload = getNativeToolPayload(msg);
    if (metadata?.message_type === "tool_call") {
        invocations.push({
            toolCallId: metadata.tool_call_id || `${msg.id}-tool-call`,
            toolName: metadata.tool_name || "tool",
            args: metadata.args || {},
            state: "call",
        });
    }
    if (metadata?.message_type === "tool_return") {
        invocations.push({
            toolCallId: metadata.tool_call_id || `${msg.id}-tool-result`,
            toolName: metadata.tool_name || "tool",
            args: metadata.args || {},
            state: "result",
            result: metadata.result,
        });
    }
    if (Array.isArray(msg.tool_calls)) {
        msg.tool_calls.forEach((rawTool, index) => {
            const tool = isRecord(rawTool) ? rawTool : {};
            const fn = isRecord(tool.function) ? tool.function : {};
            const toolName = ((typeof fn.name === "string" && fn.name)
                || (typeof tool.tool_name === "string" && tool.tool_name)
                || (typeof tool.name === "string" && tool.name)
                || "tool");
            const argsRaw = fn.arguments ?? tool.args ?? tool.arguments ?? tool.input;
            invocations.push({
                toolCallId: (typeof tool.id === "string" && tool.id)
                    || (typeof tool.tool_call_id === "string" && tool.tool_call_id)
                    || `${msg.id}-tool-${index}`,
                toolName,
                args: parseMaybeJsonObject(argsRaw),
                state: "call",
            });
        });
    }
    if (nativeToolPayload?.kind === "call") {
        invocations.push({
            toolCallId: nativeToolPayload.toolCallId,
            toolName: nativeToolPayload.toolName || metadata?.tool_name || "tool",
            args: nativeToolPayload.args || metadata?.args || {},
            state: "call",
        });
    }
    if (nativeToolPayload?.kind === "result") {
        invocations.push({
            toolCallId: nativeToolPayload.toolCallId,
            toolName: nativeToolPayload.toolName || metadata?.tool_name || "tool",
            args: metadata?.args || {},
            state: "result",
            result: nativeToolPayload.result || {},
        });
    }
    const seen = new Set();
    return invocations.filter((invocation) => {
        const key = toolInvocationKey(invocation);
        if (seen.has(key))
            return false;
        seen.add(key);
        return true;
    });
}
function mapConversationMessage(msg, options) {
    const toolInvocations = mapToolInvocations(msg);
    const createdAtMs = parseTimestampMs(msg.created_at) ?? undefined;
    const parts = [];
    let content = "";
    // Flat shape: a message is exactly one kind. Thinking renders as a reasoning
    // part; text/notification render as a text part; tool_call/tool_return render
    // via toolInvocations below.
    const thinkingPart = extractThinkingPart(msg);
    if (thinkingPart) {
        parts.push({
            id: `${msg.id}-reasoning`,
            type: "reasoning",
            text: thinkingPart.text,
            state: thinkingPart.state,
            durationMs: thinkingPart.durationMs ?? options?.thinkingDurationMs,
            startedAtMs: createdAtMs,
        });
    }
    else if (msg.kind === "TEXT" || msg.kind === "NOTIFICATION") {
        content = typeof msg.text === "string" ? msg.text.trim() : "";
        if (content) {
            parts.push({
                id: `${msg.id}-text`,
                type: "text",
                text: content,
            });
        }
    }
    toolInvocations.forEach((toolInvocation, index) => {
        parts.push({
            id: `${msg.id}-tool-${index}`,
            type: "tool",
            toolInvocation,
        });
    });
    return {
        id: msg.id,
        role: msg.role === "user" ? "user" : "assistant",
        content,
        toolInvocations,
        parts,
        createdAt: msg.created_at ? new Date(msg.created_at) : new Date(),
        conversation_id: msg.conversation_id,
        sequence: msg.sequence,
        agent_run_id: msg.agent_run_id,
        metadata: msg.metadata ?? null,
        message_metadata: msg.message_metadata ?? null,
        kind: msg.kind,
        tool_call_id: msg.tool_call_id ?? null,
        tool_name: msg.tool_name ?? null,
        tool_args: msg.tool_args ?? null,
        tool_result: msg.tool_result ?? null,
    };
}
function mapConversationMessages(messages) {
    const mappedMessages = [];
    const pendingToolCalls = new Map();
    const estimateThinkingDurationMs = (index) => {
        const message = messages[index];
        if (!message || !extractThinkingPart(message))
            return undefined;
        const startedAtMs = parseTimestampMs(message.created_at);
        if (!startedAtMs)
            return undefined;
        for (let i = index + 1; i < messages.length; i += 1) {
            const nextCreatedAtMs = parseTimestampMs(messages[i]?.created_at);
            if (!nextCreatedAtMs || nextCreatedAtMs <= startedAtMs)
                continue;
            const durationMs = nextCreatedAtMs - startedAtMs;
            if (durationMs > 0 && durationMs <= 30 * 60 * 1000) {
                return durationMs;
            }
            break;
        }
        return undefined;
    };
    messages.forEach((rawMessage, index) => {
        const mappedMessage = mapConversationMessage(rawMessage, {
            thinkingDurationMs: estimateThinkingDurationMs(index),
        });
        mappedMessage.toolInvocations?.forEach((invocation) => {
            if (invocation.state === "call") {
                pendingToolCalls.set(invocation.toolCallId, invocation);
            }
        });
        const nativePayload = getNativeToolPayload(rawMessage);
        const isToolRole = rawMessage.role === "tool";
        if (isToolRole && nativePayload?.kind === "result" && mappedMessage.toolInvocations && mappedMessage.toolInvocations.length > 0) {
            let mergedIntoPriorCall = false;
            mappedMessage.toolInvocations.forEach((resultInvocation) => {
                if (resultInvocation.state !== "result")
                    return;
                const pendingInvocation = pendingToolCalls.get(resultInvocation.toolCallId);
                if (!pendingInvocation)
                    return;
                pendingInvocation.state = "result";
                pendingInvocation.result = resultInvocation.result || {};
                if (pendingInvocation.toolName === "tool" && resultInvocation.toolName !== "tool") {
                    pendingInvocation.toolName = resultInvocation.toolName;
                }
                mergedIntoPriorCall = true;
            });
            if (mergedIntoPriorCall) {
                return;
            }
        }
        if (mappedMessage.toolInvocations) {
            mappedMessage.toolInvocations.forEach((invocation) => {
                if (invocation.state === "result") {
                    const pendingInvocation = pendingToolCalls.get(invocation.toolCallId);
                    if (pendingInvocation) {
                        if ((invocation.toolName === "tool" || !invocation.toolName) && pendingInvocation.toolName) {
                            invocation.toolName = pendingInvocation.toolName;
                        }
                        if (Object.keys(invocation.args).length === 0 && Object.keys(pendingInvocation.args).length > 0) {
                            invocation.args = pendingInvocation.args;
                        }
                    }
                }
            });
        }
        mappedMessages.push(mappedMessage);
    });
    return mappedMessages;
}
function sortConversationsByUpdatedAt(conversations) {
    return [...conversations].sort((a, b) => {
        const aTime = new Date(a.updated_at || a.created_at).getTime();
        const bTime = new Date(b.updated_at || b.created_at).getTime();
        return bTime - aTime;
    });
}
function sortMessagesByCreatedAt(messages) {
    return [...messages].sort((a, b) => {
        const aTime = Number.isFinite(new Date(a.created_at).getTime()) ? new Date(a.created_at).getTime() : 0;
        const bTime = Number.isFinite(new Date(b.created_at).getTime()) ? new Date(b.created_at).getTime() : 0;
        return aTime - bTime;
    });
}
function isConversationRunning(status) {
    if (typeof status !== "string")
        return false;
    const normalized = status.trim().toLowerCase();
    if (!normalized)
        return false;
    if (normalized === "waiting"
        || normalized === "completed"
        || normalized === "failed"
        || normalized === "cancelled"
        || normalized === "stopped") {
        return false;
    }
    return true;
}
function resolveScopedClient(client, podId) {
    if (podId && podId !== client.podId) {
        return client.withPod(podId);
    }
    return client;
}
function conversationUploadDirectory(conversationId) {
    return `/me/conversations/${conversationId}`;
}
function shouldIgnoreFolderEnsureError(error) {
    const message = error instanceof Error ? error.message.toLowerCase() : String(error ?? "").toLowerCase();
    return message.includes("already exists")
        || message.includes("already in use")
        || message.includes("path unavailable")
        || message.includes("path already")
        || message.includes("409");
}
async function ensureFolder(client, name, directoryPath) {
    try {
        await client.files.folder.create(name, { directoryPath });
    }
    catch (error) {
        if (!shouldIgnoreFolderEnsureError(error))
            throw error;
    }
}
async function ensureConversationUploadDirectory(client, conversationId) {
    await ensureFolder(client, "conversations", "/me");
    await ensureFolder(client, conversationId, "/me/conversations");
    return conversationUploadDirectory(conversationId);
}
async function uploadConversationFiles(client, conversationId, uploads, onStatus) {
    const directoryPath = await ensureConversationUploadDirectory(client, conversationId);
    const uploaded = [];
    for (const upload of uploads) {
        onStatus?.(upload.key, { status: "uploading", error: undefined });
        try {
            const response = await client.files.upload(upload.file, {
                name: upload.file.name,
                directoryPath,
                searchEnabled: true,
            });
            onStatus?.(upload.key, { status: "uploaded", path: response.path, error: undefined });
            uploaded.push(response);
        }
        catch (error) {
            onStatus?.(upload.key, {
                status: "failed",
                error: error instanceof Error ? error.message : "Upload failed",
            });
            throw error;
        }
    }
    return uploaded;
}
function formatPersonalFileReferences(files) {
    return files
        .map((file) => {
        const pathParts = file.path.split("/").filter(Boolean);
        const name = file.name || pathParts[pathParts.length - 1] || file.path;
        return `- ${name}: ${file.path}`;
    })
        .join("\n");
}
function appendPersonalFileReferences(content, files) {
    if (files.length === 0)
        return content;
    const references = formatPersonalFileReferences(files);
    return `${content}\n\nPersonal files available to this run:\n${references}`;
}
export function useAssistantController({ client, podId, agentName, assistantName, assistantId, organizationId, enabled = true, instructions, autoLoadMessages = true, }) {
    const [localError, setLocalError] = useState(null);
    const [conversations, setConversations] = useState([]);
    const [activeConversationId, setActiveConversationId] = useState(null);
    const [availableModels, setAvailableModels] = useState([]);
    const [conversationModel, setConversationModelState] = useState(null);
    const [conversationRuntime, setConversationRuntimeState] = useState(null);
    const [isStreaming, setIsStreaming] = useState(false);
    const [isLoadingConversations, setIsLoadingConversations] = useState(false);
    const [isLoadingMoreConversations, setIsLoadingMoreConversations] = useState(false);
    const [conversationsCursor, setConversationsCursor] = useState(null);
    const [isLoadingMessages, setIsLoadingMessages] = useState(false);
    const [isLoadingOlderMessages, setIsLoadingOlderMessages] = useState(false);
    const [isUploadingFiles, setIsUploadingFiles] = useState(false);
    const [pendingFileUploads, setPendingFileUploads] = useState([]);
    const [olderMessagesCursor, setOlderMessagesCursor] = useState(null);
    const activeConversationIdRef = useRef(null);
    const conversationsRef = useRef([]);
    const isStreamingRef = useRef(false);
    const sessionIsStreamingRef = useRef(false);
    const suppressAutoSelectRef = useRef(false);
    const lastAutoLoadedConversationIdRef = useRef(null);
    const loadingConversationIdRef = useRef(null);
    const skipInitialLoadConversationIdsRef = useRef(new Set());
    const loadConversationMessagesRef = useRef(null);
    const resumeIfRunningRef = useRef(null);
    const scope = useMemo(() => ({
        podId: podId ?? null,
        agentName: agentName ?? assistantName ?? assistantId ?? null,
        assistantName: assistantName ?? assistantId ?? null,
        assistantId: assistantId ?? null,
        organizationId: organizationId ?? null,
    }), [agentName, assistantId, assistantName, organizationId, podId]);
    const pendingFiles = useMemo(() => pendingFileUploads.map((upload) => upload.file), [pendingFileUploads]);
    const scopeKey = useMemo(() => JSON.stringify({
        podId: scope.podId ?? null,
        agentName: scope.agentName ?? null,
        assistantName: scope.assistantName ?? null,
        assistantId: scope.assistantId ?? null,
        organizationId: scope.organizationId ?? null,
    }), [scope.agentName, scope.assistantId, scope.assistantName, scope.organizationId, scope.podId]);
    const handleAssistantSessionError = useCallback((sessionError) => {
        setLocalError((prev) => prev || (sessionError instanceof Error ? sessionError.message : "Agent session failed"));
    }, []);
    const assistantSession = useAssistantSession({
        client,
        podId: scope.podId ?? undefined,
        agentName: scope.agentName ?? undefined,
        assistantName: scope.assistantName ?? undefined,
        assistantId: scope.assistantId ?? undefined,
        organizationId: scope.organizationId ?? undefined,
        instructions,
        conversationId: activeConversationId ?? undefined,
        autoLoad: false,
        onError: handleAssistantSessionError,
    });
    const { conversationId: sessionConversationId, listConversations: sessionListConversations, loadMessages: sessionLoadMessages, sendMessage: sessionSendMessage, createConversation: sessionCreateConversation, resumeIfRunning: sessionResumeIfRunning, stop: sessionStop, cancel: sessionCancel, isStreaming: sessionIsStreaming, messages: sessionMessages, streamingText: sessionStreamingText, streamingTool: sessionStreamingTool, status: sessionStatus, } = assistantSession;
    const { runtimeMessages, appendOptimisticUserMessage, replaceLoadedMessages, mergeMessages, clear: clearRuntimeMessages, } = useAssistantRuntime({
        conversationId: activeConversationId,
        sessionConversationId,
        sessionMessages,
    });
    const error = localError;
    const isLoading = isStreaming || sessionIsStreaming;
    const touchConversation = useCallback((conversationId, updates) => {
        setConversations((prev) => {
            const now = new Date().toISOString();
            let found = false;
            const next = prev.map((conversation) => {
                if (conversation.id !== conversationId)
                    return conversation;
                found = true;
                return {
                    ...conversation,
                    ...updates,
                    updated_at: typeof updates?.updated_at === "undefined"
                        ? conversation.updated_at
                        : updates.updated_at || now,
                };
            });
            return found ? sortConversationsByUpdatedAt(next) : next;
        });
    }, []);
    const setConversationModel = useCallback(async (model, runtime) => {
        const nextRuntime = typeof runtime === "undefined"
            ? availableModels.find((entry) => entry.id === model)?.runtime ?? null
            : runtime;
        setConversationModelState(model);
        setConversationRuntimeState(nextRuntime);
        const conversationId = activeConversationIdRef.current;
        if (!conversationId)
            return;
        const knownConversation = conversationsRef.current.find((conversation) => conversation.id === conversationId);
        const resolvedPodId = knownConversation?.pod_id ?? scope.podId;
        const previousModel = knownConversation?.model ?? null;
        const previousRuntime = knownConversation?.agent_runtime ?? null;
        touchConversation(conversationId, {
            model: model,
            agent_runtime: nextRuntime,
        });
        try {
            const updatedConversation = await client.conversations.update(conversationId, model
                ? { model: model, agent_runtime: nextRuntime }
                : { agent_runtime: null }, { pod_id: resolvedPodId ?? undefined });
            touchConversation(conversationId, {
                model: (updatedConversation.model ?? model),
                agent_runtime: updatedConversation.agent_runtime ?? nextRuntime,
                updated_at: updatedConversation.updated_at,
            });
            setConversationModelState((updatedConversation.model ?? model));
            setConversationRuntimeState(updatedConversation.agent_runtime ?? nextRuntime);
        }
        catch (error) {
            touchConversation(conversationId, {
                model: previousModel,
                agent_runtime: previousRuntime,
            });
            setConversationModelState(previousModel);
            setConversationRuntimeState(previousRuntime);
            throw error;
        }
    }, [availableModels, client, scope.podId, touchConversation]);
    const loadConversations = useCallback(async () => {
        setIsLoadingConversations(true);
        try {
            const response = await sessionListConversations({ scope, limit: CONVERSATIONS_PAGE_SIZE });
            const nextConversations = sortConversationsByUpdatedAt(response.items || []);
            setConversations(nextConversations);
            setConversationsCursor(response.next_page_token ?? null);
            setActiveConversationId((current) => {
                if (current && nextConversations.some((conversation) => conversation.id === current)) {
                    return current;
                }
                if (suppressAutoSelectRef.current) {
                    return null;
                }
                return nextConversations[0]?.id ?? null;
            });
        }
        catch (err) {
            setLocalError((prev) => prev || (err instanceof Error ? err.message : "Failed to load conversations"));
        }
        finally {
            setIsLoadingConversations(false);
        }
    }, [scope, sessionListConversations]);
    const loadMoreConversations = useCallback(async () => {
        if (!conversationsCursor || isLoadingConversations || isLoadingMoreConversations) {
            return [];
        }
        setIsLoadingMoreConversations(true);
        try {
            const response = await sessionListConversations({
                scope,
                limit: CONVERSATIONS_PAGE_SIZE,
                pageToken: conversationsCursor,
            });
            const moreConversations = response.items || [];
            setConversations((prev) => {
                const byId = new Map(prev.map((conversation) => [conversation.id, conversation]));
                for (const conversation of moreConversations) {
                    byId.set(conversation.id, conversation);
                }
                return sortConversationsByUpdatedAt(Array.from(byId.values()));
            });
            setConversationsCursor(response.next_page_token ?? null);
            return moreConversations;
        }
        catch (err) {
            setLocalError((prev) => prev || (err instanceof Error ? err.message : "Failed to load more conversations"));
            return [];
        }
        finally {
            setIsLoadingMoreConversations(false);
        }
    }, [conversationsCursor, isLoadingConversations, isLoadingMoreConversations, scope, sessionListConversations]);
    const loadAvailableModels = useCallback(async () => {
        try {
            const response = await client.conversations.listModels({
                orgId: scope.organizationId ?? undefined,
            });
            return response.items ?? [];
        }
        catch {
            return [];
        }
    }, [client, scope.organizationId]);
    const loadConversationMessages = useCallback(async (conversationId) => {
        setIsLoadingMessages(true);
        try {
            const response = await sessionLoadMessages({
                conversationId,
                limit: 100,
            });
            if (activeConversationIdRef.current !== conversationId) {
                return;
            }
            const sorted = sortMessagesByCreatedAt((response.items || []));
            replaceLoadedMessages(sorted);
            setOlderMessagesCursor(response.next_page_token ?? null);
        }
        catch (err) {
            setLocalError((prev) => prev || (err instanceof Error ? err.message : "Failed to load messages"));
            setOlderMessagesCursor(null);
        }
        finally {
            setIsLoadingMessages(false);
        }
    }, [replaceLoadedMessages, sessionLoadMessages]);
    const loadOlderMessages = useCallback(async () => {
        const conversationId = activeConversationIdRef.current;
        const cursor = olderMessagesCursor;
        if (!conversationId || !cursor || isLoadingMessages || isLoadingOlderMessages) {
            return false;
        }
        setIsLoadingOlderMessages(true);
        try {
            const response = await sessionLoadMessages({
                conversationId,
                limit: 100,
                pageToken: cursor,
            });
            if (activeConversationIdRef.current !== conversationId) {
                return false;
            }
            const older = sortMessagesByCreatedAt((response.items || []));
            mergeMessages(older);
            setOlderMessagesCursor(response.next_page_token ?? null);
            return older.length > 0;
        }
        catch (err) {
            setLocalError((prev) => prev || (err instanceof Error ? err.message : "Failed to load older messages"));
            return false;
        }
        finally {
            setIsLoadingOlderMessages(false);
        }
    }, [isLoadingMessages, isLoadingOlderMessages, mergeMessages, olderMessagesCursor, sessionLoadMessages]);
    useEffect(() => {
        loadConversationMessagesRef.current = loadConversationMessages;
    }, [loadConversationMessages]);
    useEffect(() => {
        resumeIfRunningRef.current = sessionResumeIfRunning;
    }, [sessionResumeIfRunning]);
    useEffect(() => {
        activeConversationIdRef.current = activeConversationId;
    }, [activeConversationId]);
    useEffect(() => {
        conversationsRef.current = conversations;
    }, [conversations]);
    useEffect(() => {
        isStreamingRef.current = isStreaming;
    }, [isStreaming]);
    useEffect(() => {
        sessionIsStreamingRef.current = sessionIsStreaming;
    }, [sessionIsStreaming]);
    useEffect(() => {
        if (!enabled) {
            setAvailableModels([]);
            return;
        }
        let cancelled = false;
        void loadAvailableModels()
            .then((models) => {
            if (cancelled)
                return;
            setAvailableModels(models);
        })
            .catch(() => undefined);
        return () => {
            cancelled = true;
        };
    }, [enabled, loadAvailableModels]);
    const messages = useMemo(() => {
        if (!activeConversationId)
            return [];
        const normalized = sortMessagesByCreatedAt(runtimeMessages)
            .filter((message) => message.conversation_id === activeConversationId);
        if (normalized.length === 0)
            return [];
        const nextMessages = mapConversationMessages(normalized);
        const pendingText = sessionStreamingText.trim();
        if (pendingText.length > 0) {
            const streamingId = `streaming-${activeConversationId}`;
            nextMessages.push({
                id: streamingId,
                role: "assistant",
                content: pendingText,
                createdAt: new Date(),
                parts: [{ id: `${streamingId}-text`, type: "text", text: pendingText }],
            });
        }
        return nextMessages;
    }, [activeConversationId, runtimeMessages, sessionStreamingText]);
    useEffect(() => {
        if (!activeConversationId)
            return;
        if (!sessionStatus)
            return;
        touchConversation(activeConversationId, {
            status: sessionStatus,
        });
    }, [activeConversationId, sessionStatus, touchConversation]);
    useEffect(() => {
        if (!activeConversationId)
            return;
        const activeConversation = conversations.find((conversation) => conversation.id === activeConversationId);
        if (!activeConversation)
            return;
        setConversationModelState(activeConversation.model ?? null);
        setConversationRuntimeState(activeConversation.agent_runtime ?? null);
    }, [activeConversationId, conversations]);
    useEffect(() => {
        if (!enabled) {
            sessionCancel();
            clearRuntimeMessages();
            suppressAutoSelectRef.current = false;
            activeConversationIdRef.current = null;
            lastAutoLoadedConversationIdRef.current = null;
            loadingConversationIdRef.current = null;
            skipInitialLoadConversationIdsRef.current.clear();
            setActiveConversationId(null);
            setAvailableModels([]);
            setConversationModelState(null);
            setConversationRuntimeState(null);
            setConversations([]);
            setConversationsCursor(null);
            setLocalError(null);
            setOlderMessagesCursor(null);
            setIsLoadingConversations(false);
            setIsLoadingMoreConversations(false);
            setIsLoadingMessages(false);
            setIsLoadingOlderMessages(false);
            return;
        }
        suppressAutoSelectRef.current = false;
        activeConversationIdRef.current = null;
        lastAutoLoadedConversationIdRef.current = null;
        loadingConversationIdRef.current = null;
        skipInitialLoadConversationIdsRef.current.clear();
        setActiveConversationId(null);
        setConversationModelState(null);
        setConversationRuntimeState(null);
        setConversations([]);
        setConversationsCursor(null);
        setLocalError(null);
        clearRuntimeMessages();
        setOlderMessagesCursor(null);
        if (scopeKey !== EMPTY_SCOPE_KEY) {
            void loadConversations();
        }
    }, [clearRuntimeMessages, enabled, loadConversations, scopeKey, sessionCancel]);
    useEffect(() => {
        if (!enabled || !activeConversationId) {
            clearRuntimeMessages();
            lastAutoLoadedConversationIdRef.current = null;
            loadingConversationIdRef.current = null;
            setOlderMessagesCursor(null);
            setIsLoadingMessages(false);
            return;
        }
        if (!autoLoadMessages) {
            clearRuntimeMessages();
            lastAutoLoadedConversationIdRef.current = null;
            loadingConversationIdRef.current = null;
            setOlderMessagesCursor(null);
            setIsLoadingMessages(false);
            return;
        }
        if (skipInitialLoadConversationIdsRef.current.has(activeConversationId)) {
            skipInitialLoadConversationIdsRef.current.delete(activeConversationId);
            lastAutoLoadedConversationIdRef.current = activeConversationId;
            return;
        }
        if (lastAutoLoadedConversationIdRef.current === activeConversationId) {
            return;
        }
        if (loadingConversationIdRef.current === activeConversationId) {
            return;
        }
        let cancelled = false;
        loadingConversationIdRef.current = activeConversationId;
        const loadConversation = async () => {
            setOlderMessagesCursor(null);
            await loadConversationMessagesRef.current?.(activeConversationId);
            if (cancelled)
                return;
            lastAutoLoadedConversationIdRef.current = activeConversationId;
            try {
                await resumeIfRunningRef.current?.(activeConversationId);
            }
            catch (error) {
                if (cancelled)
                    return;
                setLocalError((prev) => prev || (error instanceof Error ? error.message : "Failed to resume conversation"));
            }
        };
        void loadConversation().finally(() => {
            if (loadingConversationIdRef.current === activeConversationId) {
                loadingConversationIdRef.current = null;
            }
        });
        return () => {
            cancelled = true;
        };
    }, [activeConversationId, autoLoadMessages, clearRuntimeMessages, enabled]);
    const stop = useCallback(() => {
        const hadActiveStream = sessionIsStreamingRef.current || isStreamingRef.current;
        sessionCancel();
        setIsStreaming(false);
        const conversationId = activeConversationIdRef.current;
        if (!conversationId)
            return;
        const activeConversation = conversationsRef.current.find((conversation) => conversation.id === conversationId);
        const conversationIsRunning = isConversationRunning(activeConversation?.status);
        if (!hadActiveStream && !conversationIsRunning)
            return;
        const previousStatus = activeConversation?.status;
        touchConversation(conversationId, { status: "waiting" });
        void sessionStop(conversationId).catch((error) => {
            touchConversation(conversationId, { status: previousStatus });
            setLocalError((prev) => prev || (error instanceof Error ? error.message : "Failed to stop conversation"));
        });
    }, [sessionCancel, sessionStop, touchConversation]);
    const selectConversation = useCallback((conversationId) => {
        if (sessionIsStreamingRef.current || isStreamingRef.current) {
            sessionCancel();
            setIsStreaming(false);
        }
        const currentConversationId = activeConversationIdRef.current;
        if (conversationId && conversationId === currentConversationId) {
            if (!autoLoadMessages) {
                setLocalError(null);
                setOlderMessagesCursor(null);
                setIsLoadingMessages(false);
                return;
            }
            if (loadingConversationIdRef.current === conversationId
                || lastAutoLoadedConversationIdRef.current === conversationId) {
                return;
            }
            loadingConversationIdRef.current = conversationId;
            setLocalError(null);
            setOlderMessagesCursor(null);
            setIsLoadingMessages(true);
            void loadConversationMessagesRef.current?.(conversationId)
                .then(() => resumeIfRunningRef.current?.(conversationId))
                .catch((error) => {
                setLocalError((prev) => prev || (error instanceof Error ? error.message : "Failed to resume conversation"));
            })
                .finally(() => {
                if (loadingConversationIdRef.current === conversationId) {
                    loadingConversationIdRef.current = null;
                }
                lastAutoLoadedConversationIdRef.current = conversationId;
            });
            return;
        }
        suppressAutoSelectRef.current = conversationId === null;
        setLocalError(null);
        activeConversationIdRef.current = conversationId;
        lastAutoLoadedConversationIdRef.current = null;
        loadingConversationIdRef.current = null;
        setOlderMessagesCursor(null);
        clearRuntimeMessages();
        setIsLoadingMessages(Boolean(conversationId && autoLoadMessages));
        setActiveConversationId(conversationId);
    }, [autoLoadMessages, clearRuntimeMessages, sessionCancel]);
    const resetConversationState = useCallback((keepPendingFiles = false) => {
        stop();
        clearRuntimeMessages();
        suppressAutoSelectRef.current = true;
        activeConversationIdRef.current = null;
        lastAutoLoadedConversationIdRef.current = null;
        loadingConversationIdRef.current = null;
        skipInitialLoadConversationIdsRef.current.clear();
        setActiveConversationId(null);
        setLocalError(null);
        setOlderMessagesCursor(null);
        setIsLoadingMessages(false);
        if (!keepPendingFiles) {
            setPendingFileUploads([]);
        }
    }, [clearRuntimeMessages, stop]);
    const clearMessages = useCallback(() => {
        resetConversationState(false);
    }, [resetConversationState]);
    const ensureConversation = useCallback(async (titleSeed, options = {}) => {
        const existingConversationId = activeConversationIdRef.current;
        if (existingConversationId) {
            return existingConversationId;
        }
        const createdConversation = await sessionCreateConversation({
            title: titleSeed.slice(0, 120),
            instructions: typeof options.instructions === "undefined" ? instructions : options.instructions,
            metadata: options.metadata ?? undefined,
            model: conversationModel,
            agentRuntime: conversationRuntime,
            ...scope,
        });
        suppressAutoSelectRef.current = false;
        setConversations((prev) => sortConversationsByUpdatedAt([
            createdConversation,
            ...prev.filter((conversation) => conversation.id !== createdConversation.id),
        ]));
        activeConversationIdRef.current = createdConversation.id;
        lastAutoLoadedConversationIdRef.current = createdConversation.id;
        loadingConversationIdRef.current = null;
        skipInitialLoadConversationIdsRef.current.add(createdConversation.id);
        setActiveConversationId(createdConversation.id);
        setConversationModelState((createdConversation.model ?? conversationModel ?? null));
        setConversationRuntimeState(createdConversation.agent_runtime ?? conversationRuntime ?? null);
        clearRuntimeMessages();
        setOlderMessagesCursor(null);
        return createdConversation.id;
    }, [clearRuntimeMessages, conversationModel, conversationRuntime, instructions, scope, sessionCreateConversation]);
    const queuePendingFiles = useCallback((files) => {
        if (files.length === 0)
            return;
        setPendingFileUploads((prev) => {
            const byKey = new Map();
            prev.forEach((upload) => byKey.set(upload.key, upload));
            files.forEach((file) => {
                const key = getFileKey(file);
                byKey.set(key, {
                    key,
                    file,
                    status: "queued",
                });
            });
            return Array.from(byKey.values());
        });
    }, []);
    const removePendingFile = useCallback((fileKey) => {
        setPendingFileUploads((prev) => prev.filter((upload) => upload.key !== fileKey));
    }, []);
    const clearPendingFiles = useCallback(() => {
        setPendingFileUploads([]);
    }, []);
    const updatePendingFileUpload = useCallback((key, next) => {
        setPendingFileUploads((prev) => prev.map((upload) => (upload.key === key
            ? { ...upload, ...next }
            : upload)));
    }, []);
    const sendMessage = useCallback(async (content, options = {}) => {
        const trimmed = content.trim();
        const uploadsToSend = pendingFileUploads.filter((upload) => upload.status !== "uploaded");
        if (!enabled || (!trimmed && uploadsToSend.length === 0) || isStreaming || sessionIsStreaming)
            return;
        const forceNewConversation = options.forceNewConversation === true;
        setLocalError(null);
        if (forceNewConversation) {
            resetConversationState(true);
        }
        let conversationId = forceNewConversation ? null : activeConversationId;
        try {
            if (!conversationId) {
                conversationId = await ensureConversation(trimmed, {
                    instructions: options.instructions,
                    metadata: options.conversationMetadata,
                });
            }
            if (!conversationId) {
                throw new Error("Conversation could not be initialized");
            }
            const finalConversationId = conversationId;
            let messageContent = trimmed || "Please use the attached files.";
            let uploadedFiles = [];
            if (uploadsToSend.length > 0) {
                setIsUploadingFiles(true);
                try {
                    const fileClient = resolveScopedClient(client, scope.podId);
                    uploadedFiles = await uploadConversationFiles(fileClient, finalConversationId, uploadsToSend, updatePendingFileUpload);
                    messageContent = appendPersonalFileReferences(messageContent, uploadedFiles);
                    setPendingFileUploads([]);
                    touchConversation(finalConversationId, { updated_at: new Date().toISOString() });
                }
                finally {
                    setIsUploadingFiles(false);
                }
            }
            appendOptimisticUserMessage(messageContent, {
                conversationId: finalConversationId,
            });
            setIsStreaming(true);
            touchConversation(finalConversationId, { status: "running" });
            await sessionSendMessage(messageContent, {
                conversationId: finalConversationId,
                metadata: uploadedFiles.length > 0
                    ? {
                        ...(options.metadata ?? {}),
                        attachments: uploadedFiles.map((file) => ({
                            id: file.id,
                            name: file.name,
                            path: file.path,
                            namespace: "PERSONAL",
                            mime_type: file.mime_type,
                        })),
                    }
                    : options.metadata ?? undefined,
            });
            touchConversation(finalConversationId, { updated_at: new Date().toISOString() });
        }
        catch (err) {
            if (err instanceof DOMException && err.name === "AbortError") {
                return;
            }
            setLocalError(err instanceof Error ? err.message : "Failed to send message");
        }
        finally {
            setIsStreaming(false);
        }
    }, [
        activeConversationId,
        appendOptimisticUserMessage,
        enabled,
        ensureConversation,
        isStreaming,
        pendingFileUploads,
        resetConversationState,
        scope.podId,
        sessionIsStreaming,
        sessionSendMessage,
        touchConversation,
        updatePendingFileUpload,
    ]);
    const uploadFiles = useCallback(async (files, options) => {
        const normalizedFiles = files.filter((file) => file instanceof File);
        if (!enabled || normalizedFiles.length === 0 || isLoading || isUploadingFiles)
            return;
        void options;
        setLocalError(null);
        queuePendingFiles(normalizedFiles);
    }, [
        enabled,
        isLoading,
        isUploadingFiles,
        queuePendingFiles,
    ]);
    const resolveUserApproval = useCallback(async (approvalId, decision, response) => {
        if (!enabled)
            return;
        const conversationId = activeConversationIdRef.current;
        if (!conversationId) {
            throw new Error("An active conversation is required to resolve this approval.");
        }
        const knownConversation = conversationsRef.current.find((conversation) => conversation.id === conversationId);
        const resolvedPodId = knownConversation?.pod_id ?? scope.podId;
        setLocalError(null);
        try {
            await client.conversations.approvals.resolve(conversationId, approvalId, { decision, response: response ?? {} }, { pod_id: resolvedPodId ?? undefined });
            await loadConversationMessages(conversationId);
            void sessionResumeIfRunning(conversationId).catch((error) => {
                setLocalError((prev) => prev || (error instanceof Error ? error.message : "Failed to resume conversation"));
            });
        }
        catch (err) {
            setLocalError(err instanceof Error ? err.message : "Failed to resolve approval");
            throw err;
        }
    }, [client, enabled, loadConversationMessages, scope.podId, sessionResumeIfRunning]);
    const { pendingActions, completedActions } = useMemo(() => {
        const pending = [];
        const completed = [];
        messages.forEach((message) => {
            if (!message.toolInvocations)
                return;
            message.toolInvocations.forEach((toolInvocation) => {
                const status = toolInvocation.state === "result"
                    ? (toolInvocation.result?.success === false ? "failed" : "completed")
                    : "executing";
                const action = {
                    id: toolInvocation.toolCallId,
                    type: "tool_call",
                    status,
                    toolName: toolInvocation.toolName,
                    toolArgs: toolInvocation.args,
                    result: toolInvocation.result,
                    timestamp: message.createdAt || new Date(),
                };
                if (status === "executing") {
                    pending.push(action);
                }
                else {
                    completed.push(action);
                }
            });
        });
        return { pendingActions: pending, completedActions: completed };
    }, [messages]);
    const isActiveConversationRunning = useMemo(() => {
        if (!activeConversationId)
            return false;
        const activeConversation = conversations.find((conversation) => conversation.id === activeConversationId);
        return isConversationRunning(activeConversation?.status);
    }, [activeConversationId, conversations]);
    return useMemo(() => ({
        messages,
        conversations,
        activeConversationId,
        availableModels,
        conversationModel,
        conversationRuntime,
        isActiveConversationRunning,
        isLoading,
        isLoadingConversations,
        isLoadingMoreConversations,
        hasMoreConversations: !!conversationsCursor,
        isLoadingMessages,
        isLoadingOlderMessages,
        hasOlderMessages: !!olderMessagesCursor,
        isUploadingFiles,
        pendingFiles,
        pendingFileUploads,
        error,
        pendingActions,
        completedActions,
        streamingTool: sessionStreamingTool,
        selectConversation,
        setConversationModel,
        sendMessage,
        uploadFiles,
        removePendingFile,
        clearPendingFiles,
        loadOlderMessages,
        loadMoreConversations,
        resolveUserApproval,
        clearMessages,
        stop,
    }), [
        activeConversationId,
        availableModels,
        clearMessages,
        clearPendingFiles,
        completedActions,
        conversationRuntime,
        conversationModel,
        conversations,
        error,
        conversationsCursor,
        isActiveConversationRunning,
        isLoading,
        isLoadingConversations,
        isLoadingMoreConversations,
        isLoadingMessages,
        isLoadingOlderMessages,
        isUploadingFiles,
        loadMoreConversations,
        loadOlderMessages,
        messages,
        olderMessagesCursor,
        pendingActions,
        pendingFileUploads,
        pendingFiles,
        removePendingFile,
        resolveUserApproval,
        selectConversation,
        sendMessage,
        sessionStreamingTool,
        setConversationModel,
        stop,
        uploadFiles,
    ]);
}
