// lemma-sdk/dist/react/useAgentTask.js
import { useCallback, useEffect, useMemo, useRef, useSyncExternalStore } from "react";
import { AgentController, selectAgentTask, } from "../core/agent/index.js";
function toControllerOptions(options) {
    return {
        client: options.client,
        scope: { podId: options.podId ?? null, agentName: options.agentName ?? null },
        instructions: options.instructions ?? null,
        onError: options.onError,
    };
}
export function useAgentTask(options) {
    const { parseOutput = true, title, instructions, metadata } = options;
    const controllerRef = useRef(null);
    if (controllerRef.current === null) {
        controllerRef.current = new AgentController(toControllerOptions(options));
    }
    const controller = controllerRef.current;
    useEffect(() => {
        controller.setOptions(toControllerOptions(options));
    });
    const state = useSyncExternalStore(controller.subscribe, controller.getState, controller.getState);
    useEffect(() => () => controller.destroy(), [controller]);
    const run = useCallback(async (input) => {
        const createInput = {
            title: title ?? null,
            instructions: instructions ?? undefined,
            metadata: metadata ?? null,
            setActive: true,
        };
        await controller.createConversation(createInput);
        const content = typeof input === "string" ? input : JSON.stringify(input);
        await controller.sendMessage(content);
    }, [controller, title, instructions, metadata]);
    const stop = useCallback(() => {
        controller.cancel();
        if (controller.getState().conversationId) {
            void controller.stop().catch(() => {
                // Best-effort server stop; the local stream is already aborted.
            });
        }
    }, [controller]);
    const reset = useCallback(() => {
        controller.setConversationId(null);
    }, [controller]);
    const task = useMemo(() => selectAgentTask(state, { parseOutput }), [state, parseOutput]);
    return { ...task, run, stop, reset };
}
