// lemma-sdk/dist/react/useAuth.js
import { useState, useEffect } from "react"; // peer dependency
/**
 * React hook for subscribing to Lemma auth state.
 *
 * Usage:
 *   const { isAuthenticated, isLoading, redirectToAuth } = useAuth(client);
 */
export function useAuth(client) {
    const [state, setState] = useState(client.auth.getState());
    useEffect(() => {
        // Subscribe to future state changes
        const unsubscribe = client.auth.subscribe((next) => setState(next));
        // If still in loading state, trigger the auth check
        if (state.status === "loading") {
            client.auth.checkAuth().catch(() => {
                // checkAuth already handles errors internally
            });
        }
        return unsubscribe;
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [client]);
    return {
        status: state.status,
        user: state.user,
        isLoading: state.status === "loading",
        isAuthenticated: state.status === "authenticated",
        redirectToAuth: (options) => client.auth.redirectToAuth(options),
    };
}
