// lemma-sdk/dist/react/useFlowSession.js
import { useEffect, useRef, useSyncExternalStore } from "react";
import { WorkflowController, } from "../core/workflow/index.js";
function toControllerOptions(options) {
    return {
        client: options.client,
        scope: { podId: options.podId ?? null, flowName: options.flowName ?? null },
        autoPoll: options.autoPoll ?? true,
        pollIntervalMs: options.pollIntervalMs ?? 2000,
        initialRunId: options.runId ?? null,
        onRun: options.onRun,
        onError: options.onError,
    };
}
export function useFlowSession(options) {
    const { runId: initialRunId = null } = options;
    const controllerRef = useRef(null);
    if (controllerRef.current === null) {
        controllerRef.current = new WorkflowController(toControllerOptions(options));
    }
    const controller = controllerRef.current;
    // Keep the controller's live options (client/scope/poll/callbacks) current.
    useEffect(() => {
        controller.setOptions(toControllerOptions(options));
    });
    const state = useSyncExternalStore(controller.subscribe, controller.getState, controller.getState);
    // Attach to the controlled runId (also kicks off refresh + polling).
    useEffect(() => {
        if (initialRunId) {
            controller.setRunId(initialRunId);
        }
    }, [controller, initialRunId]);
    // Stop polling and drop listeners on unmount.
    useEffect(() => () => controller.destroy(), [controller]);
    return {
        runId: state.runId,
        run: state.run,
        status: state.status,
        isPolling: state.isPolling,
        error: state.error,
        setRunId: controller.setRunId,
        start: controller.start,
        resume: controller.resume,
        refresh: controller.refresh,
        listHistory: controller.listHistory,
        cancel: controller.cancel,
    };
}
