// lemma-sdk/dist/react/useForeignKeyOptions.js
import { useCallback, useEffect, useMemo, useState } from "react";
import { parseForeignKeyReference } from "../datastore-query.js";
import { normalizeError, resolvePodId } from "./utils.js";
const EMPTY_LABEL_FIELDS = [];
function readRecordValue(record, field) {
    if (!field)
        return undefined;
    return record[field];
}
function pickResolvedLabelField(records, referenceColumn, explicitLabelField, explicitLabelFields) {
    const candidates = [
        explicitLabelField,
        ...(explicitLabelFields ?? []),
        "name",
        "title",
        "label",
        "email",
        "slug",
        referenceColumn,
        "id",
    ].filter((value) => typeof value === "string" && value.trim().length > 0);
    for (const candidate of candidates) {
        if (records.some((record) => {
            const value = record[candidate];
            return typeof value === "string"
                ? value.trim().length > 0
                : typeof value !== "undefined" && value !== null;
        })) {
            return candidate;
        }
    }
    return null;
}
function matchesSearch(record, search, fields) {
    const normalized = search.trim().toLowerCase();
    if (!normalized)
        return true;
    return fields.some((field) => {
        const value = record[field];
        if (typeof value === "string") {
            return value.toLowerCase().includes(normalized);
        }
        if (typeof value === "number" || typeof value === "boolean") {
            return String(value).toLowerCase().includes(normalized);
        }
        return false;
    });
}
export function useForeignKeyOptions({ client, podId, tableName, columnName, labelField, labelFields = EMPTY_LABEL_FIELDS, search, limit = 50, enabled = true, autoLoad = true, }) {
    const [table, setTable] = useState(null);
    const [column, setColumn] = useState(null);
    const [reference, setReference] = useState(null);
    const [resolvedLabelField, setResolvedLabelField] = useState(null);
    const [options, setOptions] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [error, setError] = useState(null);
    const labelFieldsKey = useMemo(() => JSON.stringify(labelFields), [labelFields]);
    const stableLabelFields = useMemo(() => labelFields, [labelFieldsKey]);
    const refresh = useCallback(async (signal) => {
        if (!enabled)
            return [];
        setIsLoading(true);
        setError(null);
        try {
            const resolvedPodId = resolvePodId(client, podId);
            const scopedClient = resolvedPodId === client.podId ? client : client.withPod(resolvedPodId);
            const nextTable = await scopedClient.tables.get(tableName);
            if (signal?.aborted)
                return [];
            const nextColumn = nextTable.columns.find((entry) => entry.name === columnName) ?? null;
            const nextReference = nextColumn?.foreign_key?.references
                ? parseForeignKeyReference(nextColumn.foreign_key.references)
                : null;
            setTable(nextTable);
            setColumn(nextColumn);
            setReference(nextReference);
            if (!nextColumn) {
                throw new Error(`Column "${columnName}" was not found on table "${tableName}".`);
            }
            if (!nextReference) {
                setResolvedLabelField(null);
                setOptions([]);
                return [];
            }
            const canFilterOnServer = !!labelField && !!search?.trim();
            const response = await scopedClient.records.list(nextReference.table, {
                limit: canFilterOnServer ? Math.max(limit, 100) : (search ? Math.max(limit * 5, 100) : limit),
                filters: canFilterOnServer
                    ? [{ field: labelField, op: "ilike", value: `%${search?.trim()}%` }]
                    : undefined,
            });
            if (signal?.aborted)
                return [];
            const records = response.items ?? [];
            const nextResolvedLabelField = pickResolvedLabelField(records, nextReference.column, labelField, stableLabelFields);
            const searchableFields = Array.from(new Set([nextResolvedLabelField, ...stableLabelFields, nextReference.column, "id"]
                .filter((value) => typeof value === "string" && value.trim().length > 0)));
            const filteredRecords = canFilterOnServer || !search?.trim()
                ? records
                : records.filter((record) => matchesSearch(record, search, searchableFields));
            const nextOptions = filteredRecords
                .slice(0, limit)
                .map((record) => {
                const value = readRecordValue(record, nextReference.column);
                const labelValue = readRecordValue(record, nextResolvedLabelField);
                return {
                    value,
                    label: typeof labelValue === "string" && labelValue.trim().length > 0
                        ? labelValue
                        : String(value ?? ""),
                    record,
                };
            })
                .filter((option) => typeof option.value !== "undefined" && option.value !== null);
            setResolvedLabelField(nextResolvedLabelField);
            setOptions(nextOptions);
            return nextOptions;
        }
        catch (refreshError) {
            if (signal?.aborted)
                return [];
            const normalized = normalizeError(refreshError, "Failed to load foreign key options.");
            setError(normalized);
            return [];
        }
        finally {
            if (!signal?.aborted)
                setIsLoading(false);
        }
    }, [client, columnName, enabled, labelField, limit, podId, search, stableLabelFields, tableName]);
    useEffect(() => {
        if (!enabled || !autoLoad)
            return;
        const controller = new AbortController();
        let cancelled = false;
        (async () => {
            try {
                await refresh(controller.signal);
            }
            catch {
                if (!cancelled) {
                    setError(normalizeError(new Error("Failed to load foreign key options."), "Failed to load foreign key options."));
                }
            }
        })();
        return () => {
            cancelled = true;
            controller.abort();
        };
    }, [autoLoad, enabled, refresh]);
    return useMemo(() => ({
        table,
        column,
        reference,
        labelField: resolvedLabelField,
        options,
        isLoading,
        error,
        refresh,
    }), [column, error, isLoading, options, reference, refresh, resolvedLabelField, table]);
}
