// lemma-sdk/dist/react/useGlobalSearch.js
import { useCallback, useEffect, useMemo, useState } from "react";
import { escapeSqlString, quoteIdentifierPath, renderIdentifierPath, renderRecordFilters } from "./sql-utils.js";
import { normalizeError, resolvePodClient, stringifyComparable } from "./utils.js";
function buildTableSearchQuery(source, query) {
    const fields = Array.from(new Set([
        "id",
        ...source.searchFields,
        source.displayField ?? "id",
        source.subtitleField ?? "",
    ].filter((value) => value.trim().length > 0)));
    const searchClauses = source.searchFields.map((field) => `${renderIdentifierPath(field)} ILIKE '%${escapeSqlString(query)}%'`);
    const filterClause = renderRecordFilters(source.filters);
    return [
        `SELECT ${fields.map((field) => renderIdentifierPath(field)).join(", ")}`,
        `FROM ${quoteIdentifierPath(source.tableName)}`,
        `WHERE (${searchClauses.join(" OR ")})${filterClause ? ` AND ${filterClause}` : ""}`,
        `LIMIT ${source.limit ?? 8}`,
    ].join(" ");
}
function readString(value) {
    if (typeof value === "string" && value.trim().length > 0)
        return value;
    if (typeof value === "number" || typeof value === "boolean")
        return String(value);
    return null;
}
export function useGlobalSearch({ client, podId, query = "", tables = [], files, enabled = true, autoLoad = true, minQueryLength = 1, }) {
    const [results, setResults] = useState([]);
    const [sourceErrors, setSourceErrors] = useState({});
    const [isLoading, setIsLoading] = useState(false);
    const [error, setError] = useState(null);
    const trimmedQuery = query.trim();
    const tablesKey = stringifyComparable(tables);
    const filesKey = stringifyComparable(files);
    const stableTables = useMemo(() => tables, [tablesKey]);
    const stableFiles = useMemo(() => files, [filesKey]);
    const reset = useCallback(() => {
        setResults([]);
        setSourceErrors({});
        setError(null);
        setIsLoading(false);
    }, []);
    const search = useCallback(async (overrides = {}, signal) => {
        const nextQuery = (overrides.query ?? trimmedQuery).trim();
        if (!enabled || nextQuery.length < minQueryLength) {
            reset();
            return [];
        }
        setIsLoading(true);
        setError(null);
        try {
            const scopedClient = resolvePodClient(client, podId);
            const searchTasks = [];
            stableTables.forEach((source, index) => {
                const sourceKey = source.key?.trim() || source.tableName;
                searchTasks.push((async () => {
                    try {
                        const sql = buildTableSearchQuery(source, nextQuery);
                        const response = await scopedClient.datastore.query(sql);
                        const rows = response.items ?? [];
                        const mapped = rows.map((record) => {
                            const displayField = source.displayField ?? source.searchFields[0] ?? "id";
                            const subtitleField = source.subtitleField;
                            const id = readString(record.id) ?? `${source.tableName}-${index}`;
                            return {
                                kind: "record",
                                sourceKey,
                                sourceLabel: source.label ?? source.tableName,
                                tableName: source.tableName,
                                id,
                                title: readString(record[displayField]) ?? id,
                                subtitle: subtitleField ? readString(record[subtitleField]) : null,
                                record,
                            };
                        });
                        return { key: sourceKey, results: mapped };
                    }
                    catch (tableError) {
                        return {
                            key: sourceKey,
                            results: [],
                            error: normalizeError(tableError, `Failed to search ${source.tableName}.`),
                        };
                    }
                })());
            });
            if (stableFiles !== false && stableFiles?.enabled !== false) {
                searchTasks.push((async () => {
                    const sourceKey = "files";
                    try {
                        const response = await scopedClient.files.search(nextQuery, {
                            limit: stableFiles?.limit ?? 8,
                            searchMethod: stableFiles?.searchMethod,
                        });
                        const mapped = (response.items ?? []).map((result) => ({
                            kind: "file",
                            sourceKey,
                            sourceLabel: stableFiles?.label ?? "Files",
                            path: result.path,
                            title: result.path.split("/").filter(Boolean).pop() || result.path,
                            subtitle: typeof result.content === "string" && result.content.trim().length > 0
                                ? result.content.slice(0, 160)
                                : null,
                            result,
                        }));
                        return { key: sourceKey, results: mapped };
                    }
                    catch (fileError) {
                        return {
                            key: sourceKey,
                            results: [],
                            error: normalizeError(fileError, "Failed to search files."),
                        };
                    }
                })());
            }
            const settled = await Promise.all(searchTasks);
            if (signal?.aborted)
                return [];
            const nextErrors = {};
            const nextResults = settled.flatMap((entry) => {
                if (entry.error) {
                    nextErrors[entry.key] = entry.error;
                }
                return entry.results;
            });
            setSourceErrors(nextErrors);
            setResults(nextResults);
            return nextResults;
        }
        catch (searchError) {
            if (signal?.aborted)
                return [];
            const normalized = normalizeError(searchError, "Failed to run global search.");
            setError(normalized);
            setResults([]);
            return [];
        }
        finally {
            if (!signal?.aborted)
                setIsLoading(false);
        }
    }, [client, enabled, minQueryLength, podId, reset, stableFiles, stableTables, trimmedQuery]);
    useEffect(() => {
        if (!enabled || !autoLoad)
            return;
        if (trimmedQuery.length < minQueryLength) {
            reset();
            return;
        }
        const controller = new AbortController();
        void search({}, controller.signal);
        return () => controller.abort();
    }, [autoLoad, enabled, minQueryLength, reset, search, trimmedQuery]);
    return useMemo(() => {
        const recordResults = results.filter((entry) => entry.kind === "record");
        const fileResults = results.filter((entry) => entry.kind === "file");
        return {
            results,
            recordResults,
            fileResults,
            totalResults: results.length,
            sourceErrors,
            isLoading,
            error,
            search,
            reset,
        };
    }, [error, isLoading, reset, results, search, sourceErrors]);
}
