// lemma-sdk/dist/react/useJoinedRecords.js
import { useCallback, useEffect, useMemo, useState } from "react";
import { buildJoinedRecordsQuery, parseForeignKeyReference, } from "../datastore-query.js";
import { normalizeError, resolvePodId, stringifyComparable } from "./utils.js";
async function buildShorthandQuery(client, podId, baseTableName, shorthandJoins) {
    const resolvedPodId = resolvePodId(client, podId);
    const scopedClient = resolvedPodId === client.podId ? client : client.withPod(resolvedPodId);
    const baseTable = await scopedClient.tables.get(baseTableName.trim());
    const joins = await Promise.all(shorthandJoins.map(async (entry) => {
        const baseColumn = baseTable.columns.find((col) => col.name === entry.on) ?? null;
        const reference = baseColumn?.foreign_key?.references
            ? parseForeignKeyReference(baseColumn.foreign_key.references)
            : null;
        if (!baseColumn) {
            throw new Error(`Column "${entry.on}" was not found on table "${baseTableName}".`);
        }
        if (!reference) {
            throw new Error(`Column "${entry.on}" on "${baseTableName}" is not a foreign key. Use the full query API for non-FK joins.`);
        }
        return {
            type: entry.type ?? "left",
            table: entry.table,
            alias: entry.alias,
            on: {
                left: { table: baseTableName, column: entry.on },
                right: { table: entry.alias ?? entry.table, column: reference.column },
            },
        };
    }));
    return {
        from: { table: baseTableName, alias: baseTableName },
        joins,
    };
}
export function useJoinedRecords({ client, podId, query, baseTable, joins: shorthandJoins, enabled = true, autoLoad = true, }) {
    const [records, setRecords] = useState([]);
    const [total, setTotal] = useState(0);
    const [sql, setSql] = useState("");
    const [isLoading, setIsLoading] = useState(false);
    const [error, setError] = useState(null);
    const hasShorthand = !query && !!baseTable && !!shorthandJoins?.length;
    const queryKey = stringifyComparable(query);
    const shorthandKey = stringifyComparable({ baseTable, shorthandJoins });
    const stableQuery = useMemo(() => query, [queryKey]);
    const refresh = useCallback(async (signal) => {
        if (!enabled) {
            setRecords([]);
            setTotal(0);
            setSql("");
            setError(null);
            setIsLoading(false);
            return [];
        }
        setIsLoading(true);
        setError(null);
        try {
            const resolvedPodId = resolvePodId(client, podId);
            const scopedClient = resolvedPodId === client.podId ? client : client.withPod(resolvedPodId);
            let resolvedQuery = stableQuery;
            if (hasShorthand && baseTable && shorthandJoins) {
                resolvedQuery = await buildShorthandQuery(client, podId, baseTable, shorthandJoins);
            }
            if (!resolvedQuery) {
                setRecords([]);
                setTotal(0);
                setSql("");
                setIsLoading(false);
                return [];
            }
            if (signal?.aborted)
                return [];
            const nextSql = buildJoinedRecordsQuery(resolvedQuery);
            setSql(nextSql);
            const response = await scopedClient.datastore.query(nextSql);
            if (signal?.aborted)
                return [];
            const nextRecords = (response.items ?? []);
            setRecords(nextRecords);
            setTotal(response.total ?? nextRecords.length);
            return nextRecords;
        }
        catch (refreshError) {
            if (signal?.aborted)
                return [];
            const normalized = normalizeError(refreshError, "Failed to load joined records.");
            setError(normalized);
            return [];
        }
        finally {
            if (!signal?.aborted)
                setIsLoading(false);
        }
    }, [client, enabled, hasShorthand, baseTable, podId, shorthandJoins, stableQuery]);
    useEffect(() => {
        if (!enabled) {
            setRecords([]);
            setTotal(0);
            setSql("");
            setError(null);
            setIsLoading(false);
            return;
        }
        if (!autoLoad)
            return;
        const controller = new AbortController();
        let cancelled = false;
        (async () => {
            try {
                await refresh(controller.signal);
            }
            catch {
                if (!cancelled) {
                    setError(normalizeError(new Error("Failed to load joined records."), "Failed to load joined records."));
                }
            }
        })();
        return () => {
            cancelled = true;
            controller.abort();
        };
    }, [autoLoad, enabled, refresh]);
    return useMemo(() => ({
        records,
        total,
        sql,
        isLoading,
        error,
        refresh,
    }), [error, isLoading, records, refresh, sql, total]);
}
