// lemma-sdk/dist/react/useLiveRecords.js
import { useEffect, useMemo, useRef, useState } from "react";
import { useRecords } from "./useRecords.js";
import { useWatchChanges } from "./useWatchChanges.js";
import { applyDatastoreChange, makeRecordComparator } from "./datastoreChangeReducer.js";
import { stringifyComparable } from "./utils.js";
/**
 * A record list that stays live without polling: an initial fetch (via `useRecords`,
 * with the same filters/sort/pagination) plus a WebSocket subscription that merges
 * insert/update/delete deltas **in place**, keyed by id. Rows update without a
 * full-list refetch, so the UI doesn't flicker and the API isn't hammered.
 *
 * ```tsx
 * const { records, isLoading, liveStatus } = useLiveRecords({ client, tableName: "tickets" });
 * return <ul>{records.map((r) => <li key={r.id as string}>{r.title as string}</li>)}</ul>;
 * ```
 *
 * Filtered views: the stream carries every change you can see, so pass `accept` (a
 * client-side predicate) or `reconcile: "refetch"` to keep a filtered list exact.
 */
export function useLiveRecords(options) {
    const { live = true, reconcile = "merge", idKey = "id", accept, ...recordsOptions } = options;
    const base = useRecords(recordsOptions);
    // Local view of the rows: seeded from the base fetch, then kept live by the stream.
    const [rows, setRows] = useState(base.records);
    // Re-seed whenever the base list is replaced (initial load, refresh, loadMore,
    // filter/sort change). Live merges below touch only `rows`, never `base.records`,
    // so they are not clobbered between refetches.
    useEffect(() => {
        setRows(base.records);
    }, [base.records]);
    const sortKey = stringifyComparable(recordsOptions.sort);
    const compare = useMemo(() => makeRecordComparator(recordsOptions.sort), 
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [sortKey]);
    // Debounced refetch for reconcile:"refetch" — coalesces change bursts into one query.
    const refreshRef = useRef(base.refresh);
    refreshRef.current = base.refresh;
    const timerRef = useRef(null);
    useEffect(() => () => {
        if (timerRef.current)
            clearTimeout(timerRef.current);
    }, []);
    const { status } = useWatchChanges({
        client: recordsOptions.client,
        podId: recordsOptions.podId,
        table: recordsOptions.tableName,
        enabled: live && (recordsOptions.enabled ?? true),
        onChange: (frame) => {
            if (reconcile === "refetch") {
                if (timerRef.current)
                    clearTimeout(timerRef.current);
                timerRef.current = setTimeout(() => {
                    void refreshRef.current();
                }, 150);
                return;
            }
            setRows((current) => applyDatastoreChange(current, frame, { idKey, accept, compare }));
        },
    });
    return useMemo(() => ({ ...base, records: rows, liveStatus: status }), [base, rows, status]);
}
