// lemma-sdk/dist/react/useReferencingRecords.js
import { useCallback, useEffect, useMemo, useState } from "react";
import { normalizeError, resolvePodClient, stringifyComparable } from "./utils.js";
function sentenceCase(value) {
    return value
        .replace(/[_\.]/g, " ")
        .replace(/\s+/g, " ")
        .trim()
        .replace(/\b\w/g, (match) => match.toUpperCase());
}
function pickDefaultFields(table, foreignKey) {
    const names = table.columns
        .map((column) => column.name)
        .filter((name) => name !== "created_at" && name !== "updated_at");
    const prioritized = ["id", "name", "title", "label", "status", foreignKey];
    const next = [];
    prioritized.forEach((name) => {
        if (names.includes(name) && !next.includes(name)) {
            next.push(name);
        }
    });
    names.forEach((name) => {
        if (!next.includes(name)) {
            next.push(name);
        }
    });
    return next.slice(0, 6);
}
export function useReferencingRecords({ client, podId, table, foreignKey, recordId = null, fields, limit = 20, offset, sort, enabled = true, autoLoad = true, }) {
    const [referencedTable, setReferencedTable] = useState(null);
    const [columns, setColumns] = useState([]);
    const [records, setRecords] = useState([]);
    const [total, setTotal] = useState(0);
    const [nextPageToken, setNextPageToken] = useState(null);
    const [isLoading, setIsLoading] = useState(false);
    const [error, setError] = useState(null);
    const trimmedTable = table.trim();
    const trimmedRecordId = typeof recordId === "string" ? recordId.trim() : "";
    const fieldsKey = stringifyComparable(fields);
    const sortKey = stringifyComparable(sort);
    const stableFields = useMemo(() => fields, [fieldsKey]);
    const stableSort = useMemo(() => sort, [sortKey]);
    const isEnabled = enabled && trimmedTable.length > 0 && trimmedRecordId.length > 0;
    const refresh = useCallback(async (signal) => {
        if (!isEnabled) {
            setReferencedTable(null);
            setColumns([]);
            setRecords([]);
            setTotal(0);
            setNextPageToken(null);
            setError(null);
            setIsLoading(false);
            return [];
        }
        setIsLoading(true);
        setError(null);
        try {
            const scopedClient = resolvePodClient(client, podId);
            const tableResponse = await scopedClient.tables.get(trimmedTable);
            if (signal?.aborted)
                return [];
            setReferencedTable(tableResponse);
            const resolvedFields = (stableFields?.length ? stableFields : pickDefaultFields(tableResponse, foreignKey))
                .filter((field, index, allFields) => field.trim().length > 0 && allFields.indexOf(field) === index);
            const response = await scopedClient.records.list(trimmedTable, {
                filters: [{
                        field: foreignKey,
                        op: "eq",
                        value: trimmedRecordId,
                    }],
                limit,
                offset,
                sort: stableSort,
            });
            if (signal?.aborted)
                return [];
            const nextRecords = (response.items ?? []);
            setColumns(resolvedFields.map((field) => ({
                key: field,
                field,
                label: sentenceCase(field),
            })));
            setRecords(nextRecords);
            setTotal(response.total ?? nextRecords.length);
            setNextPageToken(response.next_page_token ?? null);
            return nextRecords;
        }
        catch (refreshError) {
            if (signal?.aborted)
                return [];
            const normalized = normalizeError(refreshError, "Failed to load referencing records.");
            setError(normalized);
            return [];
        }
        finally {
            if (!signal?.aborted)
                setIsLoading(false);
        }
    }, [
        client,
        foreignKey,
        isEnabled,
        limit,
        offset,
        podId,
        stableFields,
        stableSort,
        trimmedRecordId,
        trimmedTable,
    ]);
    useEffect(() => {
        if (!isEnabled) {
            setReferencedTable(null);
            setColumns([]);
            setRecords([]);
            setTotal(0);
            setNextPageToken(null);
            setError(null);
            setIsLoading(false);
            return;
        }
        if (!autoLoad)
            return;
        const controller = new AbortController();
        let cancelled = false;
        (async () => {
            try {
                await refresh(controller.signal);
            }
            catch {
                if (!cancelled) {
                    setError(normalizeError(new Error("Failed to load referencing records."), "Failed to load referencing records."));
                }
            }
        })();
        return () => {
            cancelled = true;
            controller.abort();
        };
    }, [autoLoad, isEnabled, refresh]);
    return useMemo(() => ({
        referencedTable,
        columns,
        records,
        total,
        nextPageToken,
        isLoading,
        error,
        refresh,
    }), [
        columns,
        error,
        isLoading,
        nextPageToken,
        records,
        refresh,
        referencedTable,
        total,
    ]);
}
