// lemma-sdk/dist/react/useRecordForm.js
import { useCallback, useEffect, useMemo, useState } from "react";
import { buildRecordFormValues, buildRecordPayload, } from "../record-form.js";
import { normalizeError, resolvePodClient, stringifyComparable } from "./utils.js";
import { useRecordSchema, } from "./useRecordSchema.js";
const EMPTY_VALUES = {};
export function useRecordForm({ client, podId, tableName, recordId = null, initialValues = EMPTY_VALUES, mode = "auto", enabled = true, autoLoad = true, visibleFields, hiddenFields, submitVia = "direct", submitFunctionName, submitFunctionInput, onSubmitSuccess, onError, }) {
    const schema = useRecordSchema({
        client,
        podId,
        tableName,
        enabled,
        autoLoad,
    });
    const visibleSet = useMemo(() => visibleFields ? new Set(visibleFields) : null, [visibleFields]);
    const hiddenSet = useMemo(() => hiddenFields ? new Set(hiddenFields) : null, [hiddenFields]);
    const filteredFields = useMemo(() => {
        if (!visibleSet && !hiddenSet)
            return schema.fields;
        return schema.fields.filter((field) => {
            if (visibleSet)
                return visibleSet.has(field.name);
            if (hiddenSet)
                return !hiddenSet.has(field.name);
            return true;
        });
    }, [hiddenSet, schema.fields, visibleSet]);
    const filteredEditableFields = useMemo(() => {
        if (!visibleSet && !hiddenSet)
            return schema.editableFields;
        return schema.editableFields.filter((field) => {
            if (visibleSet)
                return visibleSet.has(field.name);
            if (hiddenSet)
                return !hiddenSet.has(field.name);
            return true;
        });
    }, [hiddenSet, schema.editableFields, visibleSet]);
    const [record, setRecord] = useState(null);
    const [values, setValuesState] = useState({});
    const [baselineValues, setBaselineValues] = useState({});
    const [fieldErrors, setFieldErrors] = useState({});
    const [recordError, setRecordError] = useState(null);
    const [isLoadingRecord, setIsLoadingRecord] = useState(false);
    const [isSubmitting, setIsSubmitting] = useState(false);
    const initialValuesKey = stringifyComparable(initialValues);
    const stableInitialValues = useMemo(() => initialValues, [initialValuesKey]);
    const schemaTable = schema.table;
    const refreshSchema = schema.refresh;
    const hydrateValues = useCallback((source) => {
        if (!schemaTable)
            return;
        const nextValues = buildRecordFormValues(schemaTable, source);
        setValuesState(nextValues);
        setBaselineValues(nextValues);
        setFieldErrors({});
    }, [schemaTable]);
    const refreshRecord = useCallback(async () => {
        if (!enabled || !recordId) {
            setRecord(null);
            return null;
        }
        setIsLoadingRecord(true);
        setRecordError(null);
        try {
            const scopedClient = resolvePodClient(client, podId);
            const response = await scopedClient.records.get(tableName, recordId);
            const nextRecord = response ?? null;
            setRecord(nextRecord);
            return nextRecord;
        }
        catch (refreshError) {
            const normalized = normalizeError(refreshError, "Failed to load record.");
            setRecordError(normalized);
            onError?.(refreshError);
            return null;
        }
        finally {
            setIsLoadingRecord(false);
        }
    }, [client, enabled, onError, podId, recordId, tableName]);
    useEffect(() => {
        if (!enabled) {
            setRecord(null);
            setValuesState({});
            setBaselineValues({});
            setFieldErrors({});
            setRecordError(null);
            setIsLoadingRecord(false);
            setIsSubmitting(false);
            return;
        }
        if (!autoLoad || !recordId)
            return;
        void refreshRecord();
    }, [autoLoad, enabled, recordId, refreshRecord]);
    const recordKey = stringifyComparable(record);
    useEffect(() => {
        if (!schemaTable)
            return;
        const source = {
            ...(record ?? {}),
            ...stableInitialValues,
        };
        hydrateValues(source);
    }, [hydrateValues, recordKey, schemaTable, stableInitialValues]);
    const refresh = useCallback(async () => {
        const nextTable = await refreshSchema();
        if (recordId) {
            const nextRecord = await refreshRecord();
            if (nextTable && nextRecord) {
                hydrateValues({
                    ...nextRecord,
                    ...stableInitialValues,
                });
            }
            return;
        }
        if (nextTable) {
            hydrateValues(stableInitialValues);
        }
    }, [hydrateValues, recordId, refreshRecord, refreshSchema, stableInitialValues]);
    const setValue = useCallback((fieldName, value) => {
        setValuesState((current) => ({
            ...current,
            [fieldName]: value,
        }));
        setFieldErrors((current) => {
            if (!(fieldName in current))
                return current;
            const next = { ...current };
            delete next[fieldName];
            return next;
        });
    }, []);
    const setValues = useCallback((nextValues) => {
        setValuesState((current) => ({
            ...current,
            ...nextValues,
        }));
    }, []);
    const reset = useCallback((nextValues) => {
        if (!schemaTable)
            return;
        const source = nextValues ?? {
            ...(record ?? {}),
            ...stableInitialValues,
        };
        hydrateValues(source);
    }, [hydrateValues, record, schemaTable, stableInitialValues]);
    const validate = useCallback(() => {
        if (!schemaTable)
            return false;
        const resolvedMode = mode === "auto" ? (recordId ? "update" : "create") : mode;
        const result = buildRecordPayload(schemaTable, values, { mode: resolvedMode });
        setFieldErrors(result.errors);
        return result.isValid;
    }, [mode, recordId, schemaTable, values]);
    const submit = useCallback(async (overrides = {}) => {
        if (!schemaTable) {
            const error = new Error("Record schema is not loaded.");
            setRecordError(error);
            return null;
        }
        const resolvedMode = overrides.mode ?? (mode === "auto" ? (recordId ? "update" : "create") : mode);
        const payload = buildRecordPayload(schemaTable, values, { mode: resolvedMode });
        setFieldErrors(payload.errors);
        if (!payload.isValid) {
            return null;
        }
        setIsSubmitting(true);
        setRecordError(null);
        try {
            const scopedClient = resolvePodClient(client, podId);
            if (submitVia === "function") {
                const functionName = submitFunctionName ?? tableName;
                const functionInput = submitFunctionInput ? submitFunctionInput(payload.data) : payload.data;
                const run = await scopedClient.functions.runs.create(functionName, { input: functionInput });
                const nextRecord = run.output_data ?? { id: run.id, ...functionInput };
                setRecord(nextRecord);
                hydrateValues({
                    ...(nextRecord ?? {}),
                    ...stableInitialValues,
                });
                onSubmitSuccess?.(nextRecord ?? {}, run);
                return nextRecord;
            }
            const response = resolvedMode === "update" && recordId
                ? await scopedClient.records.update(tableName, recordId, payload.data)
                : await scopedClient.records.create(tableName, payload.data);
            const nextRecord = response ?? null;
            setRecord(nextRecord);
            hydrateValues({
                ...(nextRecord ?? {}),
                ...stableInitialValues,
            });
            onSubmitSuccess?.(nextRecord ?? {}, response);
            return nextRecord;
        }
        catch (submitError) {
            const normalized = normalizeError(submitError, "Failed to save record.");
            setRecordError(normalized);
            onError?.(submitError);
            return null;
        }
        finally {
            setIsSubmitting(false);
        }
    }, [client, hydrateValues, mode, onError, onSubmitSuccess, podId, recordId, schemaTable, stableInitialValues, submitFunctionInput, submitFunctionName, submitVia, tableName, values]);
    const isDirty = useMemo(() => {
        return stringifyComparable(values) !== stringifyComparable(baselineValues);
    }, [baselineValues, values]);
    return useMemo(() => ({
        table: schema.table,
        fields: filteredFields,
        editableFields: filteredEditableFields,
        defaults: schema.defaults,
        values,
        baselineValues,
        record,
        fieldErrors,
        isLoadingSchema: schema.isLoading,
        isLoadingRecord,
        isSubmitting,
        isDirty,
        error: schema.error ?? recordError,
        refreshSchema: schema.refresh,
        refreshRecord,
        refresh,
        setValue,
        setValues,
        reset,
        validate,
        submit,
    }), [
        baselineValues,
        fieldErrors,
        filteredEditableFields,
        filteredFields,
        isDirty,
        isLoadingRecord,
        isSubmitting,
        record,
        recordError,
        refresh,
        refreshRecord,
        reset,
        schema.defaults,
        schema.error,
        schema.isLoading,
        schema.refresh,
        schema.table,
        setValue,
        setValues,
        submit,
        validate,
        values,
    ]);
}
