// lemma-sdk/dist/react/useWorkflowRun.js
import { useCallback, useMemo } from "react";
import { selectWorkflowOutputs } from "../core/workflow/index.js";
import { useFlowSession, } from "./useFlowSession.js";
function resolveWorkflowName(base, override) {
    const resolved = override ?? base;
    if (!resolved) {
        throw new Error("workflowName is required.");
    }
    return resolved;
}
export function useWorkflowRun({ workflowName, ...options }) {
    const session = useFlowSession({
        ...options,
        flowName: workflowName,
    });
    const start = useCallback(async (inputs, startOptions = {}) => {
        return session.start({
            flowName: resolveWorkflowName(workflowName, startOptions.workflowName),
            inputs,
            connect: startOptions.connect,
        });
    }, [session, workflowName]);
    const listRuns = useCallback(async (listOptions = {}) => {
        return session.listHistory({
            flowName: resolveWorkflowName(workflowName, listOptions.workflowName),
            limit: listOptions.limit,
            pageToken: listOptions.pageToken,
        });
    }, [session, workflowName]);
    return useMemo(() => {
        const { isFinished, isWaitingForInput, output, finalOutput } = selectWorkflowOutputs(session.run);
        return {
            ...session,
            output,
            finalOutput,
            isWaitingForInput,
            isFinished,
            start,
            listRuns,
        };
    }, [listRuns, session, start]);
}
